{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "I need to remake the variation excel sheet so that it creates distinct lines for each type of mutation \n",
    "and the location that it affects. So if it is intergenic, I have a line for upstream and the downstream gene\n",
    "or if its in the promoter of one gene and in the coding sequence of another these get seperate lines.\n",
    "I will need to manually add some things like whether or not the gene is affected or not by the mutation type\n",
    "whether the innoculum has it \n",
    "\n",
    "then for every line I make a promoter region, an downstream region and coding region primers. \n",
    "\n",
    "if there is a similar mutation in the region, then the same effect will be predicted e.g. if there \n",
    "is an insertion of % 3 != 0 in the innoculum within the coding region, then the gene is off \n",
    "if there is not, then the gene is back ON (as in the reference)\n",
    "\n",
    "if there is a different mutation, i will need to interpret it on the fly i.e. and snv in the coding region will need to be compared by in-silico translation whereas an indel (%3!=0) can be assumed to cause a frameshift, where as ANYthing in a promoter can be interpreted as an \"unkown\" effect \n",
    "\n",
    "I then need to determine whether the homolog in the peruvian genome corresponds to a known gene \n",
    "I will need to add the predicted length of the protein and compare it to original \n",
    "\n",
    "then I need to go back and determine whether there reference version is functional or not \n",
    "\n",
    "\n",
    "BUG TO DO: in Crofts_Relapse_wo_Region for pseD is incorrect? \n",
    "\n",
    "FEATURE TO ADD: populations expected not to have the wild type \n",
    "\n",
    "Feature to add: see if existing clusters homologs in the abundances in which we detect them \n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### The purpose of this notebook is to analyze the variants reported in Crofts et al. It seeks to reproduce their results with their own data, as well as in recrudescent strains collected from community acquired infections.\n",
    "\n",
    "#### In this first block, we will simply import the packages we need to analyze the data."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import os\n",
    "from Bio import SeqIO"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### We will ultimately call SNPs using the library of a relapsed isolate mapped against the draft assembly of a paired primary infection. This differs from the SNP calling pipeline used in the Crofts et al paper becuase they call SNPs against a complete reference genome. To gauge the posibility for false positives and negatives, we simulated a draft assembly of their complete genome using `wgsim`. In this step, we will examine its quality using BLASTn.\n",
    "\n",
    "#### From this we can see that the reference genome fully covered identically by the assembly. We can also see that there are a number of smaller contigs that are repeatedly aligned with the reference at different points. Some of these repeats stich over portion of the assembly that would be otherwised uncovered (eg. contigs 10 and 12) but others map to regions that are already covered (eg. contig 18). The latter implies that there may be a gene that is repeated, but which has diverged internally, which can aberrations in coverage in the region. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "OUTFMT=\"6 qseqid sseqid pident length qcovs mismatch gapopen qstart qend qlen sstart send slen evalue bitscore sstrand\"\n",
      "\n",
      "makeblastdb -in /Volumes/KeithSSD/variant-calling/ref_fastas/CG8421_bb.fa -dbtype nucl -out /Volumes/KeithSSD/variant-calling/simulated_reference/complete_ref_db\n",
      "\n",
      "blastn -evalue 1e-3 -db /Volumes/KeithSSD/variant-calling/simulated_reference/complete_ref_db -query /Volumes/KeithSSD/variant-calling/simulated_reference/CG8421_sim_contigs_fixed.fa -outfmt \"$OUTFMT\" -qcov_hsp_perc 90 -out /Volumes/KeithSSD/variant-calling/simulated_reference/simulation_matches.txt -num_threads 3\n",
      "\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>primer</th>\n",
       "      <th>sseqid</th>\n",
       "      <th>pident</th>\n",
       "      <th>length</th>\n",
       "      <th>qcov</th>\n",
       "      <th>mismatch</th>\n",
       "      <th>gapopen</th>\n",
       "      <th>qstart</th>\n",
       "      <th>qend</th>\n",
       "      <th>qlen</th>\n",
       "      <th>sstart2</th>\n",
       "      <th>send2</th>\n",
       "      <th>sstrand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>c_000000000008</td>\n",
       "      <td>CG8421-NZ_CP005388-x</td>\n",
       "      <td>100.000</td>\n",
       "      <td>38777</td>\n",
       "      <td>100</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>38777</td>\n",
       "      <td>38777</td>\n",
       "      <td>1</td>\n",
       "      <td>38777</td>\n",
       "      <td>plus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>c_000000000010</td>\n",
       "      <td>CG8421-NZ_CP005388-x</td>\n",
       "      <td>99.834</td>\n",
       "      <td>6007</td>\n",
       "      <td>100</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>6007</td>\n",
       "      <td>6007</td>\n",
       "      <td>38678</td>\n",
       "      <td>44674</td>\n",
       "      <td>plus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>c_000000000005</td>\n",
       "      <td>CG8421-NZ_CP005388-x</td>\n",
       "      <td>100.000</td>\n",
       "      <td>104606</td>\n",
       "      <td>100</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>104606</td>\n",
       "      <td>104606</td>\n",
       "      <td>44548</td>\n",
       "      <td>149153</td>\n",
       "      <td>minus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>c_000000000012</td>\n",
       "      <td>CG8421-NZ_CP005388-x</td>\n",
       "      <td>100.000</td>\n",
       "      <td>1050</td>\n",
       "      <td>100</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1050</td>\n",
       "      <td>1050</td>\n",
       "      <td>149027</td>\n",
       "      <td>150076</td>\n",
       "      <td>minus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>c_000000000007</td>\n",
       "      <td>CG8421-NZ_CP005388-x</td>\n",
       "      <td>100.000</td>\n",
       "      <td>89457</td>\n",
       "      <td>100</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>89457</td>\n",
       "      <td>89457</td>\n",
       "      <td>150077</td>\n",
       "      <td>239533</td>\n",
       "      <td>minus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>c_000000000012</td>\n",
       "      <td>CG8421-NZ_CP005388-x</td>\n",
       "      <td>100.000</td>\n",
       "      <td>1050</td>\n",
       "      <td>100</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1050</td>\n",
       "      <td>1050</td>\n",
       "      <td>239534</td>\n",
       "      <td>240583</td>\n",
       "      <td>plus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>c_000000000004</td>\n",
       "      <td>CG8421-NZ_CP005388-x</td>\n",
       "      <td>100.000</td>\n",
       "      <td>153010</td>\n",
       "      <td>100</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>153010</td>\n",
       "      <td>153010</td>\n",
       "      <td>240457</td>\n",
       "      <td>393466</td>\n",
       "      <td>plus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>c_000000000010</td>\n",
       "      <td>CG8421-NZ_CP005388-x</td>\n",
       "      <td>99.917</td>\n",
       "      <td>6009</td>\n",
       "      <td>100</td>\n",
       "      <td>3</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>6007</td>\n",
       "      <td>6007</td>\n",
       "      <td>393340</td>\n",
       "      <td>399348</td>\n",
       "      <td>plus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>c_000000000002</td>\n",
       "      <td>CG8421-NZ_CP005388-x</td>\n",
       "      <td>100.000</td>\n",
       "      <td>287092</td>\n",
       "      <td>100</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>287092</td>\n",
       "      <td>287092</td>\n",
       "      <td>399286</td>\n",
       "      <td>686377</td>\n",
       "      <td>plus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>c_000000000010</td>\n",
       "      <td>CG8421-NZ_CP005388-x</td>\n",
       "      <td>100.000</td>\n",
       "      <td>6007</td>\n",
       "      <td>100</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>6007</td>\n",
       "      <td>6007</td>\n",
       "      <td>686251</td>\n",
       "      <td>692257</td>\n",
       "      <td>plus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>c_000000000001</td>\n",
       "      <td>CG8421-NZ_CP005388-x</td>\n",
       "      <td>100.000</td>\n",
       "      <td>565758</td>\n",
       "      <td>100</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>565758</td>\n",
       "      <td>565758</td>\n",
       "      <td>692131</td>\n",
       "      <td>1257888</td>\n",
       "      <td>plus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>c_000000000014</td>\n",
       "      <td>CG8421-NZ_CP005388-x</td>\n",
       "      <td>99.814</td>\n",
       "      <td>539</td>\n",
       "      <td>100</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>539</td>\n",
       "      <td>539</td>\n",
       "      <td>1257762</td>\n",
       "      <td>1258300</td>\n",
       "      <td>minus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>c_000000000017</td>\n",
       "      <td>CG8421-NZ_CP005388-x</td>\n",
       "      <td>100.000</td>\n",
       "      <td>176</td>\n",
       "      <td>100</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>176</td>\n",
       "      <td>176</td>\n",
       "      <td>1258182</td>\n",
       "      <td>1258357</td>\n",
       "      <td>plus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>c_000000000016</td>\n",
       "      <td>CG8421-NZ_CP005388-x</td>\n",
       "      <td>100.000</td>\n",
       "      <td>254</td>\n",
       "      <td>100</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>254</td>\n",
       "      <td>254</td>\n",
       "      <td>1258231</td>\n",
       "      <td>1258484</td>\n",
       "      <td>minus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>c_000000000018</td>\n",
       "      <td>CG8421-NZ_CP005388-x</td>\n",
       "      <td>100.000</td>\n",
       "      <td>167</td>\n",
       "      <td>100</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>167</td>\n",
       "      <td>167</td>\n",
       "      <td>1258358</td>\n",
       "      <td>1258524</td>\n",
       "      <td>plus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>c_000000000011</td>\n",
       "      <td>CG8421-NZ_CP005388-x</td>\n",
       "      <td>100.000</td>\n",
       "      <td>3244</td>\n",
       "      <td>100</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>3244</td>\n",
       "      <td>3244</td>\n",
       "      <td>1258421</td>\n",
       "      <td>1261664</td>\n",
       "      <td>plus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>c_000000000014</td>\n",
       "      <td>CG8421-NZ_CP005388-x</td>\n",
       "      <td>100.000</td>\n",
       "      <td>539</td>\n",
       "      <td>100</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>539</td>\n",
       "      <td>539</td>\n",
       "      <td>1261538</td>\n",
       "      <td>1262076</td>\n",
       "      <td>minus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>c_000000000017</td>\n",
       "      <td>CG8421-NZ_CP005388-x</td>\n",
       "      <td>100.000</td>\n",
       "      <td>176</td>\n",
       "      <td>100</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>176</td>\n",
       "      <td>176</td>\n",
       "      <td>1261958</td>\n",
       "      <td>1262133</td>\n",
       "      <td>plus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>c_000000000015</td>\n",
       "      <td>CG8421-NZ_CP005388-x</td>\n",
       "      <td>100.000</td>\n",
       "      <td>288</td>\n",
       "      <td>100</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>288</td>\n",
       "      <td>288</td>\n",
       "      <td>1262007</td>\n",
       "      <td>1262294</td>\n",
       "      <td>minus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>c_000000000018</td>\n",
       "      <td>CG8421-NZ_CP005388-x</td>\n",
       "      <td>100.000</td>\n",
       "      <td>167</td>\n",
       "      <td>100</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>167</td>\n",
       "      <td>167</td>\n",
       "      <td>1262168</td>\n",
       "      <td>1262334</td>\n",
       "      <td>plus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>c_000000000006</td>\n",
       "      <td>CG8421-NZ_CP005388-x</td>\n",
       "      <td>99.997</td>\n",
       "      <td>92334</td>\n",
       "      <td>100</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>92334</td>\n",
       "      <td>92334</td>\n",
       "      <td>1262229</td>\n",
       "      <td>1354562</td>\n",
       "      <td>minus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>c_000000000018</td>\n",
       "      <td>CG8421-NZ_CP005388-x</td>\n",
       "      <td>97.024</td>\n",
       "      <td>168</td>\n",
       "      <td>100</td>\n",
       "      <td>4</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>167</td>\n",
       "      <td>167</td>\n",
       "      <td>1264096</td>\n",
       "      <td>1264263</td>\n",
       "      <td>plus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>c_000000000013</td>\n",
       "      <td>CG8421-NZ_CP005388-x</td>\n",
       "      <td>100.000</td>\n",
       "      <td>745</td>\n",
       "      <td>100</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>745</td>\n",
       "      <td>745</td>\n",
       "      <td>1354436</td>\n",
       "      <td>1355180</td>\n",
       "      <td>minus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>c_000000000009</td>\n",
       "      <td>CG8421-NZ_CP005388-x</td>\n",
       "      <td>100.000</td>\n",
       "      <td>14343</td>\n",
       "      <td>100</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>14343</td>\n",
       "      <td>14343</td>\n",
       "      <td>1355054</td>\n",
       "      <td>1369396</td>\n",
       "      <td>minus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>c_000000000013</td>\n",
       "      <td>CG8421-NZ_CP005388-x</td>\n",
       "      <td>100.000</td>\n",
       "      <td>745</td>\n",
       "      <td>100</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>745</td>\n",
       "      <td>745</td>\n",
       "      <td>1369270</td>\n",
       "      <td>1370014</td>\n",
       "      <td>minus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>c_000000000003</td>\n",
       "      <td>CG8421-NZ_CP005388-x</td>\n",
       "      <td>100.000</td>\n",
       "      <td>266142</td>\n",
       "      <td>100</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>266142</td>\n",
       "      <td>266142</td>\n",
       "      <td>1369888</td>\n",
       "      <td>1636029</td>\n",
       "      <td>minus</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            primer                sseqid   pident  length  qcov  mismatch  \\\n",
       "7   c_000000000008  CG8421-NZ_CP005388-x  100.000   38777   100         0   \n",
       "11  c_000000000010  CG8421-NZ_CP005388-x   99.834    6007   100         0   \n",
       "4   c_000000000005  CG8421-NZ_CP005388-x  100.000  104606   100         0   \n",
       "13  c_000000000012  CG8421-NZ_CP005388-x  100.000    1050   100         0   \n",
       "6   c_000000000007  CG8421-NZ_CP005388-x  100.000   89457   100         0   \n",
       "14  c_000000000012  CG8421-NZ_CP005388-x  100.000    1050   100         0   \n",
       "3   c_000000000004  CG8421-NZ_CP005388-x  100.000  153010   100         0   \n",
       "10  c_000000000010  CG8421-NZ_CP005388-x   99.917    6009   100         3   \n",
       "1   c_000000000002  CG8421-NZ_CP005388-x  100.000  287092   100         0   \n",
       "9   c_000000000010  CG8421-NZ_CP005388-x  100.000    6007   100         0   \n",
       "0   c_000000000001  CG8421-NZ_CP005388-x  100.000  565758   100         0   \n",
       "18  c_000000000014  CG8421-NZ_CP005388-x   99.814     539   100         1   \n",
       "21  c_000000000017  CG8421-NZ_CP005388-x  100.000     176   100         0   \n",
       "20  c_000000000016  CG8421-NZ_CP005388-x  100.000     254   100         0   \n",
       "23  c_000000000018  CG8421-NZ_CP005388-x  100.000     167   100         0   \n",
       "12  c_000000000011  CG8421-NZ_CP005388-x  100.000    3244   100         0   \n",
       "17  c_000000000014  CG8421-NZ_CP005388-x  100.000     539   100         0   \n",
       "22  c_000000000017  CG8421-NZ_CP005388-x  100.000     176   100         0   \n",
       "19  c_000000000015  CG8421-NZ_CP005388-x  100.000     288   100         0   \n",
       "24  c_000000000018  CG8421-NZ_CP005388-x  100.000     167   100         0   \n",
       "5   c_000000000006  CG8421-NZ_CP005388-x   99.997   92334   100         3   \n",
       "25  c_000000000018  CG8421-NZ_CP005388-x   97.024     168   100         4   \n",
       "15  c_000000000013  CG8421-NZ_CP005388-x  100.000     745   100         0   \n",
       "8   c_000000000009  CG8421-NZ_CP005388-x  100.000   14343   100         0   \n",
       "16  c_000000000013  CG8421-NZ_CP005388-x  100.000     745   100         0   \n",
       "2   c_000000000003  CG8421-NZ_CP005388-x  100.000  266142   100         0   \n",
       "\n",
       "    gapopen  qstart    qend    qlen  sstart2    send2 sstrand  \n",
       "7         0       1   38777   38777        1    38777    plus  \n",
       "11        1       1    6007    6007    38678    44674    plus  \n",
       "4         0       1  104606  104606    44548   149153   minus  \n",
       "13        0       1    1050    1050   149027   150076   minus  \n",
       "6         0       1   89457   89457   150077   239533   minus  \n",
       "14        0       1    1050    1050   239534   240583    plus  \n",
       "3         0       1  153010  153010   240457   393466    plus  \n",
       "10        2       1    6007    6007   393340   399348    plus  \n",
       "1         0       1  287092  287092   399286   686377    plus  \n",
       "9         0       1    6007    6007   686251   692257    plus  \n",
       "0         0       1  565758  565758   692131  1257888    plus  \n",
       "18        0       1     539     539  1257762  1258300   minus  \n",
       "21        0       1     176     176  1258182  1258357    plus  \n",
       "20        0       1     254     254  1258231  1258484   minus  \n",
       "23        0       1     167     167  1258358  1258524    plus  \n",
       "12        0       1    3244    3244  1258421  1261664    plus  \n",
       "17        0       1     539     539  1261538  1262076   minus  \n",
       "22        0       1     176     176  1261958  1262133    plus  \n",
       "19        0       1     288     288  1262007  1262294   minus  \n",
       "24        0       1     167     167  1262168  1262334    plus  \n",
       "5         0       1   92334   92334  1262229  1354562   minus  \n",
       "25        1       1     167     167  1264096  1264263    plus  \n",
       "15        0       1     745     745  1354436  1355180   minus  \n",
       "8         0       1   14343   14343  1355054  1369396   minus  \n",
       "16        0       1     745     745  1369270  1370014   minus  \n",
       "2         0       1  266142  266142  1369888  1636029   minus  "
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sim_assem = \"/Volumes/KeithSSD/variant-calling/simulated_reference/CG8421_sim_contigs_fixed.fa\"\n",
    "sim_check = \"/Volumes/KeithSSD/variant-calling/simulated_reference/simulation_matches.txt\"\n",
    "complete_ref = \"/Volumes/KeithSSD/variant-calling/ref_fastas/CG8421_bb.fa\"\n",
    "complete_db = \"/Volumes/KeithSSD/variant-calling/simulated_reference/complete_ref_db\"\n",
    "\n",
    "print('\\nOUTFMT=\"6 qseqid sseqid pident length qcovs mismatch gapopen qstart qend qlen sstart send slen evalue bitscore sstrand\"\\n')\n",
    "print(\"makeblastdb -in {} -dbtype nucl -out {}\\n\".format(complete_ref, complete_db))\n",
    "bp1 = 'blastn -evalue 1e-3 -db {} -query {} -outfmt '.format(complete_db, sim_assem)\n",
    "bp2 = '\"$OUTFMT\" -qcov_hsp_perc 90 -out {} -num_threads 3\\n\\n'.format(sim_check)\n",
    "print(bp1+bp2)\n",
    "\n",
    "blast_cols1 = ['primer', 'sseqid', 'pident', 'length', 'qcov', 'mismatch', 'gapopen', 'qstart', 'qend',\n",
    "              'qlen', 'sstart', 'send', 'slen', 'evalue', 'bitscore', 'sstrand']\n",
    "sim_check_df = pd.read_csv(sim_check, sep=\"\\t\", header=None)\n",
    "sim_check_df.columns = blast_cols1\n",
    "#sim_check_df['reversed'] = sim_check_df[['sstart', 'send']].apply(lambda x: min(x), axis=1)\n",
    "sim_check_df['sstart2'] = sim_check_df[['sstart', 'send']].apply(lambda x: min(x), axis=1)\n",
    "sim_check_df['send2'] = sim_check_df[['sstart', 'send']].apply(lambda x: max(x), axis=1)\n",
    "sim_check_df.sort_values(by='sstart2', inplace=True)\n",
    "\n",
    "#sim_check_df.loc[[7, 11, 4, 13, 6, 14, 10, 21, 20, 23, 12], ['primer', 'sseqid', 'pident', 'length', 'qcov', 'mismatch', 'gapopen', 'qstart', 'qend',\n",
    "#              'qlen', 'sstart2', 'send2', 'sstrand']]\n",
    "\n",
    "sim_check_df.loc[:, ['primer', 'sseqid', 'pident', 'length', 'qcov', 'mismatch', 'gapopen', 'qstart', 'qend',\n",
    "              'qlen', 'sstart2', 'send2', 'sstrand']]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Next, we read in a Genbank file for the reference strain (CG8421) from Crofts et al. We want a dataframe containing gene feature positions, annotations, protein and DNA sequence"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Parsed 3433 features from NZ_CP005388 to a dataframe size (3433, 35)\n",
      "Full genbank dataframe sized (3433, 35)\n"
     ]
    }
   ],
   "source": [
    "gb_file = \"/Volumes/KeithSSD/variant-calling/refs/GCF_000171795.2_ASM17179v2_genomic.gbff\"\n",
    "\n",
    "gb_records = {i.name:i for i in list(SeqIO.parse(open(gb_file,\"r\"), \"genbank\"))}\n",
    "\n",
    "gb_subdfs = []\n",
    "for chrom, gb_record in gb_records.items():\n",
    "    feat_idx = list(range(len(gb_record.features)))\n",
    "    exp_cols = set(['molecule', 'type', 'strand', 'start', 'end', 'sequence'])\n",
    "    exp_cols.update([k for i in feat_idx for k in gb_record.features[i].qualifiers.keys()])\n",
    "    gb_subdf = pd.DataFrame(index=feat_idx, columns=exp_cols)\n",
    "    for i in gb_subdf.index:\n",
    "        feature_i = gb_record.features[i]\n",
    "        gb_subdf.at[i, 'molecule'] = chrom\n",
    "        gb_subdf.at[i, 'type'] = feature_i.type\n",
    "        gb_subdf.at[i, 'strand'] = feature_i.location.strand\n",
    "        gb_subdf.at[i, 'start'] = int(feature_i.location.start)\n",
    "        gb_subdf.at[i, 'end'] = int(feature_i.location.end)\n",
    "        # This already does reverse complementing of crick strand features\n",
    "        gb_subdf.at[i, 'sequence'] = str(feature_i.extract(gb_record.seq))\n",
    "        for qk, qv in feature_i.qualifiers.items():\n",
    "            if len(qv) != 1:\n",
    "                qv = \", \".join(qv)\n",
    "            gb_subdf.at[i, qk] = qv[0]\n",
    "    gb_subdfs.append(gb_subdf.copy())\n",
    "    print(\"Parsed {} features from {} to a dataframe size {}\".format(len(feat_idx), chrom, gb_subdf.shape))\n",
    "    \n",
    "gb_df = pd.concat(gb_subdfs, axis=0, ignore_index=True)\n",
    "print(\"Full genbank dataframe sized {}\".format(gb_df.shape))\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Next, we will read in some information about the mutations detected in the Crofts study taken from their supplement. This block contains a sanity test whereby we match the reference allele nucleotides specified at locii where deletions and SNVs were discovered can be located in the genome sequence we extracted from our Genbank file. \n",
    "\n",
    "#### Positional information about insertions and multinucleotide deletions is given as a string with two coordinates seperated by a carot or two periods. We will reformat this information into a python tuple containing integers so that it can be used to index the string object containing the full genome sequence. \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Variant Position</th>\n",
       "      <th>Variant Type</th>\n",
       "      <th>Reference Sequence</th>\n",
       "      <th>Variant Allele</th>\n",
       "      <th>Affected CG8421 Gene(s)</th>\n",
       "      <th>CG8421 Genome Annotation</th>\n",
       "      <th>On/Off State In inoculum</th>\n",
       "      <th>On/Off State in Isolates</th>\n",
       "      <th>Variant Called in Inoculum or Isolate Population?</th>\n",
       "      <th>Isolate Population Samples That Differed From Inoculum</th>\n",
       "      <th>Relapse Samples with Variant (out of 6)</th>\n",
       "      <th>Relapse Sample Frequency</th>\n",
       "      <th>Primary Infection Samples with Variant (out of 43)</th>\n",
       "      <th>Primary Infection Sample Frequency</th>\n",
       "      <th>Annotation Used for this Study</th>\n",
       "      <th>Variant Location</th>\n",
       "      <th>var_locii</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>67639,67640</td>\n",
       "      <td>Insertion</td>\n",
       "      <td>-</td>\n",
       "      <td>C</td>\n",
       "      <td>CJ8421_RS00265, Promoter of CJ8421_RS00260</td>\n",
       "      <td>Hypothetical Protein</td>\n",
       "      <td>&lt;10% Truncation</td>\n",
       "      <td>On</td>\n",
       "      <td>Inoculum</td>\n",
       "      <td>Relapse 8.2 , Relapse 8.1 , Relapse 1 , Relaps...</td>\n",
       "      <td>4</td>\n",
       "      <td>66.666667</td>\n",
       "      <td>12</td>\n",
       "      <td>27.906977</td>\n",
       "      <td>fedA Hemerythrin</td>\n",
       "      <td>IN GENE, IN PROMOTER REGION</td>\n",
       "      <td>(67639, 67640)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>67639,67640</td>\n",
       "      <td>Insertion</td>\n",
       "      <td>-</td>\n",
       "      <td>CC</td>\n",
       "      <td>CJ8421_RS00265, Promoter of CJ8421_RS00260</td>\n",
       "      <td>Hypothetical Protein</td>\n",
       "      <td>&lt;10% Truncation</td>\n",
       "      <td>On</td>\n",
       "      <td>Inoculum</td>\n",
       "      <td>Relapse 8.1 , Relapse 1 , Relapse 4 , Relapse ...</td>\n",
       "      <td>4</td>\n",
       "      <td>66.666667</td>\n",
       "      <td>43</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>fedA Hemerythrin</td>\n",
       "      <td>IN GENE, IN PROMOTER REGION</td>\n",
       "      <td>(67639, 67640)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>100530,100531</td>\n",
       "      <td>Deletion</td>\n",
       "      <td>TA</td>\n",
       "      <td>-</td>\n",
       "      <td>CJ8421_RS00460</td>\n",
       "      <td>C4-dicarboxylate ABC transporter</td>\n",
       "      <td>On</td>\n",
       "      <td>Off</td>\n",
       "      <td>Infection Isolate(s)</td>\n",
       "      <td>16D5, 16D6</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>2</td>\n",
       "      <td>4.651163</td>\n",
       "      <td>dcuA Anaerobic C4-dicarboxylate Transporter</td>\n",
       "      <td>IN GENE</td>\n",
       "      <td>(100530, 100531)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>168690,168691</td>\n",
       "      <td>Insertion</td>\n",
       "      <td>-</td>\n",
       "      <td>G</td>\n",
       "      <td>CJ8421_RS00870 &amp; CJ8421_RS00875</td>\n",
       "      <td>hypothetical protein</td>\n",
       "      <td>On</td>\n",
       "      <td>&lt;10% Truncation</td>\n",
       "      <td>Infection Isolate(s)</td>\n",
       "      <td>Relapse 4</td>\n",
       "      <td>1</td>\n",
       "      <td>16.666667</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>Conserved Hypothetical Proteins</td>\n",
       "      <td>IN GENE, IN PROMOTER REGION</td>\n",
       "      <td>(168690, 168691)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>168690,168691</td>\n",
       "      <td>Insertion</td>\n",
       "      <td>-</td>\n",
       "      <td>GG</td>\n",
       "      <td>CJ8421_RS00870</td>\n",
       "      <td>hypothetical protein</td>\n",
       "      <td>On</td>\n",
       "      <td>&lt;10% Truncation</td>\n",
       "      <td>Infection Isolate(s)</td>\n",
       "      <td>Relapse 4 , Relapse 31</td>\n",
       "      <td>2</td>\n",
       "      <td>33.333333</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>Conserved Hypothetical Proteins</td>\n",
       "      <td>IN GENE</td>\n",
       "      <td>(168690, 168691)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>168691</td>\n",
       "      <td>Deletion</td>\n",
       "      <td>G</td>\n",
       "      <td>-</td>\n",
       "      <td>CJ8421_RS00870</td>\n",
       "      <td>hypothetical protein</td>\n",
       "      <td>On</td>\n",
       "      <td>&lt;10% Truncation</td>\n",
       "      <td>Infection Isolate(s)</td>\n",
       "      <td>Relapse 8.2 , Relapse 9</td>\n",
       "      <td>2</td>\n",
       "      <td>33.333333</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>Hypothetical Protein</td>\n",
       "      <td>IN GENE</td>\n",
       "      <td>(168691,)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>268276</td>\n",
       "      <td>Deletion</td>\n",
       "      <td>T</td>\n",
       "      <td>-</td>\n",
       "      <td>CJ8421_RS01430</td>\n",
       "      <td>thiF family protein</td>\n",
       "      <td>On</td>\n",
       "      <td>Off</td>\n",
       "      <td>Infection Isolate(s)</td>\n",
       "      <td>Relapse 31 , 36D1, 8D3</td>\n",
       "      <td>1</td>\n",
       "      <td>16.666667</td>\n",
       "      <td>2</td>\n",
       "      <td>4.651163</td>\n",
       "      <td>thiF Family Protein</td>\n",
       "      <td>IN GENE</td>\n",
       "      <td>(268276,)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>302194</td>\n",
       "      <td>SNV</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>CJ8421_RS01640</td>\n",
       "      <td>flagellar motor protein MotA</td>\n",
       "      <td>On</td>\n",
       "      <td>On</td>\n",
       "      <td>Infection Isolate(s)</td>\n",
       "      <td>11D5, 11D6</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>2</td>\n",
       "      <td>4.651163</td>\n",
       "      <td>motA Falgellar Motor Protein</td>\n",
       "      <td>IN GENE</td>\n",
       "      <td>(302194,)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>335945,335946</td>\n",
       "      <td>Insertion</td>\n",
       "      <td>-</td>\n",
       "      <td>A</td>\n",
       "      <td>CJ8421_RS01805</td>\n",
       "      <td>TetR family transcriptional regulator</td>\n",
       "      <td>On</td>\n",
       "      <td>Off</td>\n",
       "      <td>Infection Isolate(s)</td>\n",
       "      <td>Relapse 31</td>\n",
       "      <td>1</td>\n",
       "      <td>16.666667</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>cmeR Bile Sensing Transcriptional Regulator</td>\n",
       "      <td>IN GENE</td>\n",
       "      <td>(335945, 335946)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>335972</td>\n",
       "      <td>Deletion</td>\n",
       "      <td>T</td>\n",
       "      <td>-</td>\n",
       "      <td>CJ8421_RS01805</td>\n",
       "      <td>TetR family transcriptional regulator</td>\n",
       "      <td>On</td>\n",
       "      <td>Off</td>\n",
       "      <td>Infection Isolate(s)</td>\n",
       "      <td>Relapse 4</td>\n",
       "      <td>1</td>\n",
       "      <td>16.666667</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>cmeR Bile Sensing Transcriptional Regulator</td>\n",
       "      <td>IN GENE</td>\n",
       "      <td>(335972,)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>517620,517621</td>\n",
       "      <td>Insertion</td>\n",
       "      <td>-</td>\n",
       "      <td>G</td>\n",
       "      <td>CJ8421_RS02760 &amp; CJ8421_RS02765</td>\n",
       "      <td>Membrane Protein, GTP Binding Protein</td>\n",
       "      <td>On</td>\n",
       "      <td>On</td>\n",
       "      <td>Infection Isolate(s)</td>\n",
       "      <td>Relapse 8.2</td>\n",
       "      <td>1</td>\n",
       "      <td>16.666667</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>Membrane GTP Binding Protein</td>\n",
       "      <td>BETWEEN GENES, 22BP DOWNSTREAM/56BP UPSTREAM</td>\n",
       "      <td>(517620, 517621)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>517621</td>\n",
       "      <td>Deletion</td>\n",
       "      <td>G</td>\n",
       "      <td>-</td>\n",
       "      <td>CJ8421_RS02760 &amp; CJ8421_RS02765</td>\n",
       "      <td>Membrane Protein, GTP Binding Protein</td>\n",
       "      <td>On</td>\n",
       "      <td>On</td>\n",
       "      <td>Infection Isolate(s)</td>\n",
       "      <td>36D2, 32D2</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>2</td>\n",
       "      <td>4.651163</td>\n",
       "      <td>Membrane GTP Binding Protein</td>\n",
       "      <td>BETWEEN GENES, 22BP DOWNSTREAM/56BP UPSTREAM</td>\n",
       "      <td>(517621,)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>565752</td>\n",
       "      <td>Deletion</td>\n",
       "      <td>G</td>\n",
       "      <td>-</td>\n",
       "      <td>CJ8421_RS03020</td>\n",
       "      <td>Carbonic Anyhydrase, disrupted</td>\n",
       "      <td>On</td>\n",
       "      <td>Off</td>\n",
       "      <td>Inoculum</td>\n",
       "      <td>Relapse 8.2 , 14D1, 21D1, 24D1, 16D0, 16D3, 34...</td>\n",
       "      <td>1</td>\n",
       "      <td>16.666667</td>\n",
       "      <td>8</td>\n",
       "      <td>18.604651</td>\n",
       "      <td>Flagella Modification Gene*</td>\n",
       "      <td>IN CDS</td>\n",
       "      <td>(565752,)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>565752,565753</td>\n",
       "      <td>Deletion</td>\n",
       "      <td>GG</td>\n",
       "      <td>-</td>\n",
       "      <td>CJ8421_RS03020</td>\n",
       "      <td>Carbonic Anyhydrase, disrupted</td>\n",
       "      <td>On</td>\n",
       "      <td>Off</td>\n",
       "      <td>Infection Isolate(s)</td>\n",
       "      <td>Relapse 8.2 , 32D2</td>\n",
       "      <td>1</td>\n",
       "      <td>16.666667</td>\n",
       "      <td>1</td>\n",
       "      <td>2.325581</td>\n",
       "      <td>Flagella Modification Gene*</td>\n",
       "      <td>IN CDS</td>\n",
       "      <td>(565752, 565753)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>620468</td>\n",
       "      <td>Deletion</td>\n",
       "      <td>G</td>\n",
       "      <td>-</td>\n",
       "      <td>CJ8421_RS03305</td>\n",
       "      <td>potassium transporter KtrB; disrupted</td>\n",
       "      <td>Off</td>\n",
       "      <td>Off</td>\n",
       "      <td>Infection Isolate(s)</td>\n",
       "      <td>Relapse 1 , Relapse 9 , 36D1, 1D3, 8D3</td>\n",
       "      <td>2</td>\n",
       "      <td>33.333333</td>\n",
       "      <td>3</td>\n",
       "      <td>6.976744</td>\n",
       "      <td>Potassium Transporter KtrB; disrupted</td>\n",
       "      <td>IN CDS</td>\n",
       "      <td>(620468,)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>629686,629687</td>\n",
       "      <td>Insertion</td>\n",
       "      <td>-</td>\n",
       "      <td>G</td>\n",
       "      <td>CJ8421_RS03350</td>\n",
       "      <td>invasion protein</td>\n",
       "      <td>Off</td>\n",
       "      <td>On</td>\n",
       "      <td>Inoculum</td>\n",
       "      <td>Relapse 8.2 , Relapse 8.1 , Relapse 1 , Relaps...</td>\n",
       "      <td>6</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>8</td>\n",
       "      <td>18.604651</td>\n",
       "      <td>cipA Invasion Protein</td>\n",
       "      <td>IN GENE</td>\n",
       "      <td>(629686, 629687)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>629687</td>\n",
       "      <td>Deletion</td>\n",
       "      <td>G</td>\n",
       "      <td>-</td>\n",
       "      <td>CJ8421_RS03350</td>\n",
       "      <td>invasion protein</td>\n",
       "      <td>Off</td>\n",
       "      <td>Off</td>\n",
       "      <td>Infection Isolate(s)</td>\n",
       "      <td>Relapse 8.2</td>\n",
       "      <td>1</td>\n",
       "      <td>16.666667</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>cipA Invasion Protein</td>\n",
       "      <td>IN GENE</td>\n",
       "      <td>(629687,)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>680285</td>\n",
       "      <td>Deletion</td>\n",
       "      <td>G</td>\n",
       "      <td>-</td>\n",
       "      <td>CJ8421_RS03600 &amp; CJ8421_RS03605</td>\n",
       "      <td>Hypothetical disrupted proteins</td>\n",
       "      <td>&lt;10% Truncation</td>\n",
       "      <td>&lt;10% Truncation</td>\n",
       "      <td>Infection Isolate(s)</td>\n",
       "      <td>11D1</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1</td>\n",
       "      <td>2.325581</td>\n",
       "      <td>Conserved Hypothetical Proteins</td>\n",
       "      <td>IN GENE, IN PROMOTER REGION 53BP UPSTREAM</td>\n",
       "      <td>(680285,)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>686224</td>\n",
       "      <td>Deletion</td>\n",
       "      <td>C</td>\n",
       "      <td>-</td>\n",
       "      <td>CJ8421_RS03635</td>\n",
       "      <td>Membrane Protein; disrupted</td>\n",
       "      <td>On</td>\n",
       "      <td>On</td>\n",
       "      <td>Inoculum</td>\n",
       "      <td>Relapse 4 , Relapse 9 , Relapse 31 , 16D3</td>\n",
       "      <td>3</td>\n",
       "      <td>50.000000</td>\n",
       "      <td>1</td>\n",
       "      <td>2.325581</td>\n",
       "      <td>Conserved Hypothetical Proteins</td>\n",
       "      <td>21BP DOWNSTREAM</td>\n",
       "      <td>(686224,)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>861171</td>\n",
       "      <td>SNV</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>CJ8421_RS04555</td>\n",
       "      <td>adenine phosphoribosyltransferase</td>\n",
       "      <td>On</td>\n",
       "      <td>On</td>\n",
       "      <td>Inoculum</td>\n",
       "      <td>Relapse 8.2 , Relapse 8.1 , Relapse 1 , Relaps...</td>\n",
       "      <td>6</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>43</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>apt Adenine Phosphoribosyltransferase</td>\n",
       "      <td>IN GENE</td>\n",
       "      <td>(861171,)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>982992</td>\n",
       "      <td>Deletion</td>\n",
       "      <td>C</td>\n",
       "      <td>-</td>\n",
       "      <td>CJ8421_RS05220</td>\n",
       "      <td>Pseudogene, Restriction Endonculease;disrupted</td>\n",
       "      <td>Off</td>\n",
       "      <td>On</td>\n",
       "      <td>Infection Isolate(s)</td>\n",
       "      <td>Relapse 8.2 , Relapse 8.1 , Relapse 4 , Relaps...</td>\n",
       "      <td>5</td>\n",
       "      <td>83.333333</td>\n",
       "      <td>39</td>\n",
       "      <td>90.697674</td>\n",
       "      <td>cjeI Restriction Enzyme</td>\n",
       "      <td>IN CDS</td>\n",
       "      <td>(982992,)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>1070389</td>\n",
       "      <td>Deletion</td>\n",
       "      <td>G</td>\n",
       "      <td>-</td>\n",
       "      <td>CJ8421_RS05645 &amp; CJ8421_RS05650</td>\n",
       "      <td>dTDP-6-deoxy-3,4-keto-hexulose isomerase</td>\n",
       "      <td>On</td>\n",
       "      <td>&lt;10% Truncation</td>\n",
       "      <td>Infection Isolate(s)</td>\n",
       "      <td>Relapse 8.2 , 14D1</td>\n",
       "      <td>1</td>\n",
       "      <td>16.666667</td>\n",
       "      <td>1</td>\n",
       "      <td>2.325581</td>\n",
       "      <td>dTDP-6-deoxy-3,4-keto-hexulose isomerase</td>\n",
       "      <td>IN GENE, IN PROMOTER</td>\n",
       "      <td>(1070389,)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>1071193</td>\n",
       "      <td>Deletion</td>\n",
       "      <td>G</td>\n",
       "      <td>-</td>\n",
       "      <td>CJ8421_RS05655</td>\n",
       "      <td>dtdp-6-deoxy-3,4-keto-hexulose isomerase</td>\n",
       "      <td>On</td>\n",
       "      <td>Off</td>\n",
       "      <td>Infection Isolate(s)</td>\n",
       "      <td>Relapse 8.1 , Relapse 4 , Relapse 9 , Relapse ...</td>\n",
       "      <td>4</td>\n",
       "      <td>66.666667</td>\n",
       "      <td>2</td>\n",
       "      <td>4.651163</td>\n",
       "      <td>dTDP-6-deoxy-3,4-keto-hexulose isomerase</td>\n",
       "      <td>IN GENE</td>\n",
       "      <td>(1071193,)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>1227256,1227257</td>\n",
       "      <td>Insertion</td>\n",
       "      <td>-</td>\n",
       "      <td>GG</td>\n",
       "      <td>CJ8421_RS06440</td>\n",
       "      <td>Aminopeptidase; disrupted (overlaps with degT ...</td>\n",
       "      <td>Off</td>\n",
       "      <td>On</td>\n",
       "      <td>Infection Isolate(s)</td>\n",
       "      <td>Relapse 9</td>\n",
       "      <td>1</td>\n",
       "      <td>16.666667</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>O-linked Glycosylation Gene</td>\n",
       "      <td>IN CDS</td>\n",
       "      <td>(1227256, 1227257)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>1227257</td>\n",
       "      <td>Deletion</td>\n",
       "      <td>G</td>\n",
       "      <td>-</td>\n",
       "      <td>CJ8421_RS06440</td>\n",
       "      <td>Aminopeptidase; disrupted (overlaps with degT ...</td>\n",
       "      <td>Off</td>\n",
       "      <td>On</td>\n",
       "      <td>Infection Isolate(s)</td>\n",
       "      <td>Relapse 8.2 , Relapse 1 , Relapse 4 , Relapse 31</td>\n",
       "      <td>4</td>\n",
       "      <td>66.666667</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>O-linked Glycosylation Gene</td>\n",
       "      <td>IN CDS</td>\n",
       "      <td>(1227257,)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>1234215,1234216</td>\n",
       "      <td>Insertion</td>\n",
       "      <td>-</td>\n",
       "      <td>C</td>\n",
       "      <td>CJ8421_RS06480</td>\n",
       "      <td>carbonic anyhydrase</td>\n",
       "      <td>On</td>\n",
       "      <td>On</td>\n",
       "      <td>Infection Isolate(s)</td>\n",
       "      <td>Relapse 31</td>\n",
       "      <td>1</td>\n",
       "      <td>16.666667</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>O-linked Glycosylation Gene</td>\n",
       "      <td>51BP UPSTREAM</td>\n",
       "      <td>(1234215, 1234216)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>1234216</td>\n",
       "      <td>Deletion</td>\n",
       "      <td>C</td>\n",
       "      <td>-</td>\n",
       "      <td>CJ8421_RS06480</td>\n",
       "      <td>carbonic anyhydrase</td>\n",
       "      <td>On</td>\n",
       "      <td>On</td>\n",
       "      <td>Infection Isolate(s)</td>\n",
       "      <td>Relapse 8.2 , Relapse 1</td>\n",
       "      <td>2</td>\n",
       "      <td>33.333333</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>O-linked Glycosylation Gene</td>\n",
       "      <td>51BP UPSTREAM</td>\n",
       "      <td>(1234216,)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>1248836,1248837</td>\n",
       "      <td>Insertion</td>\n",
       "      <td>-</td>\n",
       "      <td>G</td>\n",
       "      <td>CJ8421_RS06560</td>\n",
       "      <td>transferase</td>\n",
       "      <td>On</td>\n",
       "      <td>On</td>\n",
       "      <td>Inoculum</td>\n",
       "      <td>Relapse 4 , 14D0, 14D1, 14D2, 14D3, 1D5, 21D1,...</td>\n",
       "      <td>1</td>\n",
       "      <td>16.666667</td>\n",
       "      <td>42</td>\n",
       "      <td>97.674419</td>\n",
       "      <td>O-linked Glycosylation Gene</td>\n",
       "      <td>IN PROMOTER REGION</td>\n",
       "      <td>(1248836, 1248837)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>1248837</td>\n",
       "      <td>Deletion</td>\n",
       "      <td>G</td>\n",
       "      <td>-</td>\n",
       "      <td>CJ8421_RS06560</td>\n",
       "      <td>transferase</td>\n",
       "      <td>On</td>\n",
       "      <td>On</td>\n",
       "      <td>Infection Isolate(s)</td>\n",
       "      <td>16D0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1</td>\n",
       "      <td>2.325581</td>\n",
       "      <td>O-linked Glycosylation Gene</td>\n",
       "      <td>IN PROMOTER REGION</td>\n",
       "      <td>(1248837,)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>1251713,1251714</td>\n",
       "      <td>Insertion</td>\n",
       "      <td>-</td>\n",
       "      <td>G</td>\n",
       "      <td>CJ8421_RS06580 &amp; CJ8421_RS06585</td>\n",
       "      <td>Methyltransferase</td>\n",
       "      <td>On</td>\n",
       "      <td>&lt;10% Truncation</td>\n",
       "      <td>Infection Isolate(s)</td>\n",
       "      <td>Relapse 4 , Relapse 9</td>\n",
       "      <td>2</td>\n",
       "      <td>33.333333</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>O-linked Glycosylation ptmH</td>\n",
       "      <td>IN GENE</td>\n",
       "      <td>(1251713, 1251714)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>1251714</td>\n",
       "      <td>Deletion</td>\n",
       "      <td>G</td>\n",
       "      <td>-</td>\n",
       "      <td>CJ8421_RS06580</td>\n",
       "      <td>Methyltransferase</td>\n",
       "      <td>On</td>\n",
       "      <td>&lt;10% Truncation</td>\n",
       "      <td>Infection Isolate(s)</td>\n",
       "      <td>Relapse 31 , 16D0, 34D0</td>\n",
       "      <td>1</td>\n",
       "      <td>16.666667</td>\n",
       "      <td>2</td>\n",
       "      <td>4.651163</td>\n",
       "      <td>O-linked Glycosylation ptmH</td>\n",
       "      <td>IN GENE</td>\n",
       "      <td>(1251714,)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>1262397,1262398</td>\n",
       "      <td>Insertion</td>\n",
       "      <td>-</td>\n",
       "      <td>T</td>\n",
       "      <td>CJ8421_RS06630</td>\n",
       "      <td>PseD protein</td>\n",
       "      <td>On</td>\n",
       "      <td>Off</td>\n",
       "      <td>Infection Isolate(s)</td>\n",
       "      <td>Relapse 4</td>\n",
       "      <td>1</td>\n",
       "      <td>16.666667</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>O-linked Glycosylation pseD</td>\n",
       "      <td>IN GENE</td>\n",
       "      <td>(1262397, 1262398)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>1262399</td>\n",
       "      <td>Deletion</td>\n",
       "      <td>A</td>\n",
       "      <td>-</td>\n",
       "      <td>CJ8421_RS06630</td>\n",
       "      <td>PseD protein</td>\n",
       "      <td>On</td>\n",
       "      <td>Off</td>\n",
       "      <td>Infection Isolate(s)</td>\n",
       "      <td>Relapse 4</td>\n",
       "      <td>1</td>\n",
       "      <td>16.666667</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>O-linked Glycosylation pseD</td>\n",
       "      <td>IN GENE</td>\n",
       "      <td>(1262399,)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>1263109,1263110</td>\n",
       "      <td>Insertion</td>\n",
       "      <td>-</td>\n",
       "      <td>G</td>\n",
       "      <td>CJ8421_RS06630</td>\n",
       "      <td>PseD protein</td>\n",
       "      <td>On</td>\n",
       "      <td>Off</td>\n",
       "      <td>Infection Isolate(s)</td>\n",
       "      <td>Relapse 8.2</td>\n",
       "      <td>1</td>\n",
       "      <td>16.666667</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>O-linked Glycosylation pseD</td>\n",
       "      <td>IN GENE</td>\n",
       "      <td>(1263109, 1263110)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>1272799,1272800</td>\n",
       "      <td>Insertion</td>\n",
       "      <td>-</td>\n",
       "      <td>C</td>\n",
       "      <td>CJ8421_RS06660</td>\n",
       "      <td>hypothetical protein</td>\n",
       "      <td>On</td>\n",
       "      <td>Off</td>\n",
       "      <td>Infection Isolate(s)</td>\n",
       "      <td>Relapse 8.1 , Relapse 9</td>\n",
       "      <td>2</td>\n",
       "      <td>33.333333</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>O-linked Glycosylation maf7</td>\n",
       "      <td>IN GENE</td>\n",
       "      <td>(1272799, 1272800)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>1272800</td>\n",
       "      <td>Deletion</td>\n",
       "      <td>C</td>\n",
       "      <td>-</td>\n",
       "      <td>CJ8421_RS06660</td>\n",
       "      <td>hypothetical protein</td>\n",
       "      <td>On</td>\n",
       "      <td>Off</td>\n",
       "      <td>Infection Isolate(s)</td>\n",
       "      <td>Relapse 1</td>\n",
       "      <td>1</td>\n",
       "      <td>16.666667</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>O-linked Glycosylation maf7</td>\n",
       "      <td>IN GENE</td>\n",
       "      <td>(1272800,)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>1293109,1293110</td>\n",
       "      <td>Insertion</td>\n",
       "      <td>-</td>\n",
       "      <td>C</td>\n",
       "      <td>CJ8421_RS06770</td>\n",
       "      <td>C4-dicarboxylate ABC transporter</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Infection Isolate(s)</td>\n",
       "      <td>Relapse 8.2 , Relapse 8.1 , 34D0</td>\n",
       "      <td>2</td>\n",
       "      <td>33.333333</td>\n",
       "      <td>1</td>\n",
       "      <td>2.325581</td>\n",
       "      <td>C4-dicarboxylate ABC transporter</td>\n",
       "      <td>40BP UPSTREAM</td>\n",
       "      <td>(1293109, 1293110)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>1293110</td>\n",
       "      <td>Deletion</td>\n",
       "      <td>C</td>\n",
       "      <td>-</td>\n",
       "      <td>CJ8421_RS06770</td>\n",
       "      <td>C4-dicarboxylate ABC transporter</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Infection Isolate(s)</td>\n",
       "      <td>14D0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1</td>\n",
       "      <td>2.325581</td>\n",
       "      <td>C4-dicarboxylate ABC transporter</td>\n",
       "      <td>40BP UPSTREAM</td>\n",
       "      <td>(1293110,)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>1352935,1352936</td>\n",
       "      <td>Insertion</td>\n",
       "      <td>-</td>\n",
       "      <td>C</td>\n",
       "      <td>CJ8421_RS07045</td>\n",
       "      <td>methyltransferase</td>\n",
       "      <td>On</td>\n",
       "      <td>Off</td>\n",
       "      <td>Infection Isolate(s)</td>\n",
       "      <td>Relapse 31 , 16D6</td>\n",
       "      <td>1</td>\n",
       "      <td>16.666667</td>\n",
       "      <td>1</td>\n",
       "      <td>2.325581</td>\n",
       "      <td>Putative Methyltransferase</td>\n",
       "      <td>IN GENE</td>\n",
       "      <td>(1352935, 1352936)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>1352936</td>\n",
       "      <td>Deletion</td>\n",
       "      <td>C</td>\n",
       "      <td>-</td>\n",
       "      <td>CJ8421_RS07045</td>\n",
       "      <td>methyltransferase</td>\n",
       "      <td>On</td>\n",
       "      <td>Off</td>\n",
       "      <td>Infection Isolate(s)</td>\n",
       "      <td>Relapse 8.2 , Relapse 1 , 24D2, 4D2, 16D0, 33D...</td>\n",
       "      <td>2</td>\n",
       "      <td>33.333333</td>\n",
       "      <td>7</td>\n",
       "      <td>16.279070</td>\n",
       "      <td>Putative Methyltransferase</td>\n",
       "      <td>IN GENE</td>\n",
       "      <td>(1352936,)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>1358677,1358678</td>\n",
       "      <td>Insertion</td>\n",
       "      <td>-</td>\n",
       "      <td>C</td>\n",
       "      <td>CJ8421_RS07075</td>\n",
       "      <td>GDP-mannose 4,6-dehydratase</td>\n",
       "      <td>On</td>\n",
       "      <td>Off</td>\n",
       "      <td>Infection Isolate(s)</td>\n",
       "      <td>Relapse 9</td>\n",
       "      <td>1</td>\n",
       "      <td>16.666667</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>wcbK, GDP-mannose 4,6-dehydratase</td>\n",
       "      <td>IN GENE</td>\n",
       "      <td>(1358677, 1358678)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>1361225,1361226</td>\n",
       "      <td>Insertion</td>\n",
       "      <td>-</td>\n",
       "      <td>C</td>\n",
       "      <td>CJ8421_RS07085 &amp; CJ8421_RS07090</td>\n",
       "      <td>Hypothetical proteins</td>\n",
       "      <td>On</td>\n",
       "      <td>Off</td>\n",
       "      <td>Infection Isolate(s)</td>\n",
       "      <td>Relapse 31</td>\n",
       "      <td>1</td>\n",
       "      <td>16.666667</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>Conserved Hypothetical Protein</td>\n",
       "      <td>IN GENE</td>\n",
       "      <td>(1361225, 1361226)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>1361226</td>\n",
       "      <td>Deletion</td>\n",
       "      <td>C</td>\n",
       "      <td>-</td>\n",
       "      <td>CJ8421_RS07085 &amp; CJ8421_RS07090</td>\n",
       "      <td>Hypothetical proteins</td>\n",
       "      <td>On</td>\n",
       "      <td>Off</td>\n",
       "      <td>Infection Isolate(s)</td>\n",
       "      <td>Relapse 8.2 , Relapse 8.1 , 14D3, 24D1, 31D2, ...</td>\n",
       "      <td>2</td>\n",
       "      <td>33.333333</td>\n",
       "      <td>12</td>\n",
       "      <td>27.906977</td>\n",
       "      <td>Conserved Hypothetical Protein</td>\n",
       "      <td>IN GENE</td>\n",
       "      <td>(1361226,)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>1363819</td>\n",
       "      <td>Deletion</td>\n",
       "      <td>C</td>\n",
       "      <td>-</td>\n",
       "      <td>CJ8421_RS07100</td>\n",
       "      <td>sugar transferase</td>\n",
       "      <td>On</td>\n",
       "      <td>On</td>\n",
       "      <td>Infection Isolate(s)</td>\n",
       "      <td>Relapse 8.1 , Relapse 1 , Relapse 4 , Relapse ...</td>\n",
       "      <td>4</td>\n",
       "      <td>66.666667</td>\n",
       "      <td>37</td>\n",
       "      <td>86.046512</td>\n",
       "      <td>wcA Sugar Transferase</td>\n",
       "      <td>INTERGENIC - 43BP UPSTREAM</td>\n",
       "      <td>(1363819,)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>1363819,1363820</td>\n",
       "      <td>Deletion</td>\n",
       "      <td>CC</td>\n",
       "      <td>-</td>\n",
       "      <td>CJ8421_RS07100</td>\n",
       "      <td>sugar transferase</td>\n",
       "      <td>On</td>\n",
       "      <td>On</td>\n",
       "      <td>Infection Isolate(s)</td>\n",
       "      <td>Relapse 8.2</td>\n",
       "      <td>1</td>\n",
       "      <td>16.666667</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>wcA Sugar Transferase</td>\n",
       "      <td>INTERGENIC - 43BP UPSTREAM</td>\n",
       "      <td>(1363819, 1363820)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>1416012,1416020*</td>\n",
       "      <td>Deletion</td>\n",
       "      <td>TTAAATTTT</td>\n",
       "      <td>-</td>\n",
       "      <td>CJ8421_RS07360</td>\n",
       "      <td>chemotaxis protein CheY</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>On (Wild-type reference sequence)</td>\n",
       "      <td>Inoculum</td>\n",
       "      <td>Relapse 8.2 , Relapse 8.1 , Relapse 1 , Relaps...</td>\n",
       "      <td>6</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>43</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>Cjj81176_1483 response regulator</td>\n",
       "      <td>IN GENE</td>\n",
       "      <td>(1416012, 1416020)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46</th>\n",
       "      <td>1485899</td>\n",
       "      <td>Deletion</td>\n",
       "      <td>A</td>\n",
       "      <td>-</td>\n",
       "      <td>CJ8421_RS07715</td>\n",
       "      <td>chemotaxis protein</td>\n",
       "      <td>On</td>\n",
       "      <td>Off</td>\n",
       "      <td>Infection Isolate(s)</td>\n",
       "      <td>16D5, 16D6</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>2</td>\n",
       "      <td>4.651163</td>\n",
       "      <td>Methyl-accepting chemotaxis protein</td>\n",
       "      <td>IN GENE</td>\n",
       "      <td>(1485899,)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47</th>\n",
       "      <td>1486346,1486347</td>\n",
       "      <td>Insertion</td>\n",
       "      <td>-</td>\n",
       "      <td>G</td>\n",
       "      <td>CJ8421_RS07715 &amp; CJ8421_RS07720</td>\n",
       "      <td>Chemotaxis protein, Flagellar Protein</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Infection Isolate(s)</td>\n",
       "      <td>14D1, 11D4, 1D2, 11D1</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>4</td>\n",
       "      <td>9.302326</td>\n",
       "      <td>Methyl-accepting chemotaxis protein</td>\n",
       "      <td>IN GENE</td>\n",
       "      <td>(1486346, 1486347)</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    Variant Position Variant Type Reference Sequence Variant Allele  \\\n",
       "0        67639,67640    Insertion                  -              C   \n",
       "1        67639,67640    Insertion                  -             CC   \n",
       "2      100530,100531     Deletion                 TA              -   \n",
       "3      168690,168691    Insertion                  -              G   \n",
       "4      168690,168691    Insertion                  -             GG   \n",
       "5             168691     Deletion                  G              -   \n",
       "6             268276     Deletion                  T              -   \n",
       "7             302194          SNV                  A              G   \n",
       "8      335945,335946    Insertion                  -              A   \n",
       "9             335972     Deletion                  T              -   \n",
       "10     517620,517621    Insertion                  -              G   \n",
       "11            517621     Deletion                  G              -   \n",
       "12            565752     Deletion                  G              -   \n",
       "13     565752,565753     Deletion                 GG              -   \n",
       "14            620468     Deletion                  G              -   \n",
       "15     629686,629687    Insertion                  -              G   \n",
       "16            629687     Deletion                  G              -   \n",
       "17            680285     Deletion                  G              -   \n",
       "18            686224     Deletion                  C              -   \n",
       "19            861171          SNV                  C              T   \n",
       "20            982992     Deletion                  C              -   \n",
       "21           1070389     Deletion                  G              -   \n",
       "22           1071193     Deletion                  G              -   \n",
       "23   1227256,1227257    Insertion                  -             GG   \n",
       "24           1227257     Deletion                  G              -   \n",
       "25   1234215,1234216    Insertion                  -              C   \n",
       "26           1234216     Deletion                  C              -   \n",
       "27   1248836,1248837    Insertion                  -              G   \n",
       "28           1248837     Deletion                  G              -   \n",
       "29   1251713,1251714    Insertion                  -              G   \n",
       "30           1251714     Deletion                  G              -   \n",
       "31   1262397,1262398    Insertion                  -              T   \n",
       "32           1262399     Deletion                  A              -   \n",
       "33   1263109,1263110    Insertion                  -              G   \n",
       "34   1272799,1272800    Insertion                  -              C   \n",
       "35           1272800     Deletion                  C              -   \n",
       "36   1293109,1293110    Insertion                  -              C   \n",
       "37           1293110     Deletion                  C              -   \n",
       "38   1352935,1352936    Insertion                  -              C   \n",
       "39           1352936     Deletion                  C              -   \n",
       "40   1358677,1358678    Insertion                  -              C   \n",
       "41   1361225,1361226    Insertion                  -              C   \n",
       "42           1361226     Deletion                  C              -   \n",
       "43           1363819     Deletion                  C              -   \n",
       "44   1363819,1363820     Deletion                 CC              -   \n",
       "45  1416012,1416020*     Deletion          TTAAATTTT              -   \n",
       "46           1485899     Deletion                  A              -   \n",
       "47   1486346,1486347    Insertion                  -              G   \n",
       "\n",
       "                       Affected CG8421 Gene(s)  \\\n",
       "0   CJ8421_RS00265, Promoter of CJ8421_RS00260   \n",
       "1   CJ8421_RS00265, Promoter of CJ8421_RS00260   \n",
       "2                               CJ8421_RS00460   \n",
       "3              CJ8421_RS00870 & CJ8421_RS00875   \n",
       "4                               CJ8421_RS00870   \n",
       "5                               CJ8421_RS00870   \n",
       "6                               CJ8421_RS01430   \n",
       "7                               CJ8421_RS01640   \n",
       "8                               CJ8421_RS01805   \n",
       "9                               CJ8421_RS01805   \n",
       "10             CJ8421_RS02760 & CJ8421_RS02765   \n",
       "11             CJ8421_RS02760 & CJ8421_RS02765   \n",
       "12                              CJ8421_RS03020   \n",
       "13                              CJ8421_RS03020   \n",
       "14                              CJ8421_RS03305   \n",
       "15                              CJ8421_RS03350   \n",
       "16                              CJ8421_RS03350   \n",
       "17             CJ8421_RS03600 & CJ8421_RS03605   \n",
       "18                              CJ8421_RS03635   \n",
       "19                              CJ8421_RS04555   \n",
       "20                              CJ8421_RS05220   \n",
       "21             CJ8421_RS05645 & CJ8421_RS05650   \n",
       "22                              CJ8421_RS05655   \n",
       "23                              CJ8421_RS06440   \n",
       "24                              CJ8421_RS06440   \n",
       "25                              CJ8421_RS06480   \n",
       "26                              CJ8421_RS06480   \n",
       "27                              CJ8421_RS06560   \n",
       "28                              CJ8421_RS06560   \n",
       "29            CJ8421_RS06580 & CJ8421_RS06585    \n",
       "30                              CJ8421_RS06580   \n",
       "31                              CJ8421_RS06630   \n",
       "32                              CJ8421_RS06630   \n",
       "33                              CJ8421_RS06630   \n",
       "34                              CJ8421_RS06660   \n",
       "35                              CJ8421_RS06660   \n",
       "36                              CJ8421_RS06770   \n",
       "37                              CJ8421_RS06770   \n",
       "38                              CJ8421_RS07045   \n",
       "39                              CJ8421_RS07045   \n",
       "40                              CJ8421_RS07075   \n",
       "41             CJ8421_RS07085 & CJ8421_RS07090   \n",
       "42             CJ8421_RS07085 & CJ8421_RS07090   \n",
       "43                              CJ8421_RS07100   \n",
       "44                              CJ8421_RS07100   \n",
       "45                              CJ8421_RS07360   \n",
       "46                              CJ8421_RS07715   \n",
       "47             CJ8421_RS07715 & CJ8421_RS07720   \n",
       "\n",
       "                             CG8421 Genome Annotation  \\\n",
       "0                                Hypothetical Protein   \n",
       "1                                Hypothetical Protein   \n",
       "2                    C4-dicarboxylate ABC transporter   \n",
       "3                                hypothetical protein   \n",
       "4                                hypothetical protein   \n",
       "5                                hypothetical protein   \n",
       "6                                 thiF family protein   \n",
       "7                        flagellar motor protein MotA   \n",
       "8               TetR family transcriptional regulator   \n",
       "9               TetR family transcriptional regulator   \n",
       "10              Membrane Protein, GTP Binding Protein   \n",
       "11              Membrane Protein, GTP Binding Protein   \n",
       "12                     Carbonic Anyhydrase, disrupted   \n",
       "13                     Carbonic Anyhydrase, disrupted   \n",
       "14              potassium transporter KtrB; disrupted   \n",
       "15                                   invasion protein   \n",
       "16                                   invasion protein   \n",
       "17                    Hypothetical disrupted proteins   \n",
       "18                        Membrane Protein; disrupted   \n",
       "19                  adenine phosphoribosyltransferase   \n",
       "20     Pseudogene, Restriction Endonculease;disrupted   \n",
       "21           dTDP-6-deoxy-3,4-keto-hexulose isomerase   \n",
       "22           dtdp-6-deoxy-3,4-keto-hexulose isomerase   \n",
       "23  Aminopeptidase; disrupted (overlaps with degT ...   \n",
       "24  Aminopeptidase; disrupted (overlaps with degT ...   \n",
       "25                                carbonic anyhydrase   \n",
       "26                                carbonic anyhydrase   \n",
       "27                                        transferase   \n",
       "28                                        transferase   \n",
       "29                                  Methyltransferase   \n",
       "30                                  Methyltransferase   \n",
       "31                                       PseD protein   \n",
       "32                                       PseD protein   \n",
       "33                                       PseD protein   \n",
       "34                               hypothetical protein   \n",
       "35                               hypothetical protein   \n",
       "36                   C4-dicarboxylate ABC transporter   \n",
       "37                   C4-dicarboxylate ABC transporter   \n",
       "38                                  methyltransferase   \n",
       "39                                  methyltransferase   \n",
       "40                        GDP-mannose 4,6-dehydratase   \n",
       "41                              Hypothetical proteins   \n",
       "42                              Hypothetical proteins   \n",
       "43                                  sugar transferase   \n",
       "44                                  sugar transferase   \n",
       "45                            chemotaxis protein CheY   \n",
       "46                                 chemotaxis protein   \n",
       "47              Chemotaxis protein, Flagellar Protein   \n",
       "\n",
       "   On/Off State In inoculum            On/Off State in Isolates  \\\n",
       "0            <10% Truncation                                 On   \n",
       "1            <10% Truncation                                 On   \n",
       "2                         On                                Off   \n",
       "3                         On                    <10% Truncation   \n",
       "4                         On                    <10% Truncation   \n",
       "5                         On                    <10% Truncation   \n",
       "6                         On                                Off   \n",
       "7                         On                                 On   \n",
       "8                         On                                Off   \n",
       "9                         On                                Off   \n",
       "10                        On                                 On   \n",
       "11                        On                                 On   \n",
       "12                        On                                Off   \n",
       "13                        On                                Off   \n",
       "14                       Off                                Off   \n",
       "15                       Off                                 On   \n",
       "16                       Off                                Off   \n",
       "17           <10% Truncation                    <10% Truncation   \n",
       "18                        On                                 On   \n",
       "19                        On                                 On   \n",
       "20                       Off                                 On   \n",
       "21                        On                    <10% Truncation   \n",
       "22                        On                                Off   \n",
       "23                       Off                                 On   \n",
       "24                       Off                                 On   \n",
       "25                        On                                 On   \n",
       "26                        On                                 On   \n",
       "27                        On                                 On   \n",
       "28                        On                                 On   \n",
       "29                        On                    <10% Truncation   \n",
       "30                        On                    <10% Truncation   \n",
       "31                        On                                Off   \n",
       "32                        On                                Off   \n",
       "33                        On                                Off   \n",
       "34                        On                                Off   \n",
       "35                        On                                Off   \n",
       "36                       NaN                                NaN   \n",
       "37                       NaN                                NaN   \n",
       "38                        On                                Off   \n",
       "39                        On                                Off   \n",
       "40                        On                                Off   \n",
       "41                        On                                Off   \n",
       "42                        On                                Off   \n",
       "43                        On                                 On   \n",
       "44                        On                                 On   \n",
       "45                   Unknown  On (Wild-type reference sequence)   \n",
       "46                        On                                Off   \n",
       "47                   Unknown                            Unknown   \n",
       "\n",
       "   Variant Called in Inoculum or Isolate Population?  \\\n",
       "0                                           Inoculum   \n",
       "1                                           Inoculum   \n",
       "2                               Infection Isolate(s)   \n",
       "3                               Infection Isolate(s)   \n",
       "4                               Infection Isolate(s)   \n",
       "5                               Infection Isolate(s)   \n",
       "6                               Infection Isolate(s)   \n",
       "7                               Infection Isolate(s)   \n",
       "8                               Infection Isolate(s)   \n",
       "9                               Infection Isolate(s)   \n",
       "10                              Infection Isolate(s)   \n",
       "11                              Infection Isolate(s)   \n",
       "12                                          Inoculum   \n",
       "13                              Infection Isolate(s)   \n",
       "14                              Infection Isolate(s)   \n",
       "15                                          Inoculum   \n",
       "16                              Infection Isolate(s)   \n",
       "17                              Infection Isolate(s)   \n",
       "18                                          Inoculum   \n",
       "19                                          Inoculum   \n",
       "20                              Infection Isolate(s)   \n",
       "21                              Infection Isolate(s)   \n",
       "22                              Infection Isolate(s)   \n",
       "23                              Infection Isolate(s)   \n",
       "24                              Infection Isolate(s)   \n",
       "25                              Infection Isolate(s)   \n",
       "26                              Infection Isolate(s)   \n",
       "27                                          Inoculum   \n",
       "28                              Infection Isolate(s)   \n",
       "29                              Infection Isolate(s)   \n",
       "30                              Infection Isolate(s)   \n",
       "31                              Infection Isolate(s)   \n",
       "32                              Infection Isolate(s)   \n",
       "33                              Infection Isolate(s)   \n",
       "34                              Infection Isolate(s)   \n",
       "35                              Infection Isolate(s)   \n",
       "36                              Infection Isolate(s)   \n",
       "37                              Infection Isolate(s)   \n",
       "38                              Infection Isolate(s)   \n",
       "39                              Infection Isolate(s)   \n",
       "40                              Infection Isolate(s)   \n",
       "41                              Infection Isolate(s)   \n",
       "42                              Infection Isolate(s)   \n",
       "43                              Infection Isolate(s)   \n",
       "44                              Infection Isolate(s)   \n",
       "45                                          Inoculum   \n",
       "46                              Infection Isolate(s)   \n",
       "47                              Infection Isolate(s)   \n",
       "\n",
       "   Isolate Population Samples That Differed From Inoculum  \\\n",
       "0   Relapse 8.2 , Relapse 8.1 , Relapse 1 , Relaps...       \n",
       "1   Relapse 8.1 , Relapse 1 , Relapse 4 , Relapse ...       \n",
       "2                                          16D5, 16D6       \n",
       "3                                          Relapse 4        \n",
       "4                             Relapse 4 , Relapse 31        \n",
       "5                            Relapse 8.2 , Relapse 9        \n",
       "6                              Relapse 31 , 36D1, 8D3       \n",
       "7                                          11D5, 11D6       \n",
       "8                                         Relapse 31        \n",
       "9                                          Relapse 4        \n",
       "10                                       Relapse 8.2        \n",
       "11                                         36D2, 32D2       \n",
       "12  Relapse 8.2 , 14D1, 21D1, 24D1, 16D0, 16D3, 34...       \n",
       "13                                 Relapse 8.2 , 32D2       \n",
       "14             Relapse 1 , Relapse 9 , 36D1, 1D3, 8D3       \n",
       "15  Relapse 8.2 , Relapse 8.1 , Relapse 1 , Relaps...       \n",
       "16                                       Relapse 8.2        \n",
       "17                                               11D1       \n",
       "18          Relapse 4 , Relapse 9 , Relapse 31 , 16D3       \n",
       "19  Relapse 8.2 , Relapse 8.1 , Relapse 1 , Relaps...       \n",
       "20  Relapse 8.2 , Relapse 8.1 , Relapse 4 , Relaps...       \n",
       "21                                 Relapse 8.2 , 14D1       \n",
       "22  Relapse 8.1 , Relapse 4 , Relapse 9 , Relapse ...       \n",
       "23                                         Relapse 9        \n",
       "24  Relapse 8.2 , Relapse 1 , Relapse 4 , Relapse 31        \n",
       "25                                        Relapse 31        \n",
       "26                           Relapse 8.2 , Relapse 1        \n",
       "27  Relapse 4 , 14D0, 14D1, 14D2, 14D3, 1D5, 21D1,...       \n",
       "28                                               16D0       \n",
       "29                             Relapse 4 , Relapse 9        \n",
       "30                            Relapse 31 , 16D0, 34D0       \n",
       "31                                         Relapse 4        \n",
       "32                                         Relapse 4        \n",
       "33                                       Relapse 8.2        \n",
       "34                           Relapse 8.1 , Relapse 9        \n",
       "35                                         Relapse 1        \n",
       "36                   Relapse 8.2 , Relapse 8.1 , 34D0       \n",
       "37                                               14D0       \n",
       "38                                  Relapse 31 , 16D6       \n",
       "39  Relapse 8.2 , Relapse 1 , 24D2, 4D2, 16D0, 33D...       \n",
       "40                                         Relapse 9        \n",
       "41                                        Relapse 31        \n",
       "42  Relapse 8.2 , Relapse 8.1 , 14D3, 24D1, 31D2, ...       \n",
       "43  Relapse 8.1 , Relapse 1 , Relapse 4 , Relapse ...       \n",
       "44                                       Relapse 8.2        \n",
       "45  Relapse 8.2 , Relapse 8.1 , Relapse 1 , Relaps...       \n",
       "46                                         16D5, 16D6       \n",
       "47                              14D1, 11D4, 1D2, 11D1       \n",
       "\n",
       "    Relapse Samples with Variant (out of 6)  Relapse Sample Frequency  \\\n",
       "0                                         4                 66.666667   \n",
       "1                                         4                 66.666667   \n",
       "2                                         0                  0.000000   \n",
       "3                                         1                 16.666667   \n",
       "4                                         2                 33.333333   \n",
       "5                                         2                 33.333333   \n",
       "6                                         1                 16.666667   \n",
       "7                                         0                  0.000000   \n",
       "8                                         1                 16.666667   \n",
       "9                                         1                 16.666667   \n",
       "10                                        1                 16.666667   \n",
       "11                                        0                  0.000000   \n",
       "12                                        1                 16.666667   \n",
       "13                                        1                 16.666667   \n",
       "14                                        2                 33.333333   \n",
       "15                                        6                100.000000   \n",
       "16                                        1                 16.666667   \n",
       "17                                        0                  0.000000   \n",
       "18                                        3                 50.000000   \n",
       "19                                        6                100.000000   \n",
       "20                                        5                 83.333333   \n",
       "21                                        1                 16.666667   \n",
       "22                                        4                 66.666667   \n",
       "23                                        1                 16.666667   \n",
       "24                                        4                 66.666667   \n",
       "25                                        1                 16.666667   \n",
       "26                                        2                 33.333333   \n",
       "27                                        1                 16.666667   \n",
       "28                                        0                  0.000000   \n",
       "29                                        2                 33.333333   \n",
       "30                                        1                 16.666667   \n",
       "31                                        1                 16.666667   \n",
       "32                                        1                 16.666667   \n",
       "33                                        1                 16.666667   \n",
       "34                                        2                 33.333333   \n",
       "35                                        1                 16.666667   \n",
       "36                                        2                 33.333333   \n",
       "37                                        0                  0.000000   \n",
       "38                                        1                 16.666667   \n",
       "39                                        2                 33.333333   \n",
       "40                                        1                 16.666667   \n",
       "41                                        1                 16.666667   \n",
       "42                                        2                 33.333333   \n",
       "43                                        4                 66.666667   \n",
       "44                                        1                 16.666667   \n",
       "45                                        6                100.000000   \n",
       "46                                        0                  0.000000   \n",
       "47                                        0                  0.000000   \n",
       "\n",
       "    Primary Infection Samples with Variant (out of 43)  \\\n",
       "0                                                  12    \n",
       "1                                                  43    \n",
       "2                                                   2    \n",
       "3                                                   0    \n",
       "4                                                   0    \n",
       "5                                                   0    \n",
       "6                                                   2    \n",
       "7                                                   2    \n",
       "8                                                   0    \n",
       "9                                                   0    \n",
       "10                                                  0    \n",
       "11                                                  2    \n",
       "12                                                  8    \n",
       "13                                                  1    \n",
       "14                                                  3    \n",
       "15                                                  8    \n",
       "16                                                  0    \n",
       "17                                                  1    \n",
       "18                                                  1    \n",
       "19                                                 43    \n",
       "20                                                 39    \n",
       "21                                                  1    \n",
       "22                                                  2    \n",
       "23                                                  0    \n",
       "24                                                  0    \n",
       "25                                                  0    \n",
       "26                                                  0    \n",
       "27                                                 42    \n",
       "28                                                  1    \n",
       "29                                                  0    \n",
       "30                                                  2    \n",
       "31                                                  0    \n",
       "32                                                  0    \n",
       "33                                                  0    \n",
       "34                                                  0    \n",
       "35                                                  0    \n",
       "36                                                  1    \n",
       "37                                                  1    \n",
       "38                                                  1    \n",
       "39                                                  7    \n",
       "40                                                  0    \n",
       "41                                                  0    \n",
       "42                                                 12    \n",
       "43                                                 37    \n",
       "44                                                  0    \n",
       "45                                                 43    \n",
       "46                                                  2    \n",
       "47                                                  4    \n",
       "\n",
       "    Primary Infection Sample Frequency  \\\n",
       "0                            27.906977   \n",
       "1                           100.000000   \n",
       "2                             4.651163   \n",
       "3                             0.000000   \n",
       "4                             0.000000   \n",
       "5                             0.000000   \n",
       "6                             4.651163   \n",
       "7                             4.651163   \n",
       "8                             0.000000   \n",
       "9                             0.000000   \n",
       "10                            0.000000   \n",
       "11                            4.651163   \n",
       "12                           18.604651   \n",
       "13                            2.325581   \n",
       "14                            6.976744   \n",
       "15                           18.604651   \n",
       "16                            0.000000   \n",
       "17                            2.325581   \n",
       "18                            2.325581   \n",
       "19                          100.000000   \n",
       "20                           90.697674   \n",
       "21                            2.325581   \n",
       "22                            4.651163   \n",
       "23                            0.000000   \n",
       "24                            0.000000   \n",
       "25                            0.000000   \n",
       "26                            0.000000   \n",
       "27                           97.674419   \n",
       "28                            2.325581   \n",
       "29                            0.000000   \n",
       "30                            4.651163   \n",
       "31                            0.000000   \n",
       "32                            0.000000   \n",
       "33                            0.000000   \n",
       "34                            0.000000   \n",
       "35                            0.000000   \n",
       "36                            2.325581   \n",
       "37                            2.325581   \n",
       "38                            2.325581   \n",
       "39                           16.279070   \n",
       "40                            0.000000   \n",
       "41                            0.000000   \n",
       "42                           27.906977   \n",
       "43                           86.046512   \n",
       "44                            0.000000   \n",
       "45                          100.000000   \n",
       "46                            4.651163   \n",
       "47                            9.302326   \n",
       "\n",
       "                 Annotation Used for this Study  \\\n",
       "0                              fedA Hemerythrin   \n",
       "1                              fedA Hemerythrin   \n",
       "2   dcuA Anaerobic C4-dicarboxylate Transporter   \n",
       "3               Conserved Hypothetical Proteins   \n",
       "4               Conserved Hypothetical Proteins   \n",
       "5                          Hypothetical Protein   \n",
       "6                           thiF Family Protein   \n",
       "7                  motA Falgellar Motor Protein   \n",
       "8   cmeR Bile Sensing Transcriptional Regulator   \n",
       "9   cmeR Bile Sensing Transcriptional Regulator   \n",
       "10                 Membrane GTP Binding Protein   \n",
       "11                 Membrane GTP Binding Protein   \n",
       "12                  Flagella Modification Gene*   \n",
       "13                  Flagella Modification Gene*   \n",
       "14        Potassium Transporter KtrB; disrupted   \n",
       "15                        cipA Invasion Protein   \n",
       "16                        cipA Invasion Protein   \n",
       "17              Conserved Hypothetical Proteins   \n",
       "18              Conserved Hypothetical Proteins   \n",
       "19        apt Adenine Phosphoribosyltransferase   \n",
       "20                      cjeI Restriction Enzyme   \n",
       "21     dTDP-6-deoxy-3,4-keto-hexulose isomerase   \n",
       "22     dTDP-6-deoxy-3,4-keto-hexulose isomerase   \n",
       "23                  O-linked Glycosylation Gene   \n",
       "24                  O-linked Glycosylation Gene   \n",
       "25                  O-linked Glycosylation Gene   \n",
       "26                  O-linked Glycosylation Gene   \n",
       "27                  O-linked Glycosylation Gene   \n",
       "28                  O-linked Glycosylation Gene   \n",
       "29                  O-linked Glycosylation ptmH   \n",
       "30                  O-linked Glycosylation ptmH   \n",
       "31                 O-linked Glycosylation pseD    \n",
       "32                 O-linked Glycosylation pseD    \n",
       "33                 O-linked Glycosylation pseD    \n",
       "34                  O-linked Glycosylation maf7   \n",
       "35                  O-linked Glycosylation maf7   \n",
       "36             C4-dicarboxylate ABC transporter   \n",
       "37             C4-dicarboxylate ABC transporter   \n",
       "38                   Putative Methyltransferase   \n",
       "39                   Putative Methyltransferase   \n",
       "40            wcbK, GDP-mannose 4,6-dehydratase   \n",
       "41               Conserved Hypothetical Protein   \n",
       "42               Conserved Hypothetical Protein   \n",
       "43                       wcA Sugar Transferase    \n",
       "44                       wcA Sugar Transferase    \n",
       "45             Cjj81176_1483 response regulator   \n",
       "46          Methyl-accepting chemotaxis protein   \n",
       "47          Methyl-accepting chemotaxis protein   \n",
       "\n",
       "                                Variant Location           var_locii  \n",
       "0                    IN GENE, IN PROMOTER REGION      (67639, 67640)  \n",
       "1                    IN GENE, IN PROMOTER REGION      (67639, 67640)  \n",
       "2                                        IN GENE    (100530, 100531)  \n",
       "3                    IN GENE, IN PROMOTER REGION    (168690, 168691)  \n",
       "4                                        IN GENE    (168690, 168691)  \n",
       "5                                        IN GENE           (168691,)  \n",
       "6                                        IN GENE           (268276,)  \n",
       "7                                        IN GENE           (302194,)  \n",
       "8                                        IN GENE    (335945, 335946)  \n",
       "9                                        IN GENE           (335972,)  \n",
       "10  BETWEEN GENES, 22BP DOWNSTREAM/56BP UPSTREAM    (517620, 517621)  \n",
       "11  BETWEEN GENES, 22BP DOWNSTREAM/56BP UPSTREAM           (517621,)  \n",
       "12                                        IN CDS           (565752,)  \n",
       "13                                        IN CDS    (565752, 565753)  \n",
       "14                                        IN CDS           (620468,)  \n",
       "15                                       IN GENE    (629686, 629687)  \n",
       "16                                       IN GENE           (629687,)  \n",
       "17     IN GENE, IN PROMOTER REGION 53BP UPSTREAM           (680285,)  \n",
       "18                               21BP DOWNSTREAM           (686224,)  \n",
       "19                                       IN GENE           (861171,)  \n",
       "20                                        IN CDS           (982992,)  \n",
       "21                          IN GENE, IN PROMOTER          (1070389,)  \n",
       "22                                       IN GENE          (1071193,)  \n",
       "23                                        IN CDS  (1227256, 1227257)  \n",
       "24                                        IN CDS          (1227257,)  \n",
       "25                                 51BP UPSTREAM  (1234215, 1234216)  \n",
       "26                                 51BP UPSTREAM          (1234216,)  \n",
       "27                            IN PROMOTER REGION  (1248836, 1248837)  \n",
       "28                            IN PROMOTER REGION          (1248837,)  \n",
       "29                                       IN GENE  (1251713, 1251714)  \n",
       "30                                       IN GENE          (1251714,)  \n",
       "31                                       IN GENE  (1262397, 1262398)  \n",
       "32                                       IN GENE          (1262399,)  \n",
       "33                                       IN GENE  (1263109, 1263110)  \n",
       "34                                       IN GENE  (1272799, 1272800)  \n",
       "35                                       IN GENE          (1272800,)  \n",
       "36                                 40BP UPSTREAM  (1293109, 1293110)  \n",
       "37                                 40BP UPSTREAM          (1293110,)  \n",
       "38                                       IN GENE  (1352935, 1352936)  \n",
       "39                                       IN GENE          (1352936,)  \n",
       "40                                       IN GENE  (1358677, 1358678)  \n",
       "41                                       IN GENE  (1361225, 1361226)  \n",
       "42                                       IN GENE          (1361226,)  \n",
       "43                    INTERGENIC - 43BP UPSTREAM          (1363819,)  \n",
       "44                    INTERGENIC - 43BP UPSTREAM  (1363819, 1363820)  \n",
       "45                                       IN GENE  (1416012, 1416020)  \n",
       "46                                       IN GENE          (1485899,)  \n",
       "47                                       IN GENE  (1486346, 1486347)  "
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "var_file = '/Volumes/KeithSSD/variant-calling/misc_data/Crofts_Gene_Variants.xlsx'\n",
    "var_df = pd.read_excel(var_file, sheet_name='Spec_Variants', index_col=None)\n",
    "var_df['Variant Position'] = var_df['Variant Position'].astype(str).str.replace(\"\\.\\.\", ',')\n",
    "var_df['Variant Position'] = var_df['Variant Position'].str.replace(\"^\", ',')\n",
    "\n",
    "full_genome = gb_df.loc[0, 'sequence']\n",
    "\n",
    "lociii = []\n",
    "for var_i in var_df.index:\n",
    "    locii = str(var_df.loc[var_i, 'Variant Position']).replace(\"*\", \"\").split(\",\")\n",
    "    locii = tuple([int(i) for i in locii])\n",
    "    lociii.append(locii)\n",
    "    detectable_muts = var_df.loc[var_i, 'Variant Type'] in ['Deletion', 'SNV']\n",
    "    if len(locii) == 1 and detectable_muts:\n",
    "        assert full_genome[locii[0]-1] == var_df.at[var_i, 'Reference Sequence']\n",
    "    elif len(locii) == 2 and detectable_muts:\n",
    "        assert full_genome[locii[0]-1:locii[1]] == var_df.at[var_i, 'Reference Sequence']\n",
    "\n",
    "var_df['var_locii'] = pd.Series(index=var_df.index, data=lociii)\n",
    "\n",
    "var_df['Variant Location'] = var_df['Variant Location'].apply(lambda x: x.strip().upper())\n",
    "\n",
    "var_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### The following code block performs the Fisher Exact test and bonferroni correction on the variations observed. Individually, very few rise to the "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Annotation</th>\n",
       "      <th>Reported Significant Associations</th>\n",
       "      <th>Relapse w/Variant</th>\n",
       "      <th>Relapse w/o Variant</th>\n",
       "      <th>Primary Infection w/Variant</th>\n",
       "      <th>Primary Infections wo Variant</th>\n",
       "      <th>p_val</th>\n",
       "      <th>oddsratio</th>\n",
       "      <th>correct_p_val</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>fedA Hemerythrin</td>\n",
       "      <td>NaN</td>\n",
       "      <td>6</td>\n",
       "      <td>0</td>\n",
       "      <td>43</td>\n",
       "      <td>0</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>dcuA Anaerobic C4-dicarboxylate Transporter</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>6</td>\n",
       "      <td>2</td>\n",
       "      <td>41</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Conserved Hypothetical Proteins</td>\n",
       "      <td>Relapse Isolates</td>\n",
       "      <td>4</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>43</td>\n",
       "      <td>0.000071</td>\n",
       "      <td>inf</td>\n",
       "      <td>0.002124</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>thiF Family Protein</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>5</td>\n",
       "      <td>2</td>\n",
       "      <td>41</td>\n",
       "      <td>0.330167</td>\n",
       "      <td>4.100000</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>motA Falgellar Motor Protein</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>6</td>\n",
       "      <td>2</td>\n",
       "      <td>41</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>cmeR Bile Sensing Transcriptional Regulator</td>\n",
       "      <td>Relapse Isolates</td>\n",
       "      <td>2</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>43</td>\n",
       "      <td>0.012755</td>\n",
       "      <td>inf</td>\n",
       "      <td>0.382653</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Membrane GTP Binding Protein</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>5</td>\n",
       "      <td>2</td>\n",
       "      <td>41</td>\n",
       "      <td>0.330167</td>\n",
       "      <td>4.100000</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Flagella Modification Gene*</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>5</td>\n",
       "      <td>9</td>\n",
       "      <td>34</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.755556</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Potassium Transporter KtrB; disrupted</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2</td>\n",
       "      <td>4</td>\n",
       "      <td>3</td>\n",
       "      <td>40</td>\n",
       "      <td>0.106890</td>\n",
       "      <td>6.666667</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>cipA Invasion Protein</td>\n",
       "      <td>Relapse Isolates</td>\n",
       "      <td>6</td>\n",
       "      <td>0</td>\n",
       "      <td>8</td>\n",
       "      <td>35</td>\n",
       "      <td>0.000215</td>\n",
       "      <td>inf</td>\n",
       "      <td>0.006442</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>Conserved Hypothetical Proteins</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>6</td>\n",
       "      <td>1</td>\n",
       "      <td>42</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>Conserved Hypothetical Proteins</td>\n",
       "      <td>Relapse Isolates</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>42</td>\n",
       "      <td>0.004130</td>\n",
       "      <td>42.000000</td>\n",
       "      <td>0.123893</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>apt Adenine Phosphoribosyltransferase</td>\n",
       "      <td>NaN</td>\n",
       "      <td>6</td>\n",
       "      <td>0</td>\n",
       "      <td>43</td>\n",
       "      <td>0</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>cjeI Restriction Enzyme</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5</td>\n",
       "      <td>1</td>\n",
       "      <td>39</td>\n",
       "      <td>4</td>\n",
       "      <td>0.495198</td>\n",
       "      <td>0.512821</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>dTDP-6-deoxy-3,4-keto-hexulose isomerase</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>5</td>\n",
       "      <td>1</td>\n",
       "      <td>42</td>\n",
       "      <td>0.232143</td>\n",
       "      <td>8.400000</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>dTDP-6-deoxy-3,4-keto-hexulose isomerase</td>\n",
       "      <td>Relapse Isolates</td>\n",
       "      <td>4</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>41</td>\n",
       "      <td>0.000987</td>\n",
       "      <td>41.000000</td>\n",
       "      <td>0.029614</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>O-linked Glycosylation Gene</td>\n",
       "      <td>Relapse Isolates</td>\n",
       "      <td>5</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>43</td>\n",
       "      <td>0.000003</td>\n",
       "      <td>inf</td>\n",
       "      <td>0.000094</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>O-linked Glycosylation Gene</td>\n",
       "      <td>Relapse Isolates</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>43</td>\n",
       "      <td>0.001086</td>\n",
       "      <td>inf</td>\n",
       "      <td>0.032566</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>O-linked Glycosylation Gene</td>\n",
       "      <td>Primary Infeciton Isolates</td>\n",
       "      <td>1</td>\n",
       "      <td>5</td>\n",
       "      <td>42</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000019</td>\n",
       "      <td>0.004762</td>\n",
       "      <td>0.000556</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>O-linked Glycosylation ptmH</td>\n",
       "      <td>Relapse Isolates</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>2</td>\n",
       "      <td>41</td>\n",
       "      <td>0.009812</td>\n",
       "      <td>20.500000</td>\n",
       "      <td>0.294370</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>O-linked Glycosylation pseD</td>\n",
       "      <td>Relapse Isolates</td>\n",
       "      <td>2</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>43</td>\n",
       "      <td>0.012755</td>\n",
       "      <td>inf</td>\n",
       "      <td>0.382653</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>O-linked Glycosylation maf7</td>\n",
       "      <td>Relapse Isolates</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>43</td>\n",
       "      <td>0.001086</td>\n",
       "      <td>inf</td>\n",
       "      <td>0.032566</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>C4-dicarboxylate ABC transporter</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2</td>\n",
       "      <td>4</td>\n",
       "      <td>2</td>\n",
       "      <td>41</td>\n",
       "      <td>0.068059</td>\n",
       "      <td>10.250000</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>Putative Methyltransferase</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>8</td>\n",
       "      <td>35</td>\n",
       "      <td>0.117418</td>\n",
       "      <td>4.375000</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>wcbK, GDP-mannose 4,6-dehydratase</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "      <td>43</td>\n",
       "      <td>0.122449</td>\n",
       "      <td>inf</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>Conserved Hypothetical Protein</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>12</td>\n",
       "      <td>31</td>\n",
       "      <td>0.353301</td>\n",
       "      <td>2.583333</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>wcA Sugar Transferase</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5</td>\n",
       "      <td>1</td>\n",
       "      <td>37</td>\n",
       "      <td>6</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.810811</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>Cjj81176_1483 response regulator</td>\n",
       "      <td>NaN</td>\n",
       "      <td>6</td>\n",
       "      <td>0</td>\n",
       "      <td>43</td>\n",
       "      <td>0</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>Methyl-accepting chemotaxis protein</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>6</td>\n",
       "      <td>2</td>\n",
       "      <td>41</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>Methyl-accepting chemotaxis protein</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>6</td>\n",
       "      <td>4</td>\n",
       "      <td>39</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                     Annotation  \\\n",
       "0                              fedA Hemerythrin   \n",
       "1   dcuA Anaerobic C4-dicarboxylate Transporter   \n",
       "2               Conserved Hypothetical Proteins   \n",
       "3                           thiF Family Protein   \n",
       "4                  motA Falgellar Motor Protein   \n",
       "5   cmeR Bile Sensing Transcriptional Regulator   \n",
       "6                  Membrane GTP Binding Protein   \n",
       "7                   Flagella Modification Gene*   \n",
       "8         Potassium Transporter KtrB; disrupted   \n",
       "9                         cipA Invasion Protein   \n",
       "10              Conserved Hypothetical Proteins   \n",
       "11              Conserved Hypothetical Proteins   \n",
       "12        apt Adenine Phosphoribosyltransferase   \n",
       "13                      cjeI Restriction Enzyme   \n",
       "14     dTDP-6-deoxy-3,4-keto-hexulose isomerase   \n",
       "15     dTDP-6-deoxy-3,4-keto-hexulose isomerase   \n",
       "16                  O-linked Glycosylation Gene   \n",
       "17                  O-linked Glycosylation Gene   \n",
       "18                  O-linked Glycosylation Gene   \n",
       "19                  O-linked Glycosylation ptmH   \n",
       "20                 O-linked Glycosylation pseD    \n",
       "21                  O-linked Glycosylation maf7   \n",
       "22             C4-dicarboxylate ABC transporter   \n",
       "23                   Putative Methyltransferase   \n",
       "24            wcbK, GDP-mannose 4,6-dehydratase   \n",
       "25               Conserved Hypothetical Protein   \n",
       "26                       wcA Sugar Transferase    \n",
       "27             Cjj81176_1483 response regulator   \n",
       "28          Methyl-accepting chemotaxis protein   \n",
       "29          Methyl-accepting chemotaxis protein   \n",
       "\n",
       "   Reported Significant Associations  Relapse w/Variant  Relapse w/o Variant  \\\n",
       "0                                NaN                  6                    0   \n",
       "1                                NaN                  0                    6   \n",
       "2                   Relapse Isolates                  4                    2   \n",
       "3                                NaN                  1                    5   \n",
       "4                                NaN                  0                    6   \n",
       "5                   Relapse Isolates                  2                    4   \n",
       "6                                NaN                  1                    5   \n",
       "7                                NaN                  1                    5   \n",
       "8                                NaN                  2                    4   \n",
       "9                   Relapse Isolates                  6                    0   \n",
       "10                               NaN                  0                    6   \n",
       "11                  Relapse Isolates                  3                    3   \n",
       "12                               NaN                  6                    0   \n",
       "13                               NaN                  5                    1   \n",
       "14                               NaN                  1                    5   \n",
       "15                  Relapse Isolates                  4                    2   \n",
       "16                  Relapse Isolates                  5                    1   \n",
       "17                  Relapse Isolates                  3                    3   \n",
       "18        Primary Infeciton Isolates                  1                    5   \n",
       "19                  Relapse Isolates                  3                    3   \n",
       "20                  Relapse Isolates                  2                    4   \n",
       "21                  Relapse Isolates                  3                    3   \n",
       "22                               NaN                  2                    4   \n",
       "23                               NaN                  3                    3   \n",
       "24                               NaN                  1                    5   \n",
       "25                               NaN                  3                    3   \n",
       "26                               NaN                  5                    1   \n",
       "27                               NaN                  6                    0   \n",
       "28                               NaN                  0                    6   \n",
       "29                               NaN                  0                    6   \n",
       "\n",
       "    Primary Infection w/Variant  Primary Infections wo Variant     p_val  \\\n",
       "0                            43                              0  1.000000   \n",
       "1                             2                             41  1.000000   \n",
       "2                             0                             43  0.000071   \n",
       "3                             2                             41  0.330167   \n",
       "4                             2                             41  1.000000   \n",
       "5                             0                             43  0.012755   \n",
       "6                             2                             41  0.330167   \n",
       "7                             9                             34  1.000000   \n",
       "8                             3                             40  0.106890   \n",
       "9                             8                             35  0.000215   \n",
       "10                            1                             42  1.000000   \n",
       "11                            1                             42  0.004130   \n",
       "12                           43                              0  1.000000   \n",
       "13                           39                              4  0.495198   \n",
       "14                            1                             42  0.232143   \n",
       "15                            2                             41  0.000987   \n",
       "16                            0                             43  0.000003   \n",
       "17                            0                             43  0.001086   \n",
       "18                           42                              1  0.000019   \n",
       "19                            2                             41  0.009812   \n",
       "20                            0                             43  0.012755   \n",
       "21                            0                             43  0.001086   \n",
       "22                            2                             41  0.068059   \n",
       "23                            8                             35  0.117418   \n",
       "24                            0                             43  0.122449   \n",
       "25                           12                             31  0.353301   \n",
       "26                           37                              6  1.000000   \n",
       "27                           43                              0  1.000000   \n",
       "28                            2                             41  1.000000   \n",
       "29                            4                             39  1.000000   \n",
       "\n",
       "    oddsratio  correct_p_val  \n",
       "0         NaN       1.000000  \n",
       "1    0.000000       1.000000  \n",
       "2         inf       0.002124  \n",
       "3    4.100000       1.000000  \n",
       "4    0.000000       1.000000  \n",
       "5         inf       0.382653  \n",
       "6    4.100000       1.000000  \n",
       "7    0.755556       1.000000  \n",
       "8    6.666667       1.000000  \n",
       "9         inf       0.006442  \n",
       "10   0.000000       1.000000  \n",
       "11  42.000000       0.123893  \n",
       "12        NaN       1.000000  \n",
       "13   0.512821       1.000000  \n",
       "14   8.400000       1.000000  \n",
       "15  41.000000       0.029614  \n",
       "16        inf       0.000094  \n",
       "17        inf       0.032566  \n",
       "18   0.004762       0.000556  \n",
       "19  20.500000       0.294370  \n",
       "20        inf       0.382653  \n",
       "21        inf       0.032566  \n",
       "22  10.250000       1.000000  \n",
       "23   4.375000       1.000000  \n",
       "24        inf       1.000000  \n",
       "25   2.583333       1.000000  \n",
       "26   0.810811       1.000000  \n",
       "27        NaN       1.000000  \n",
       "28   0.000000       1.000000  \n",
       "29   0.000000       1.000000  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from statsmodels.stats.multitest import multipletests\n",
    "from scipy.stats import fisher_exact\n",
    "\n",
    "var_file = '/Volumes/KeithSSD/variant-calling/misc_data/Crofts_Variants_Pooled.xlsx'\n",
    "pooled_var_df = pd.read_excel(var_file, sheet_name='Sheet1', index_col=None)\n",
    "\n",
    "col1 = 'Relapse Sample Count (out of 6)'\n",
    "col2 = 'Relapse w/o Variant'\n",
    "col3 = 'Primary Infection Samples with Variant(s) (out of 43)'\n",
    "col4 = 'Primary Infections wo Variant'\n",
    "\n",
    "pooled_var_df[col2] = 6 - pooled_var_df[col1]\n",
    "pooled_var_df[col4] = 43 - pooled_var_df[col3]\n",
    "\n",
    "oddsratio_fxn = lambda x: fisher_exact([[x[0], x[2]], [x[1], x[3]]])[0]\n",
    "pvalue_fxn = lambda x: fisher_exact([[x[0], x[2]], [x[1], x[3]]])[1]\n",
    "\n",
    "pooled_var_df['oddsratio'] = pooled_var_df[[col1, col2, col3, col4]].apply(oddsratio_fxn,  axis=1)\n",
    "pooled_var_df['p_val'] = pooled_var_df[[col1, col2, col3, col4]].apply(pvalue_fxn,  axis=1)\n",
    "\n",
    "corr_p_vals = multipletests(pooled_var_df['p_val'].values, method='bonferroni', alpha=0.05)\n",
    "pooled_var_df['correct_p_val'] = pd.Series(corr_p_vals[1], index=pooled_var_df.index)\n",
    "pooled_var_df[\"Relapse w/Variant\"] = pooled_var_df[col1]\n",
    "pooled_var_df[\"Primary Infection w/Variant\"] = pooled_var_df[col3]\n",
    "\n",
    "pooled_var_df['Reported Significant Associations'] = pooled_var_df['Variant Association with Human Disease State as Determined by Fisher Exact Test (FDR p-value <0.05)']\n",
    "\n",
    "pooled_var_df[['Annotation', 'Reported Significant Associations', \"Relapse w/Variant\", col2, \n",
    "               \"Primary Infection w/Variant\", col4, 'p_val', 'oddsratio', 'correct_p_val']]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### I noticed that some of the locus tags listed in their Supplemental Data file have been replaced with new ones in the latest revision of the CG8421 genome, so this code goes through the variation dataframe and makes those same updates."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "10 locus tags were either changed or removed\n"
     ]
    }
   ],
   "source": [
    "# fix formatting\n",
    "var_df['affected'] = var_df[\"Affected CG8421 Gene(s)\"].str.replace(\", Promoter of \", \" & \")\n",
    "var_df['affected'] = var_df['affected'].apply(lambda x: x.split(\" & \"))\n",
    "\n",
    "# manually verified = [12, 13, 14, 25, 26, 29, 30, 41, 42]\n",
    "updated_tags = {\"CJ8421_RS03020\": \"CJ8421_RS08905\", \n",
    "                \"CJ8421_RS03305\": \"CJ8421_RS08915\", \n",
    "                \"CJ8421_RS06480\": \"CJ8421_RS08860\",\n",
    "                'CJ8421_RS06580': 'CJ8421_RS08870',\n",
    "                'CJ8421_RS06585': 'CJ8421_RS08870',\n",
    "                \"CJ8421_RS07090\": \"CJ8421_RS07085\"}\n",
    "def fix_tags(x):\n",
    "    y = []\n",
    "    for ip in x:\n",
    "        i = ip.strip()\n",
    "        if not i in updated_tags.keys():\n",
    "            y.append(i)\n",
    "        else:\n",
    "            y.append(updated_tags[i])\n",
    "    return list(set(y))\n",
    "\n",
    "var_df['affected_2'] = var_df['affected'].apply(fix_tags)\n",
    "changes_made = var_df[['affected', 'affected_2']].apply(lambda x: len(set(x[0]) - set(x[1])), axis=1).sum()\n",
    "print(\"{} locus tags were either changed or removed\".format(changes_made))\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Here we will confirm the adjacency of each mutation and its affected gene. In the same script, we can add a column to indicate whether the mutation was detected within or between genes and another to collect the identifiers for additional nearby features that may add support for whether a paralog to the target region in the CG8421 genome is detected in our Peruvian isolates."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0. 2 possible genes affected by mutation at 67639\n",
      "1. 2 possible genes affected by mutation at 67639\n",
      "2. 1 possible genes affected by mutation at 100530\n",
      "3. 2 possible genes affected by mutation at 168690\n",
      "4. 2 possible genes affected by mutation at 168690\n",
      "5. 2 possible genes affected by mutation at 168691\n",
      "6. 1 possible genes affected by mutation at 268276\n",
      "7. 1 possible genes affected by mutation at 302194\n",
      "8. 1 possible genes affected by mutation at 335945\n",
      "9. 1 possible genes affected by mutation at 335972\n",
      "10. 2 possible genes affected by mutation at 517620\n",
      "11. 2 possible genes affected by mutation at 517621\n",
      "12. 1 possible genes affected by mutation at 565752\n",
      "13. 1 possible genes affected by mutation at 565752\n",
      "14. 1 possible genes affected by mutation at 620468\n",
      "15. 1 possible genes affected by mutation at 629686\n",
      "16. 1 possible genes affected by mutation at 629687\n",
      "17. 2 possible genes affected by mutation at 680285\n",
      "18. 1 possible genes affected by mutation at 686224\n",
      "19. 1 possible genes affected by mutation at 861171\n",
      "20. 1 possible genes affected by mutation at 982992\n",
      "21. 2 possible genes affected by mutation at 1070389\n",
      "22. 2 possible genes affected by mutation at 1071193\n",
      "23. 2 possible genes affected by mutation at 1227256\n",
      "24. 2 possible genes affected by mutation at 1227257\n",
      "25. 1 possible genes affected by mutation at 1234215\n",
      "26. 1 possible genes affected by mutation at 1234216\n",
      "27. 2 possible genes affected by mutation at 1248836\n",
      "28. 2 possible genes affected by mutation at 1248837\n",
      "29. 1 possible genes affected by mutation at 1251713\n",
      "30. 1 possible genes affected by mutation at 1251714\n",
      "31. 1 possible genes affected by mutation at 1262397\n",
      "32. 1 possible genes affected by mutation at 1262399\n",
      "33. 1 possible genes affected by mutation at 1263109\n",
      "34. 1 possible genes affected by mutation at 1272799\n",
      "35. 1 possible genes affected by mutation at 1272800\n",
      "36. 1 possible genes affected by mutation at 1293109\n",
      "37. 1 possible genes affected by mutation at 1293110\n",
      "38. 1 possible genes affected by mutation at 1352935\n",
      "39. 1 possible genes affected by mutation at 1352936\n",
      "40. 1 possible genes affected by mutation at 1358677\n",
      "41. 1 possible genes affected by mutation at 1361225\n",
      "42. 1 possible genes affected by mutation at 1361226\n",
      "43. 2 possible genes affected by mutation at 1363819\n",
      "44. 2 possible genes affected by mutation at 1363819\n",
      "45. 1 possible genes affected by mutation at 1416012\n",
      "46. 1 possible genes affected by mutation at 1485899\n",
      "47. 2 possible genes affected by mutation at 1486346\n"
     ]
    }
   ],
   "source": [
    "# all gene & cds are 1:1 in the records & location searches will be hampered by the source rec\n",
    "ngb_df = gb_df.copy()[~gb_df['type'].isin(['gene', 'source'])]\n",
    "\n",
    "var_cols = ['Variant Position', \"Variant Type\", \"Reference Sequence\", \"Variant Allele\"]\n",
    "\n",
    "anchor_dicts, anch_counter = {}, 0 \n",
    "for var_i in var_df.index:\n",
    "    mut_pos = var_df.loc[var_i, 'var_locii'][0]\n",
    "    mut_us, mut_ds = mut_pos + 150, mut_pos - 150\n",
    "    coding_bool = ngb_df['type'] == 'CDS'\n",
    "    within_bool = (ngb_df['start'] <= mut_pos) & (ngb_df['end'] >= mut_pos) & (coding_bool)\n",
    "    downstream = (ngb_df['start'] > mut_pos) & (ngb_df['start'] <= mut_us) & (coding_bool)\n",
    "    upstream = (ngb_df['end'] < mut_pos) & (ngb_df['end'] >= mut_ds) & (coding_bool)\n",
    "    sensor_bools = [(downstream) & (ngb_df['strand'] == 1),\n",
    "                    (upstream) & (ngb_df['strand'] == 1), \n",
    "                    (downstream) & (ngb_df['strand'] == -1),\n",
    "                    (upstream) & (ngb_df['strand'] == -1), \n",
    "                    within_bool]\n",
    "     \n",
    "    sensor_types = ['upstream', 'downstream', 'upstream', 'downstream', 'in_CDS']\n",
    "    keys_needed = ['locus_tag', 'start', 'end', 'strand', 'product', 'sequence', 'pseudo']\n",
    "    all_pinged = within_bool.sum()+downstream.sum()+upstream.sum()\n",
    "     \n",
    "    print(\"{}. {} possible genes affected by mutation at {}\".format(var_i, all_pinged, mut_pos))\n",
    "    #print(\"\\t= {}\".format(var_df.loc[var_i, 'affected_2']))\n",
    "    #print(\"{}, {}\".format(mut_pos, ngb_df.loc[within_bool | upstream | downstream, keys_needed].T))\n",
    "    #input()\n",
    "     \n",
    "    for sensor, senseType in zip(sensor_bools, sensor_types):\n",
    "        if sensor.sum() > 0:\n",
    "            sensee = ngb_df.loc[sensor, keys_needed]\n",
    "            for hit_i in sensee.index:\n",
    "                pos_dict = {'variation_index':var_i, 'mut_start':mut_pos, 'var_location': senseType}\n",
    "                for vc in var_cols:\n",
    "                    pos_dict[vc] = var_df.loc[var_i, vc]\n",
    "                for kn in keys_needed:\n",
    "                    pos_dict[kn] = sensee.loc[hit_i, kn]\n",
    "                if pos_dict['strand'] == 1:\n",
    "                    pos_dict['upstream_seq'] = full_genome[pos_dict['start']-20-1:pos_dict['start']+180]\n",
    "                    pos_dict['downstream_seq'] = full_genome[pos_dict['end']-20-1:pos_dict['end']+180]\n",
    "                    pos_dict['upstream_region'] = (pos_dict['start']-20, pos_dict['start']+180)\n",
    "                    pos_dict['downstream_region'] = (pos_dict['end']-20, pos_dict['end']+180)\n",
    "                else:\n",
    "                    pos_dict['upstream_seq'] = full_genome[pos_dict['end']-20-1:pos_dict['end']+180]\n",
    "                    pos_dict['downstream_seq'] = full_genome[pos_dict['start']-20-1:pos_dict['start']+180]\n",
    "                    pos_dict['upstream_region'] = (pos_dict['end']-20, pos_dict['end']+180)\n",
    "                    pos_dict['downstream_region'] = (pos_dict['start']-20, pos_dict['start']+180)\n",
    "                anchor_dicts[anch_counter] = pos_dict\n",
    "                anch_counter += 1"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Here we are going to add some genes we are fans of and some random genes to search for just to get a sense of the range of matches observed in the data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### To determine if our mutation is conserved we will look for the interrupted gene, as well as some nearby genes to determine how conserved the region is. To do this, we will extract the following anchor sequences per feature:\n",
    "#### -  75 nucleotides on either side of mutation \n",
    "#### -  Gene sequences of 2 closest genes on either side and target gene\n",
    "#### - Protein sequences for all gene anchors"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Adding 20 genes to search for, currently there are: 65\n",
      "Total of 85 genes to discover i.e. 255 primers\n",
      "Total of 85 genes to discover i.e. 255 primers\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Reference Sequence</th>\n",
       "      <th>Variant Allele</th>\n",
       "      <th>Variant Position</th>\n",
       "      <th>Variant Type</th>\n",
       "      <th>downstream_region</th>\n",
       "      <th>downstream_seq</th>\n",
       "      <th>end</th>\n",
       "      <th>locus_tag</th>\n",
       "      <th>mut_start</th>\n",
       "      <th>product</th>\n",
       "      <th>pseudo</th>\n",
       "      <th>sequence</th>\n",
       "      <th>start</th>\n",
       "      <th>strand</th>\n",
       "      <th>upstream_region</th>\n",
       "      <th>upstream_seq</th>\n",
       "      <th>var_location</th>\n",
       "      <th>variation_index</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>G</td>\n",
       "      <td>-</td>\n",
       "      <td>565752</td>\n",
       "      <td>Deletion</td>\n",
       "      <td>(566403, 566603)</td>\n",
       "      <td>ATTTTAAATATAACAAATTGAGTATATTAAATGTTTATATTTAAAA...</td>\n",
       "      <td>566423</td>\n",
       "      <td>CJ8421_RS08905</td>\n",
       "      <td>565752</td>\n",
       "      <td>DUF2920 family protein</td>\n",
       "      <td></td>\n",
       "      <td>ATGATTATAAACCAAATTTATTCTATAGATTCTTGTGATGATGTAG...</td>\n",
       "      <td>565192</td>\n",
       "      <td>1</td>\n",
       "      <td>(565172, 565372)</td>\n",
       "      <td>ATTTCCAAAAAAAGGTTTAATATGATTATAAACCAAATTTATTCTA...</td>\n",
       "      <td>in_CDS</td>\n",
       "      <td>12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>GG</td>\n",
       "      <td>-</td>\n",
       "      <td>565752,565753</td>\n",
       "      <td>Deletion</td>\n",
       "      <td>(566403, 566603)</td>\n",
       "      <td>ATTTTAAATATAACAAATTGAGTATATTAAATGTTTATATTTAAAA...</td>\n",
       "      <td>566423</td>\n",
       "      <td>CJ8421_RS08905</td>\n",
       "      <td>565752</td>\n",
       "      <td>DUF2920 family protein</td>\n",
       "      <td></td>\n",
       "      <td>ATGATTATAAACCAAATTTATTCTATAGATTCTTGTGATGATGTAG...</td>\n",
       "      <td>565192</td>\n",
       "      <td>1</td>\n",
       "      <td>(565172, 565372)</td>\n",
       "      <td>ATTTCCAAAAAAAGGTTTAATATGATTATAAACCAAATTTATTCTA...</td>\n",
       "      <td>in_CDS</td>\n",
       "      <td>13</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Reference Sequence Variant Allele Variant Position Variant Type  \\\n",
       "19                  G              -           565752     Deletion   \n",
       "20                 GG              -    565752,565753     Deletion   \n",
       "\n",
       "   downstream_region                                     downstream_seq  \\\n",
       "19  (566403, 566603)  ATTTTAAATATAACAAATTGAGTATATTAAATGTTTATATTTAAAA...   \n",
       "20  (566403, 566603)  ATTTTAAATATAACAAATTGAGTATATTAAATGTTTATATTTAAAA...   \n",
       "\n",
       "       end       locus_tag mut_start                 product pseudo  \\\n",
       "19  566423  CJ8421_RS08905    565752  DUF2920 family protein          \n",
       "20  566423  CJ8421_RS08905    565752  DUF2920 family protein          \n",
       "\n",
       "                                             sequence   start strand  \\\n",
       "19  ATGATTATAAACCAAATTTATTCTATAGATTCTTGTGATGATGTAG...  565192      1   \n",
       "20  ATGATTATAAACCAAATTTATTCTATAGATTCTTGTGATGATGTAG...  565192      1   \n",
       "\n",
       "     upstream_region                                       upstream_seq  \\\n",
       "19  (565172, 565372)  ATTTCCAAAAAAAGGTTTAATATGATTATAAACCAAATTTATTCTA...   \n",
       "20  (565172, 565372)  ATTTCCAAAAAAAGGTTTAATATGATTATAAACCAAATTTATTCTA...   \n",
       "\n",
       "   var_location variation_index  \n",
       "19       in_CDS              12  \n",
       "20       in_CDS              13  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "conserved_ = ngb_df['locus_tag'].isin(['CJ8421_RS08380', 'CJ8421_RS00020', 'CJ8421_RS00430',\n",
    "                                       'CJ8421_RS00435', 'CJ8421_RS02005', 'CJ8421_RS05885', \n",
    "                                       'CJ8421_RS06285', 'CJ8421_RS07345', 'CJ8421_RS07350', \n",
    "                                       'CJ8421_RS07355'])\n",
    "\n",
    "random_recs = ngb_df[coding_bool].sample(n=10, random_state=1)\n",
    "random_recs = random_recs.append(ngb_df[conserved_])\n",
    "\n",
    "print(\"Adding {} genes to search for, currently there are: {}\".format(len(random_recs), \n",
    "                                                                      len(anchor_dicts)))\n",
    "for hit_i in random_recs.index:\n",
    "    pos_dict = {}\n",
    "    for vc in var_cols+['variation_index', 'mut_start', 'var_location']:\n",
    "        pos_dict[vc] = np.nan\n",
    "    for kn in keys_needed:\n",
    "        pos_dict[kn] = ngb_df.loc[hit_i, kn]\n",
    "    if pos_dict['strand'] == 1:\n",
    "        pos_dict['upstream_seq'] = full_genome[pos_dict['start']-20-1:pos_dict['start']+180]\n",
    "        pos_dict['downstream_seq'] = full_genome[pos_dict['end']-20-1:pos_dict['end']+180]\n",
    "        pos_dict['upstream_region'] = (pos_dict['start']-20, pos_dict['start']+180)\n",
    "        pos_dict['downstream_region'] = (pos_dict['end']-20, pos_dict['end']+180)\n",
    "    else:\n",
    "        pos_dict['upstream_seq'] = full_genome[pos_dict['end']-20-1:pos_dict['end']+180]\n",
    "        pos_dict['downstream_seq'] = full_genome[pos_dict['start']-20-1:pos_dict['start']+180]\n",
    "        pos_dict['upstream_region'] = (pos_dict['end']-20, pos_dict['end']+180)\n",
    "        pos_dict['downstream_region'] = (pos_dict['start']-20, pos_dict['start']+180)\n",
    "    anchor_dicts[anch_counter] = pos_dict\n",
    "    anch_counter += 1\n",
    "\n",
    "anchor_df = pd.DataFrame(anchor_dicts).T\n",
    "print(\"Total of {} genes to discover i.e. {} primers\".format(len(anchor_df), len(anchor_df)*3))\n",
    "anchor_df = anchor_df.drop_duplicates()\n",
    "print(\"Total of {} genes to discover i.e. {} primers\".format(len(anchor_df), len(anchor_df)*3))\n",
    "anchor_df.loc[[19, 20], :]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Here we create a new folder for all data written to disk.\n",
    "#### Then we write out both anchor files into files called reference_anchors."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      " /Volumes/KeithSSD/variant-calling/variant_homologs/gene_primers.fa written\n"
     ]
    }
   ],
   "source": [
    "anchor_df['before_header'] = anchor_df['locus_tag'].apply(lambda x: \"{}-before\".format(x))\n",
    "anchor_df['after_header'] = anchor_df['locus_tag'].apply(lambda x: \"{}-after\".format(x))\n",
    "anchor_df['gene_header'] = anchor_df['locus_tag'].apply(lambda x: \"{}-gene\".format(x))\n",
    "\n",
    "def table_to_fasta(h_col, s_col, df, outname):\n",
    "    output = list(df[[h_col, s_col]].apply(lambda x: \">\"+str(x[0])+\"\\n\"+x[1], axis=1))\n",
    "    with open(outname, 'w') as fh:\n",
    "        fh.write(\"\\n\".join(output)+\"\\n\")\n",
    "    print(\"\\n\", outname, \"written\")\n",
    "    return\n",
    "\n",
    "data_dir = \"/Volumes/KeithSSD/variant-calling/variant_homologs\"\n",
    "if not os.path.exists(data_dir):\n",
    "    os.mkdir(data_dir)\n",
    "\n",
    "gene_file = os.path.join(data_dir, 'gene_primers.fa')\n",
    "\n",
    "col_pairs = [('downstream_seq', 'before_header'), \n",
    "             ('upstream_seq', 'after_header'), \n",
    "             ('sequence', 'gene_header')]\n",
    "\n",
    "outdf = pd.DataFrame(columns=['headers', 'nucleotides'], \n",
    "                     data=[list(x) for s, h in col_pairs for x in anchor_df[[h, s]].values]).drop_duplicates()\n",
    "\n",
    "table_to_fasta('headers', 'nucleotides', outdf, gene_file)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Here we will read in a text file containing information on the primary infection and relapse isolate pairs identified in both our and the Crofts et al. studies.  We use this information, and two columns from the Crofts supplement to define a new column, each row of which contains a list of the isolates expected to have a mutation in a given location "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "24 relapse pair strains identified, including:\n",
      "\t['CAMPY82_spr11', 'pCAMPY80', 'CAMPY87_S44', 'crofts_8R2', 'crofts_4D1', 'CAMPY15_S7', 'crofts_8D3', 'crofts_4R', 'crofts_9D2', 'pCAMPY22', 'crofts_8R1', 'CAMPY4_spr11', 'pCAMPY47', 'CAMPY28', 'crofts_1R', 'crofts_31R', 'CAMPY12_S6', 'crofts_9R', 'CAMPY41_S20', 'CAMPY10', 'crofts_8D2', 'crofts_1D2', 'crofts_31D2', 'pCAMPY86']\n",
      "\n",
      "At 34 after adding the rest of Crofts isolates.\n"
     ]
    }
   ],
   "source": [
    "metadata_file = \"/Volumes/KeithSSD/CA_V3/data/10_functional_enrichment_final/enrichment_categories.txt\"\n",
    "metadata_df = pd.read_csv(metadata_file, sep=\"\\t\", index_col=0)\n",
    "relapse_pair_isos = list(metadata_df[metadata_df['primary_v_relapse'].notnull()].index)\n",
    "print(\"{} relapse pair strains identified, including:\\n\\t{}\\n\".format(len(relapse_pair_isos), relapse_pair_isos))\n",
    "all_crofts_ = [i for i in metadata_df.index if 'crofts' in i]\n",
    "relapse_pair_isos = list(set(relapse_pair_isos + all_crofts_))\n",
    "print(\"At {} after adding the rest of Crofts isolates.\".format(len(relapse_pair_isos)))\n",
    "\n",
    "def fix_pop_col(x):\n",
    "    f1 = lambda x: x.replace(\"Relapse 8.\", \"crofts_8R\")\n",
    "    f2 = lambda x: x.replace(\"Relapse 1\", \"crofts_1R\")\n",
    "    f3 = lambda x: x.replace(\"Relapse 9\", \"crofts_31R\")\n",
    "    f4 = lambda x: x.replace(\"Relapse 4\", \"crofts_9R\")\n",
    "    f5 = lambda x: x.replace(\"Relapse 31\", \"crofts_4R\")\n",
    "    f6 = lambda x: \"crofts_\" + x if 'D' in x else x\n",
    "    f7 = lambda x: f6(f1(f2(f3(f4(f5(x.strip()))))))\n",
    "    return list(map(f7, x.split(\",\")))\n",
    "\n",
    "cn1 = 'Isolate Population Samples That Differed From Inoculum'\n",
    "cn2 = 'Variant Called in Inoculum or Isolate Population?'\n",
    "\n",
    "var_df['Isos_unlike_Innoc'] = var_df[cn1].apply(fix_pop_col)\n",
    "f8 = lambda x: set(['crofts_innoc']) if x[0] == 'Inoculum' else set(x[1]).intersection(set(relapse_pair_isos))\n",
    "f9 = lambda x: set(relapse_pair_isos).intersection(x[1]) if x[0] == 'Inoculum' else set()\n",
    "var_df['Samps_w_Mutation'] = var_df[[cn2, 'Isos_unlike_Innoc']].apply(f8, axis=1)\n",
    "var_df['Samps_off_Target'] = var_df[[cn2, 'Isos_unlike_Innoc']].apply(f9, axis=1)\n",
    "\n",
    "for add_cols_ in ['Samps_w_Mutation', 'Samps_off_Target']:\n",
    "    anchor_df[add_cols_] = anchor_df['variation_index'].map(var_df[add_cols_].to_dict())\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### We will now import full genome sequences and predicted protein sequences. The protein tables were exported from an'vio"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Reading protein and gene position files\n",
      "Reading assemblies\n",
      "Total # proteins read 147064\n",
      "Total # contigs read 34795\n",
      "              old_ids                      new_ids  \\\n",
      "34790  c_000000000014  CG8421_sim-c_000000000014-x   \n",
      "34791  c_000000000015  CG8421_sim-c_000000000015-x   \n",
      "34792  c_000000000016  CG8421_sim-c_000000000016-x   \n",
      "34793  c_000000000017  CG8421_sim-c_000000000017-x   \n",
      "34794  c_000000000018  CG8421_sim-c_000000000018-x   \n",
      "\n",
      "                                                sequence  \n",
      "34790  AAATTTTCTGCTAATTTCTTATTTAACTCCAAAACATCTTCATGAA...  \n",
      "34791  TCGTAGCTTTATTAGCATATTTTTTTAATAAAGGCAATTGTTTAGT...  \n",
      "34792  TCGTAGCTTTATTAGCATATTTTTTTAATAAAGGCAATTGTTTAGT...  \n",
      "34793  TGTTCTTCTAAACCTTTCTTTCAATTTTCTAGAATTTATTTTTTAG...  \n",
      "34794  CATCCTAGTTATAAAGAATTACTTTCTAAAAGAAAGGGTATAAGTG...  \n",
      "        gene_callers_id                                           sequence  \\\n",
      "147059             1716  MSHYYERFHEDVLELNKKLAENFKNSIVSHGNDPLDALQGIEQFVY...   \n",
      "147060             1717             VALLAYFFNKGNCLVRDGPVETMIAVSLIPFLLESNSL   \n",
      "147061             1718  MSHYYERFHEDVLELNKKLAENFKNSIVSHGNDPLDALQGIDILVI...   \n",
      "147062             1719  CSSKPFFQFSRIYFLELMSHYYERFHEDVLELNKKLAENFKNSIVS...   \n",
      "147063             1720  HPSYKELLSKRKGISDTAIIVSTGPSLTKQLPLLKKYANKATIFCA...   \n",
      "\n",
      "                contig  start  stop direction  partial    source version  \\\n",
      "147059  c_000000000015      1   286         r        1  prodigal  v2.6.3   \n",
      "147060  c_000000000016      2   119         f        1  prodigal  v2.6.3   \n",
      "147061  c_000000000016     84   252         r        1  prodigal  v2.6.3   \n",
      "147062  c_000000000017      0   174         f        1  prodigal  v2.6.3   \n",
      "147063  c_000000000018      0   165         f        1  prodigal  v2.6.3   \n",
      "\n",
      "       genome_name                         new_ids  \n",
      "147059  CG8421_sim  CG8421_sim-c_000000000015-1716  \n",
      "147060  CG8421_sim  CG8421_sim-c_000000000016-1717  \n",
      "147061  CG8421_sim  CG8421_sim-c_000000000016-1718  \n",
      "147062  CG8421_sim  CG8421_sim-c_000000000017-1719  \n",
      "147063  CG8421_sim  CG8421_sim-c_000000000018-1720  \n"
     ]
    }
   ],
   "source": [
    "prot_dir = \"/Volumes/KeithSSD/CA_V3/data/03_CNTG_DBs_final\"\n",
    "prot_paths = [os.path.join(prot_dir, i+\"_Proteins.txt\") for i in relapse_pair_isos]\n",
    "assert sum([os.path.exists(i) for i in prot_paths]) == len(prot_paths)\n",
    "pos_paths = [os.path.join(prot_dir, i+\"_GenePositions.txt\") for i in relapse_pair_isos]\n",
    "assert sum([os.path.exists(i) for i in pos_paths]) == len(pos_paths)\n",
    "contigs_dir = \"/Volumes/KeithSSD/CA_V3/data/02_ASSEMBLY_final\"\n",
    "assem_paths = [os.path.join(contigs_dir, i+\"_contigs_fixed.fa\") for i in relapse_pair_isos]\n",
    "assert sum([os.path.exists(i) for i in assem_paths])\n",
    "\n",
    "relapse_pair_isos.append('CG8421_sim')\n",
    "pos_paths.append('/Volumes/KeithSSD/variant-calling/simulated_reference/CG8421_GenePositions.txt')\n",
    "prot_paths.append('/Volumes/KeithSSD/variant-calling/simulated_reference/CG8421_Proteins.txt')\n",
    "assem_paths.append('/Volumes/KeithSSD/variant-calling/simulated_reference/CG8421_sim_contigs_fixed.fa')\n",
    "\n",
    "prot_dict = {}\n",
    "print(\"Reading protein and gene position files\")\n",
    "for samp, prot_f, pos_f in zip(relapse_pair_isos, prot_paths, pos_paths):\n",
    "    pos_df = pd.read_csv(pos_f, sep=\"\\t\", index_col=0)\n",
    "    prot_df = pd.read_csv(prot_f, sep=\"\\t\", index_col=0)\n",
    "    protpos_df = prot_df.join(pos_df, on=None, how='outer').reset_index()\n",
    "    assert len(pos_df) == len(prot_df) == len(protpos_df)\n",
    "    protpos_df['genome_name'] = pd.Series(index=protpos_df.index, \n",
    "                                       data=[samp]*len(protpos_df))\n",
    "    prot_dict[samp] = protpos_df.copy()\n",
    "\n",
    "def read_contig_file(contig_path, samp_name):\n",
    "    header_lines, contig_sequences = [], []\n",
    "    with open(contig_path, 'r') as fh:\n",
    "        for line in fh:\n",
    "            if line.startswith(\">\"):\n",
    "                if len(header_lines) > 0:\n",
    "                    contig_sequences.append(temp_sequence)\n",
    "                header_lines.append(line[1:].strip())\n",
    "                temp_sequence = \"\"\n",
    "            else:\n",
    "                temp_sequence += line.strip()\n",
    "    \n",
    "    contig_sequences.append(temp_sequence)\n",
    "    if len(header_lines) != len(contig_sequences):\n",
    "        print(len(header_lines), len(contig_sequences))\n",
    "    assert len(header_lines) == len(contig_sequences)\n",
    "    \n",
    "    genome_size = len(\"\".join(contig_sequences))\n",
    "    new_headers = [samp_name + \"-\" + j + \"-\" + 'x' for i, j in enumerate(header_lines)]\n",
    "    contig_df = pd.DataFrame(data=[header_lines, new_headers, contig_sequences], index=[\"old_ids\", 'new_ids', 'sequence']).T\n",
    "    return contig_df\n",
    "\n",
    "print(\"Reading assemblies\")\n",
    "contigs = {i:read_contig_file(j, i) for i, j in zip(relapse_pair_isos, assem_paths)}\n",
    "\n",
    "all_prot_df = pd.concat(prot_dict.values(), ignore_index=True, axis=0)\n",
    "hdr_fxn = lambda x: \"-\".join([str(i) for i in x])\n",
    "all_prot_df['new_ids'] = all_prot_df[['genome_name', 'contig', 'gene_callers_id']].apply(hdr_fxn, axis=1)\n",
    "print(\"Total # proteins read {}\".format(len(all_prot_df)))\n",
    "\n",
    "all_cntg_df = pd.concat(contigs.values(), ignore_index=True, axis=0)\n",
    "print(\"Total # contigs read {}\".format(len(all_cntg_df)))\n",
    "\n",
    "print(all_cntg_df.tail())\n",
    "print(all_prot_df.tail())\n",
    "\n",
    "gene_cluster_file = '../../CA_V3/data/07_PAN_GENOME_final/Trial_2/gene_clusters.txt'\n",
    "cluster_df = pd.read_csv(gene_cluster_file, sep=\"\\t\")\n",
    "\n",
    "cluster_df['mapper'] = cluster_df[['gene_caller_id', 'genome_name']].apply(tuple, axis=1)\n",
    "clust_map = {k:v for k, v in list(cluster_df[['mapper', 'gene_cluster_id' ]].values)}\n",
    "\n",
    "all_prot_df['mapper'] = cluster_df[['gene_caller_id', 'genome_name']].apply(tuple, axis=1)\n",
    "all_prot_df['gene_cluster_id'] = all_prot_df['mapper'].map(clust_map)\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### We will add the reference genome to these dataframes as a positive control and then we will print the correctly formatted commands to blast in the Terminal. We want to enforce 45% query coverage for both exact and full gene sequences. For the latter, it is because no matter where a frameshift occurs in a protein, at least 50% of one of the frames will still match the original. For the former, it is because the mutation is expected in the middle of the sequence and if the alignment decides to break, instead of pushing through the mismatch, a little under 50% of it should still be matched. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Contigs number upped from 34795 to 34796\n",
      "\n",
      " /Volumes/KeithSSD/variant-calling/variant_homologs/contig_pile.fa written\n",
      "\n",
      "OUTFMT=\"6 qseqid sseqid pident length qcovs mismatch gapopen qstart qend qlen sstart send slen evalue bitscore sstrand\"\n",
      "\n",
      "makeblastdb -in /Volumes/KeithSSD/variant-calling/variant_homologs/contig_pile.fa -dbtype nucl -out /Volumes/KeithSSD/variant-calling/variant_homologs/contig_pile_db\n",
      "\n",
      "tblastx -query_gencode 11 -db_gencode 11 -evalue 1e-3 -db /Volumes/KeithSSD/variant-calling/variant_homologs/contig_pile_db -query /Volumes/KeithSSD/variant-calling/variant_homologs/gene_primers.fa -outfmt \"$OUTFMT\" -qcov_hsp_perc 45 -out /Volumes/KeithSSD/variant-calling/variant_homologs/gene_hits.txt -num_threads 3 \n",
      "\n"
     ]
    }
   ],
   "source": [
    "src_index = list(gb_df[gb_df['type'] == 'source'].index)\n",
    "assert len(src_index) == 1\n",
    "header_pfx = \"-\".join(list(gb_df.loc[src_index[0], ['strain', 'molecule']]) + ['x'])\n",
    "ref_row = pd.DataFrame({all_cntg_df.shape[0]:{'old_ids':\"\", 'new_ids':header_pfx, 'sequence': full_genome}}).T\n",
    "all_cntg_df3 = all_cntg_df.append(ref_row, sort=True)\n",
    "all_cntg_df2 = all_cntg_df3[['new_ids', 'sequence']].dropna()\n",
    "print(\"Contigs number upped from {} to {}\".format(len(all_cntg_df), len(all_cntg_df2)))\n",
    "\n",
    "nt_ref_file = os.path.join(data_dir, 'contig_pile.fa')\n",
    "ntdb = nt_ref_file.replace(\".fa\", \"_db\")\n",
    "gene_out = os.path.join(data_dir, 'gene_hits.txt')\n",
    "table_to_fasta('new_ids', 'sequence', all_cntg_df2, nt_ref_file)\n",
    "\n",
    "print('\\nOUTFMT=\"6 qseqid sseqid pident length qcovs mismatch gapopen qstart qend qlen sstart send slen evalue bitscore sstrand\"\\n')\n",
    "print(\"makeblastdb -in {} -dbtype nucl -out {}\\n\".format(nt_ref_file, ntdb))\n",
    "print('tblastx -query_gencode 11 -db_gencode 11 -evalue 1e-3 -db {} -query {} -outfmt \"$OUTFMT\" -qcov_hsp_perc 45 -out {} -num_threads 3 \\n'.format(ntdb, gene_file, gene_out))\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### After running BLAST, we read back in the results. BLAST can be tricky because if you allow for very lenient alignment parameters, the program will not find optimal alignments, however if you err on the side of conservatism, you risk false negatives. With that in mind, we made sure to use BLAST command parameters and post-processing filtering that allowed the program to get a minimum of 1 perfect match. We then set an identity limit by cutting and pasting the primers into the BLAST web search tool and performing tblasx against the RefSeq Campylobacter coli genome. The hits and misses were obvious and so we set an addtional, such that all substandard alignments were removed. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Raw BLAST hits total 33823 w/ 23 columns\n",
      "\tAfter keeping only best pairs, we are left with 6690\n"
     ]
    }
   ],
   "source": [
    "from itertools import product\n",
    "\n",
    "def best_pairs(df):\n",
    "    primers_ = df['primer'].unique()\n",
    "    genomes_ = df['genome_name'].unique()\n",
    "    subdfs = [df[(df['primer'] == p) & (df['genome_name'] == g)] for p, g in product(primers_, genomes_)]\n",
    "    subdfs2 = [df[df.evalue == df.evalue.min()] for df in subdfs]\n",
    "    bestidxs = [i for df in subdfs2 for i in df[df.pident_adj == df.pident_adj.max()].index]\n",
    "    df2 = df.loc[bestidxs, :].copy()\n",
    "    df3 = df2[df2.pident > 80.0 ]\n",
    "    check_ = [df3[(df3['primer'] == p) & (df3['genome_name'] == 'CG8421')].shape[0] > 0 for p in primers_]\n",
    "    assert set(check_) == set([True])\n",
    "    return df3.copy()\n",
    "\n",
    "blast_cols = ['primer', 'sseqid', 'pident', 'length', 'qcov', 'mismatch', 'gapopen', 'qstart', 'qend',\n",
    "              'qlen', 'sstart', 'send', 'slen', 'evalue', 'bitscore', 'sstrand']\n",
    "new_cols = ['genome_name', 'contig', 'gene_callers_id', 'pident_adj', 'evalue']\n",
    "extra_cols = set(blast_cols[5:]) - set(new_cols)\n",
    "new_order = blast_cols[:5] + new_cols + list(extra_cols) + ['present', 'sstart2', 'send2']\n",
    "\n",
    "hit_df = pd.read_csv(gene_out, sep=\"\\t\", header=None)\n",
    "hit_df.columns = blast_cols\n",
    "hit_df['genome_name'] = hit_df['sseqid'].apply(lambda x: x.split(\"-\")[0])\n",
    "hit_df['contig'] = hit_df['sseqid'].apply(lambda x: x.split(\"-\")[1])\n",
    "hit_df['gene_callers_id'] = hit_df['sseqid'].apply(lambda x: x.split(\"-\")[2])\n",
    "hit_df['qcov'] = abs(hit_df['qend'] - hit_df['qstart']) / hit_df['qlen']\n",
    "hit_df['sstart2'] = hit_df[['sstart', 'send']].apply(lambda x: min(x), axis=1)\n",
    "hit_df['send2'] = hit_df[['sstart', 'send']].apply(lambda x: max(x), axis=1)\n",
    "hit_df['pident_adj'] = hit_df['pident']*hit_df['qcov']\n",
    "hit_df['present'] = pd.Series(index=hit_df.index, data=[1]*len(hit_df))\n",
    "print(\"Raw BLAST hits total {} w/ {} columns\".format(len(hit_df), hit_df.shape[1]))\n",
    "assert set(new_order) == set(hit_df.columns)\n",
    "blast_df = best_pairs(hit_df[new_order])\n",
    "blast_df['locus_tag'] = blast_df['primer'].apply(lambda x: x.split(\"-\")[0])\n",
    "blast_df['primer_position'] = blast_df['primer'].apply(lambda x: x.split(\"-\")[1])\n",
    "print(\"\\tAfter keeping only best pairs, we are left with {}\".format(blast_df.shape[0]))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Here we compile both types of BLAST searches and pivot them into a presence/absence table with primers in the columns and genomes in the rows. We will also make a similarly shaped table with all the location information for each hit."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total possible homologs = 6372, Total valid ones detected = 5707\n"
     ]
    }
   ],
   "source": [
    "crofts_genomes = list(filter(lambda x: x.startswith(\"crofts\"), blast_df['genome_name'].unique()))\n",
    "my_genomes = [i for i in set(blast_df['genome_name']) if not i.startswith(\"CG8421\") and not i in crofts_genomes]\n",
    "\n",
    "def fix_aggs(x):\n",
    "    x.index.name = None\n",
    "    x.columns = x.columns.droplevel()\n",
    "    x.columns.name = None\n",
    "    return x\n",
    "\n",
    "# Create a location to record the # of valid homologs per primer\n",
    "pre_homolog = blast_df[['genome_name', 'primer', 'present']].groupby(['genome_name', 'primer'])\n",
    "homolog_df = fix_aggs(pre_homolog.agg(np.sum).unstack(level=-1))\n",
    "\n",
    "# Create permanent + temporary locations to record the corresponding contig and location values \n",
    "loc_cols = ['sseqid', 'sstart2', 'send2', 'qstart', 'qend', 'pident']\n",
    "blast_df['loc_info'] = blast_df[loc_cols].apply(tuple, axis=1)\n",
    "pre_loc = blast_df[['genome_name', 'primer', 'loc_info']].groupby(['genome_name', 'primer'])\n",
    "location_df = fix_aggs(pre_loc.agg(list).unstack(level=-1))\n",
    "\n",
    "## Create a location for e-values and set permissive defaults \n",
    "pre_eval = blast_df[['genome_name', 'primer', 'evalue']].groupby(['genome_name', 'primer'])\n",
    "evalue_df = fix_aggs(pre_eval.agg(list).unstack(level=-1))\n",
    "\n",
    "searched_for = homolog_df.shape[0]*homolog_df.shape[1]\n",
    "detected_n = searched_for - homolog_df.isnull().sum().sum()\n",
    "print(\"Total possible homologs = {}, Total valid ones detected = {}\".format(searched_for, detected_n))\n",
    "\n",
    "coldOrder = ['CG8421', 'CG8421_sim'] + crofts_genomes + my_genomes\n",
    "assert set(coldOrder) - set(evalue_df.index) == set()\n",
    "\n",
    "homolog_group_order = []\n",
    "for aff_gene in anchor_df['locus_tag'].unique():\n",
    "    homolog_group_order += [aff_gene+'-before', aff_gene+'-gene', aff_gene+'-after']\n",
    "\n",
    "assert set(homolog_group_order) == set(homolog_df.columns)\n",
    "homolog_df.loc[coldOrder, homolog_group_order].to_csv(\"/Volumes/KeithSSD/variant-calling/misc_data/homologs.txt\", sep=\"\\t\")\n",
    "evalue_df.loc[coldOrder, homolog_group_order].to_csv(\"/Volumes/KeithSSD/variant-calling/misc_data/evalues.txt\", sep=\"\\t\")\n",
    "location_df.loc[coldOrder, homolog_group_order].to_csv(\"/Volumes/KeithSSD/variant-calling/misc_data/locations.txt\", sep=\"\\t\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Now that we know where the hits are, we need to integrate them with the variation dataframe file. We go through our BLAST hits by variant region and check that either the full gene or the exact region in which a variation was detected was on the same contig as at least one of the genes affected by that variation. Sometimes the variation is within the gene, and so this is a redundant check, but many times they are not. We also go through and check to see if BLAST was able to locate an exact copy of the wild-type allele (found in the reference genome) in the isolate assembly. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "def find_wt(gnsr, ra):\n",
    "    cntg_df, gnc = all_cntg_df3.copy(), list(gnsr)\n",
    "    if type(ra) != float:\n",
    "        cntg_match = cntg_df['new_ids'] == gnc[0]\n",
    "        if cntg_match.sum() == 1:\n",
    "            cntg_seq = cntg_df.loc[cntg_match, 'sequence'].values[0]\n",
    "            for sub_seq in ra:\n",
    "                if cntg_seq.find(sub_seq, gnc[1]-20, gnc[2]+20) != -1:\n",
    "                    return True\n",
    "            return False\n",
    "        else:\n",
    "            raise ValueError(\"what contig from where: {}\".format(gnc))\n",
    "    else:\n",
    "        return ra\n",
    "\n",
    "from Bio.Seq import Seq\n",
    "from Bio.Alphabet import generic_dna\n",
    "\n",
    "def seq_array(ra):\n",
    "    ra2 = Seq(ra, generic_dna)\n",
    "    ra_ls = [ra]\n",
    "    ra_ls.append(str(ra2.reverse_complement()))\n",
    "    ra_ls.append(str(ra2.complement()))\n",
    "    ra_ls.append(str(ra2.reverse_complement().complement()))\n",
    "    return ra_ls\n",
    "\n",
    "## Repeat the iteration from before \n",
    "data_prod_1 = []\n",
    "\n",
    "pop_cols = ['has_wildtype', 'has_homolog']\n",
    "for pop_colx in pop_cols:\n",
    "    anchor_df[pop_colx] = pd.Series(index=anchor_df.index) \n",
    "\n",
    "contigs_identified = set()\n",
    "loc_cols2 = ['sseqid', 'sstart2', 'send2']\n",
    "\n",
    "for ix in anchor_df.index:\n",
    "    gene_, _var_ = list(anchor_df.loc[ix, ['locus_tag', 'variation_index']])\n",
    "    if type(_var_) != float:\n",
    "        locii, annot_ = list(var_df.loc[_var_, ['var_locii', 'Annotation Used for this Study']])\n",
    "        start_cut = locii[0]-1-8\n",
    "        if len(locii) == 1:\n",
    "            end_cut = locii[0]+8\n",
    "        else:\n",
    "            end_cut = locii[1]+8\n",
    "        ref_set = seq_array(full_genome[start_cut:end_cut][:16])\n",
    "    else:\n",
    "        ref_set = np.nan\n",
    "        annot_ = ngb_df.loc[ngb_df['locus_tag'] == gene_, 'product'].values[0]\n",
    "        locii = (0,0)\n",
    "    \n",
    "    blast_filt = blast_df.copy()[blast_df['locus_tag'] == gene_]\n",
    "    \n",
    "    # pull out populations with the homologous region\n",
    "    reg_by_iso = blast_filt[['genome_name', 'primer_position']].groupby('genome_name').agg(lambda x: len(set(x)))\n",
    "    good_isos = list(reg_by_iso[reg_by_iso['primer_position'] > 1 ].index)\n",
    "    anchor_df.loc[ix, 'has_homolog'] = \"@#$\".join(good_isos)\n",
    "    \n",
    "    # search for the wild type allele \n",
    "    blast_filt['has_wildtype'] = blast_filt[loc_cols2].apply(lambda x: find_wt(x, ref_set), axis=1)\n",
    "    wt_havers = set(blast_filt.loc[blast_filt['has_wildtype'] == True, 'genome_name'])\n",
    "    anchor_df.loc[ix, 'has_wildtype'] = \"@#$\".join(list(wt_havers))\n",
    "\n",
    "    # aggregate all the contigs that were identified \n",
    "    contigs_identified.update(list(blast_filt['sseqid'].unique()))\n",
    "    print(\"{}\\tGene = {}\\t Samples w/it = {}\\t Samples w/WT seq = {}\\t{}\".format(annot_, gene_, len(good_isos), \n",
    "                                                                             len(wt_havers), locii))\n",
    "for pop_colx in pop_cols:\n",
    "    nn_x = anchor_df[pop_colx].notnull()\n",
    "    anchor_df.loc[nn_x, pop_colx] = anchor_df.loc[nn_x, pop_colx].apply(lambda x: x.split(\"@#$\"))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([[list(['CAMPY87_S44', 'crofts_24D2', 'crofts_31R', 'CAMPY15_S7', 'CAMPY12_S6', 'crofts_9R', 'pCAMPY22', 'crofts_32D2', 'CAMPY41_S20', 'crofts_4D1', 'crofts_11D1', 'CAMPY82_spr11', 'crofts_8R1', 'crofts_14D2', 'crofts_21D1', 'crofts_16D0', 'CG8421_sim', 'crofts_31D2', 'pCAMPY47', 'crofts_4R', 'CG8421', 'pCAMPY80', 'crofts_8D3', 'crofts_8D2', 'crofts_14D3', 'crofts_9D2', 'crofts_innoc', 'crofts_16D1', 'crofts_8R2', 'crofts_33D1', 'CAMPY4_spr11', 'crofts_1R', 'crofts_1D2']),\n",
       "        list(['CAMPY10', 'CAMPY12_S6', 'CAMPY15_S7', 'CAMPY28', 'CAMPY41_S20', 'CAMPY4_spr11', 'CAMPY82_spr11', 'CAMPY87_S44', 'CG8421', 'CG8421_sim', 'crofts_11D1', 'crofts_14D2', 'crofts_14D3', 'crofts_16D0', 'crofts_16D1', 'crofts_1D2', 'crofts_1R', 'crofts_21D1', 'crofts_24D2', 'crofts_31D2', 'crofts_31R', 'crofts_32D2', 'crofts_33D1', 'crofts_4D1', 'crofts_4R', 'crofts_8D2', 'crofts_8D3', 'crofts_8R1', 'crofts_8R2', 'crofts_9D2', 'crofts_9R', 'crofts_innoc', 'pCAMPY22', 'pCAMPY47', 'pCAMPY80'])],\n",
       "       [list(['CAMPY87_S44', 'crofts_24D2', 'crofts_31R', 'CAMPY15_S7', 'CAMPY12_S6', 'crofts_9R', 'pCAMPY22', 'crofts_32D2', 'CAMPY41_S20', 'crofts_4D1', 'crofts_11D1', 'CAMPY82_spr11', 'crofts_8R1', 'crofts_14D2', 'crofts_21D1', 'crofts_16D0', 'CG8421_sim', 'crofts_31D2', 'pCAMPY47', 'crofts_4R', 'CG8421', 'pCAMPY80', 'crofts_8D3', 'crofts_8D2', 'crofts_14D3', 'crofts_9D2', 'crofts_innoc', 'crofts_16D1', 'crofts_8R2', 'crofts_33D1', 'CAMPY4_spr11', 'crofts_1R', 'crofts_1D2']),\n",
       "        list(['CAMPY10', 'CAMPY12_S6', 'CAMPY15_S7', 'CAMPY28', 'CAMPY41_S20', 'CAMPY4_spr11', 'CAMPY82_spr11', 'CAMPY87_S44', 'CG8421', 'CG8421_sim', 'crofts_11D1', 'crofts_14D2', 'crofts_14D3', 'crofts_16D0', 'crofts_16D1', 'crofts_1D2', 'crofts_1R', 'crofts_21D1', 'crofts_24D2', 'crofts_31D2', 'crofts_31R', 'crofts_32D2', 'crofts_33D1', 'crofts_4D1', 'crofts_4R', 'crofts_8D2', 'crofts_8D3', 'crofts_8R1', 'crofts_8R2', 'crofts_9D2', 'crofts_9R', 'crofts_innoc', 'pCAMPY22', 'pCAMPY47', 'pCAMPY80'])]],\n",
       "      dtype=object)"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#gene_ = 'CJ8421_RS06660'\n",
    "#blast_filt = blast_df[blast_df['locus_tag'] == gene_]\n",
    "#blast_filt.sample(1).T\n",
    "anchor_df[pop_cols].head(2).values"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Here we print a summary table where we look at regions of variation how often each region of variation is found in in primary and relapse isolates from both groups, as well as in the complete and reassembled versions of the reference. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "22 15 6 3\n",
      "(85, 22) (85, 25) (85, 47)\n"
     ]
    }
   ],
   "source": [
    "innoc = 'crofts_innoc'\n",
    "ref_complete = 'CG8421'\n",
    "ref_sim = 'CG8421_sim'\n",
    "single_isos = ['Ref_Full', 'Ref_Sim', 'Crofts_Innoc'] \n",
    "single_keys = [ref_complete, ref_sim, innoc]\n",
    "\n",
    "relapse_crofts = list(metadata_df[metadata_df['primary_v_relapse_crofts'] == 'Relapse'].index)\n",
    "primary_crofts = [i for i in crofts_genomes if not i in single_keys and not i in relapse_crofts]\n",
    "print(len(crofts_genomes), len(primary_crofts), len(relapse_crofts), len(single_keys))\n",
    "\n",
    "peru_primary = list(metadata_df[metadata_df['primary_v_relapse_nocrofts'] == 'Primary'].index)\n",
    "peru_relapse = list(metadata_df[metadata_df['primary_v_relapse_nocrofts'] == 'Relapse'].index)\n",
    "\n",
    "group_names = ['Crofts_Primary', 'Crofts_Relapse', 'Iquitos_Primary', 'Iquitos_Relapse']\n",
    "group_keys = [primary_crofts, relapse_crofts, peru_primary, peru_relapse]\n",
    "\n",
    "procd_homologs = {}\n",
    "for anch_idx in anchor_df.index:\n",
    "    wt_isos, homolog_isos = anchor_df.loc[anch_idx, pop_cols]\n",
    "    assert type(wt_isos) == list == type(homolog_isos)\n",
    "    wt_isos, homolog_isos = set(wt_isos), set(homolog_isos)\n",
    "    rec_rows = {}\n",
    "    for g_n, g_k in zip(group_names, group_keys):\n",
    "        g_ks = set(g_k)\n",
    "        rec_rows[\"_\".join([g_n, 'w_Wildtype'])] = len(wt_isos.intersection(g_ks))\n",
    "        rec_rows[\"_\".join([g_n, 'wo_Wildtype'])] = len(g_ks) - len(wt_isos.intersection(g_ks))\n",
    "        rec_rows[\"_\".join([g_n, 'w_Region'])] = len(homolog_isos.intersection(g_ks))\n",
    "        rec_rows[\"_\".join([g_n, 'wo_Region'])] = len(g_ks) - len(homolog_isos.intersection(g_ks))\n",
    "        \n",
    "    for s_i, s_k in zip(single_isos, single_keys):\n",
    "        rec_rows[\"_\".join([s_i, 'has_Wildtype'])] = s_k in wt_isos\n",
    "        rec_rows[\"_\".join([s_i, 'Region_Count'])] = s_k in homolog_isos\n",
    "\n",
    "    procd_homologs[anch_idx] = rec_rows\n",
    "\n",
    "# Need to map the descriptive information from the original dataframe to here \n",
    "flipped_df = pd.DataFrame(procd_homologs).T\n",
    "e_anchor_df = pd.concat([anchor_df, flipped_df], axis=1)\n",
    "print(flipped_df.shape, anchor_df.shape, e_anchor_df.shape)\n",
    "\n",
    "flipped_df.to_csv('/Volumes/KeithSSD/variant-calling/misc_data/all_iso_all_region_processed.txt', sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Next we decide which of the two assemblies (primary or relapse) is more complete and then output a pair of columns in a text file of which assembly to use as a reference and which to library to map to it, in order to call variations. For the Croft's genomes, we will use the simulated and complete reference genomes, in order to reproduce their results.  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "966 contigs hit across all genomes\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>reference_assembly</th>\n",
       "      <th>mapping_query</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CG8421</td>\n",
       "      <td>crofts_9D2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CG8421</td>\n",
       "      <td>crofts_4D1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CG8421</td>\n",
       "      <td>crofts_31D2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CG8421</td>\n",
       "      <td>crofts_11D1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CG8421</td>\n",
       "      <td>crofts_9R</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>CG8421</td>\n",
       "      <td>crofts_14D2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>CG8421</td>\n",
       "      <td>crofts_8D2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>CG8421</td>\n",
       "      <td>crofts_innoc</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>CG8421</td>\n",
       "      <td>crofts_4R</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>CG8421</td>\n",
       "      <td>crofts_16D0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>CG8421</td>\n",
       "      <td>crofts_1D2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>CG8421</td>\n",
       "      <td>crofts_8D3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>CG8421</td>\n",
       "      <td>crofts_14D3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>CG8421</td>\n",
       "      <td>crofts_24D2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>CG8421</td>\n",
       "      <td>crofts_8R2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>CG8421</td>\n",
       "      <td>crofts_16D1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>CG8421</td>\n",
       "      <td>crofts_32D2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>CG8421</td>\n",
       "      <td>crofts_31R</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>CG8421</td>\n",
       "      <td>crofts_21D1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>CG8421</td>\n",
       "      <td>crofts_33D1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>CG8421</td>\n",
       "      <td>crofts_1R</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>CG8421</td>\n",
       "      <td>crofts_8R1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>CAMPY28</td>\n",
       "      <td>CAMPY10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>pCAMPY80</td>\n",
       "      <td>pCAMPY22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>CAMPY4_spr11</td>\n",
       "      <td>CAMPY15_S7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>CAMPY87_S44</td>\n",
       "      <td>CAMPY82_spr11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>CAMPY41_S20</td>\n",
       "      <td>pCAMPY47</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>CAMPY12_S6</td>\n",
       "      <td>pCAMPY86</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   reference_assembly  mapping_query\n",
       "0              CG8421     crofts_9D2\n",
       "1              CG8421     crofts_4D1\n",
       "2              CG8421    crofts_31D2\n",
       "3              CG8421    crofts_11D1\n",
       "4              CG8421      crofts_9R\n",
       "5              CG8421    crofts_14D2\n",
       "6              CG8421     crofts_8D2\n",
       "7              CG8421   crofts_innoc\n",
       "8              CG8421      crofts_4R\n",
       "9              CG8421    crofts_16D0\n",
       "10             CG8421     crofts_1D2\n",
       "11             CG8421     crofts_8D3\n",
       "12             CG8421    crofts_14D3\n",
       "13             CG8421    crofts_24D2\n",
       "14             CG8421     crofts_8R2\n",
       "15             CG8421    crofts_16D1\n",
       "16             CG8421    crofts_32D2\n",
       "17             CG8421     crofts_31R\n",
       "18             CG8421    crofts_21D1\n",
       "19             CG8421    crofts_33D1\n",
       "20             CG8421      crofts_1R\n",
       "21             CG8421     crofts_8R1\n",
       "22            CAMPY28        CAMPY10\n",
       "23           pCAMPY80       pCAMPY22\n",
       "24       CAMPY4_spr11     CAMPY15_S7\n",
       "25        CAMPY87_S44  CAMPY82_spr11\n",
       "26        CAMPY41_S20       pCAMPY47\n",
       "27         CAMPY12_S6       pCAMPY86"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# find Iquitos relapse pairings from somewhere \n",
    "relpase_pair_file = '/Volumes/KeithSSD/CA_V3/data/09_Phylogenomics/relapse_dist_pairs_120719.txt'\n",
    "iquitos_pairings = pd.read_csv(relpase_pair_file, sep=\"\\t\", index_col=0)[['Pair Member 1', 'Pair Member 2']]\n",
    "all_genomes = set(blast_df['genome_name'])\n",
    "ip_links = {i:j for i, j in iquitos_pairings.values \n",
    "                if (i in all_genomes) and (j in all_genomes) and (not i in crofts_genomes)}\n",
    "\n",
    "# read in each desired contig set from each member of a pair\n",
    "print(\"{} contigs hit across all genomes\".format(len(contigs_identified)))\n",
    "assert len(set(contigs_identified) - set(all_cntg_df2['new_ids'])) == 0\n",
    "\n",
    "all_cntg_df2['source_genome'] = all_cntg_df2['new_ids'].apply(lambda x: x.split(\"-\")[0])\n",
    "all_cntg_df4 = all_cntg_df2.copy()[all_cntg_df2['new_ids'].isin(contigs_identified)]\n",
    "all_cntg_df4['seqlen'] = all_cntg_df4['sequence'].apply(len)\n",
    "\n",
    "contig_sum = all_cntg_df4[['source_genome', 'seqlen']].groupby('source_genome').agg(np.sum)['seqlen']\n",
    "\n",
    "assem_lib_pairs = []\n",
    "for cg in crofts_genomes:\n",
    "    assem_lib_pairs.append(('CG8421', cg))\n",
    "\n",
    "# whichever has a greater number of hits write out as the \"backbone\"\n",
    "peru_isos = set(peru_primary + peru_relapse)\n",
    "for k, v in ip_links.items():\n",
    "    peru_isos.remove(k); peru_isos.remove(v);\n",
    "    ixs = np.array([k, v])\n",
    "    anylz = np.array([contig_sum[k], contig_sum[v]])\n",
    "    assem_lib_pairs.append((ixs[anylz.argmax()], ixs[anylz.argmin()]))\n",
    "\n",
    "# create a table of the name of the backbone and the name of the reads to map against it\n",
    "mapping_pairs = pd.DataFrame(index=range(len(assem_lib_pairs)), data=assem_lib_pairs,\n",
    "                             columns=['reference_assembly', 'mapping_query'])\n",
    "\n",
    "# write a `snipper` file to do the mapping but alter it to find my contigs too\n",
    "mapping_pairs.to_csv(\"/Volumes/KeithSSD/variant-calling/misc_data/ref_map_pairs.txt\", sep=\"\\t\", \n",
    "                     header=None, index=None)\n",
    "mapping_pairs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Writing 1 recs for CG8421 to /Volumes/KeithSSD/variant-calling/ref_fastas/CG8421_bb.fa\n",
      "\n",
      " /Volumes/KeithSSD/variant-calling/ref_fastas/CG8421_bb.fa written\n",
      "Writing 29 recs for CAMPY28 to /Volumes/KeithSSD/variant-calling/ref_fastas/CAMPY28_bb.fa\n",
      "\n",
      " /Volumes/KeithSSD/variant-calling/ref_fastas/CAMPY28_bb.fa written\n",
      "Writing 48 recs for pCAMPY80 to /Volumes/KeithSSD/variant-calling/ref_fastas/pCAMPY80_bb.fa\n",
      "\n",
      " /Volumes/KeithSSD/variant-calling/ref_fastas/pCAMPY80_bb.fa written\n",
      "Writing 49 recs for CAMPY4_spr11 to /Volumes/KeithSSD/variant-calling/ref_fastas/CAMPY4_spr11_bb.fa\n",
      "\n",
      " /Volumes/KeithSSD/variant-calling/ref_fastas/CAMPY4_spr11_bb.fa written\n",
      "Writing 46 recs for CAMPY87_S44 to /Volumes/KeithSSD/variant-calling/ref_fastas/CAMPY87_S44_bb.fa\n",
      "\n",
      " /Volumes/KeithSSD/variant-calling/ref_fastas/CAMPY87_S44_bb.fa written\n",
      "Writing 45 recs for CAMPY41_S20 to /Volumes/KeithSSD/variant-calling/ref_fastas/CAMPY41_S20_bb.fa\n",
      "\n",
      " /Volumes/KeithSSD/variant-calling/ref_fastas/CAMPY41_S20_bb.fa written\n",
      "Writing 54 recs for CAMPY12_S6 to /Volumes/KeithSSD/variant-calling/ref_fastas/CAMPY12_S6_bb.fa\n",
      "\n",
      " /Volumes/KeithSSD/variant-calling/ref_fastas/CAMPY12_S6_bb.fa written\n"
     ]
    }
   ],
   "source": [
    "# write contig sets from reference genomes into individual files in a directory \n",
    "ref_dir2 = \"/Volumes/KeithSSD/variant-calling/ref_fastas\"\n",
    "if not os.path.exists(ref_dir2):\n",
    "    os.mkdir(ref_dir2)\n",
    "\n",
    "ref_set = list(mapping_pairs['reference_assembly'].unique())\n",
    "out_paths = [os.path.join(ref_dir2, sn+\"_bb.fa\") for sn in ref_set]\n",
    "sub_dfs = [all_cntg_df4.copy()[all_cntg_df4['source_genome'] == sn] for sn in ref_set]\n",
    "for sn, of, sdf in zip(ref_set, out_paths, sub_dfs):\n",
    "    print(\"Writing {} recs for {} to {}\".format(len(sdf), sn, of))\n",
    "    table_to_fasta('new_ids', 'sequence', sdf, of)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Once we've run prodigal to convert the FASTA files to genbank files and then Snippy to call variations between the reference backbone files and the libraries. We can proceed to read back the variants that were discovered."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "62 polymorphism reports detected\n",
      "Peru libs have a mean=11704.17, std=3952.42 variations v. CG8421 ref (n=6)\n",
      "But a mean=6.50, std=11.33 variations v. their own assems (n=12)\n",
      "Crofts libs have a mean=19.55, std=5.10 variations v. CG8421 ref (n=22)\n",
      "and a mean=18.91, std=5.10 variations v. CG8421 assem (n=22)\n",
      "56 SNP reports retained\n"
     ]
    }
   ],
   "source": [
    "# read in the SNP tables\n",
    "snp_result_dir = \"/Volumes/KeithSSD/variant-calling/snipout3_results\"\n",
    "snp_files = os.listdir(snp_result_dir)\n",
    "snp_paths = [os.path.join(snp_result_dir, f) for f in snp_files if f.endswith(\".2.tab\")]\n",
    "print(\"{} polymorphism reports detected\".format(len(snp_paths)))\n",
    "mp2 = [os.path.basename(i).replace(\".2.tab\", \"\").split(\"-\") + [i] for i in snp_paths]\n",
    "mapping_pairs2 = pd.DataFrame(data=mp2, columns=['ix', 'ref_assem', 'mapped_lib', 'result']).drop(['ix'], axis=1)\n",
    "results_df = {i:pd.read_csv(mapping_pairs2.loc[i, 'result'], sep=\"\\t\") for i in mapping_pairs2.index}\n",
    "\n",
    "bad_maps = (mapping_pairs2['ref_assem'] == 'CG8421') & (~mapping_pairs2['mapped_lib'].str.contains(\"crofts\"))\n",
    "bad_map_indexes = list(mapping_pairs2[bad_maps].index)\n",
    "bad_map_variants = np.array([results_df[i].shape[0] for i in bad_map_indexes])\n",
    "\n",
    "better_maps = (mapping_pairs2['ref_assem'] != 'CG8421') & (~mapping_pairs2['mapped_lib'].str.contains(\"crofts\"))\n",
    "better_map_indexes = list(mapping_pairs2[better_maps].index)\n",
    "better_map_variants = np.array([results_df[i].shape[0] for i in better_map_indexes])\n",
    "\n",
    "good_maps = (mapping_pairs2['ref_assem'] == 'CG8421') & (mapping_pairs2['mapped_lib'].str.contains(\"crofts\"))\n",
    "good_map_indexes = list(mapping_pairs2[good_maps].index)\n",
    "good_map_variants = np.array([results_df[i].shape[0] for i in good_map_indexes])\n",
    "\n",
    "best_maps = (mapping_pairs2['ref_assem'] == 'CG8421_sim') & (mapping_pairs2['mapped_lib'].str.contains(\"crofts\"))\n",
    "best_map_indexes = list(mapping_pairs2[best_maps].index)\n",
    "best_map_variants = np.array([results_df[i].shape[0] for i in best_map_indexes])\n",
    "\n",
    "print(\"Peru libs have a mean={:.2f}, std={:.2f} variations v. CG8421 ref (n={})\".format(bad_map_variants.mean(),\n",
    "                                                                                        bad_map_variants.std(),\n",
    "                                                                                        bad_map_variants.shape[0]))\n",
    "\n",
    "print(\"But a mean={:.2f}, std={:.2f} variations v. their own assems (n={})\".format(better_map_variants.mean(),\n",
    "                                                                                   better_map_variants.std(),\n",
    "                                                                                   better_map_variants.shape[0]))\n",
    "\n",
    "print(\"Crofts libs have a mean={:.2f}, std={:.2f} variations v. CG8421 ref (n={})\".format(good_map_variants.mean(),\n",
    "                                                                                          good_map_variants.std(),\n",
    "                                                                                          good_map_variants.shape[0]))\n",
    "\n",
    "print(\"and a mean={:.2f}, std={:.2f} variations v. CG8421 assem (n={})\".format(best_map_variants.mean(),\n",
    "                                                                               best_map_variants.std(),\n",
    "                                                                               best_map_variants.shape[0]))\n",
    "\n",
    "results_df = {k:v for k, v in results_df.items() if not k in bad_map_indexes}\n",
    "print(\"{} SNP reports retained\".format(len(results_df)))\n",
    "#assert set(mapping_pairs['reference_assembly']) == set(mapping_pairs2['ref_assem'])\n",
    "#assert set(mapping_pairs['mapping_query']) == set(mapping_pairs2['mapped_lib'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['Insertion' 'Deletion' 'SNV' nan]\n",
      "{'complex', 'del', 'snp', 'mnp', 'ins'}\n",
      "['ins' 'del' 'snp' nan]\n"
     ]
    }
   ],
   "source": [
    "print(e_anchor_df['Variant Type'].unique())\n",
    "print(set([i for v in results_df.values() for i in v['TYPE'].unique()]))\n",
    "\n",
    "print(e_anchor_df['var_type'].unique())\n",
    "#list(results_df.values())[0]\n",
    "#e_anchor_df.iloc[0, :].T"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GCATCATCCTCCTTTATTT GCATCATCCTCCTTTATTT:34 G:2\n",
      "A A:148 AG:10\n",
      "CTGAAAG CTGAAAG:42 ATAAAAA:0\n",
      "CT CT:40 AA:0\n",
      "AATAGA AATAGA:6 GTTATG:0\n",
      "T T:100 TC:38\n",
      "TACTAGTTA TACTAGTTA:72 GAATACTTG:4\n",
      "C C:106 T:4\n",
      "T T:116 C:4\n",
      "AAAGCTGGAGTG AAAGCTGGAGTG:82 GAAAGTAGGAGTA:4\n",
      "TT TT:164 AA:12\n",
      "A A:156 AAG:0\n",
      "A A:148 AC:14\n",
      "A A:162 AG:12\n",
      "T T:172 TA:0\n",
      "CG CG:122 C:52\n",
      "G G:112 A:0\n",
      "AC AC:160 A:112\n",
      "AC AC:154 A:40\n",
      "A A:146 AC:14\n",
      "A A:142 AC:12\n",
      "A A:90 T:0\n",
      "G G:70 C:0\n",
      "GA GA:34 AT:0\n",
      "CCT CCT:32 C:0\n",
      "A A:10 T:2\n",
      "GAA GAA:10 AT:0\n",
      "TACTAGTTA TACTAGTTA:72 GAATACTTG:4\n",
      "C C:106 T:4\n",
      "T T:116 C:4\n",
      "AAAGCTGGAGTG AAAGCTGGAGTG:82 GAAAGTAGGAGTA:4\n",
      "TT TT:164 AA:12\n",
      "A A:156 AAG:0\n",
      "A A:148 AC:14\n",
      "A A:162 AG:12\n",
      "T T:172 TA:0\n",
      "CG CG:122 C:52\n",
      "A A:146 AG:10\n",
      "CTGAAAG CTGAAAG:42 ATAAAAA:0\n",
      "CT CT:40 AA:0\n",
      "AATAGA AATAGA:6 GTTATG:0\n",
      "T T:146 TC:42\n",
      "TTC TTC:10 AT:0\n",
      "T T:10 A:2\n",
      "AGGAAAATCATAACAATAAATT AGGAAAATCATAACAATAAATT:32 GAAAAATATAAGAATAAATA:0\n",
      "CGATGAAATAAAGGAGGAT CGATGAAATAAAGGAGGAT:34 C:0\n",
      "AG AG:154 A:48\n",
      "TG TG:164 T:122\n",
      "C C:112 T:0\n",
      "G G:8 T:0\n",
      "C C:132 G:0\n",
      "T T:164 TG:16\n",
      "T T:156 TG:14\n",
      "TCC TCC:88 T:28\n",
      "GCATCATCCTCCTTTATTT GCATCATCCTCCTTTATTT:34 G:2\n",
      "C C:158 CG:88\n",
      "T T:6 G:0\n",
      "A A:150 AG:10\n",
      "A A:158 AG:144\n",
      "CTGAAAG CTGAAAG:38 ATAAAAA:0\n",
      "CT CT:36 AA:0\n",
      "TACTAGTTA TACTAGTTA:72 GAATACTTG:4\n",
      "C C:106 T:4\n",
      "T T:114 C:4\n",
      "AAAGCTGGAGTG AAAGCTGGAGTG:78 GAAAGTAGGAGTA:4\n",
      "TT TT:158 AA:14\n",
      "A A:152 AAG:0\n",
      "A A:140 AC:44\n",
      "A A:160 AG:70\n",
      "T T:176 TA:0\n",
      "CGG CGG:88 C:20\n",
      "CG CG:110 C:66\n",
      "AG AG:124 A:84\n",
      "G G:112 A:0\n",
      "AC AC:156 A:88\n",
      "A A:90 T:0\n",
      "G G:70 C:0\n",
      "GA GA:36 AT:0\n",
      "CCT CCT:34 C:0\n",
      "A A:14 T:0\n",
      "GAA GAA:10 AT:0\n",
      "TACTAGTTA TACTAGTTA:72 GAATACTTG:4\n",
      "C C:106 T:4\n",
      "T T:114 C:4\n",
      "AAAGCTGGAGTG AAAGCTGGAGTG:78 GAAAGTAGGAGTA:4\n",
      "TT TT:158 AA:14\n",
      "A A:152 AAG:0\n",
      "A A:158 AC:48\n",
      "A A:160 AG:70\n",
      "T T:180 TA:0\n",
      "CGG CGG:88 C:20\n",
      "CG CG:110 C:66\n",
      "AG AG:124 A:84\n",
      "A A:150 AG:10\n",
      "A A:158 AG:144\n",
      "CTGAAAG CTGAAAG:38 ATAAAAA:0\n",
      "CT CT:36 AA:0\n",
      "TTC TTC:10 AT:0\n",
      "T T:14 A:0\n",
      "AGGAAAATCATAACAATAAATT AGGAAAATCATAACAATAAATT:34 GAAAAATATAAGAATAAATA:0\n",
      "T T:6 G:0\n",
      "CGATGAAATAAAGGAGGAT CGATGAAATAAAGGAGGAT:34 C:0\n",
      "TGG TGG:78 T:26\n",
      "TG TG:154 T:84\n",
      "C C:112 T:0\n",
      "G G:6 T:0\n",
      "C C:120 G:0\n",
      "A A:164 AC:86\n",
      "TG TG:138 T:128\n",
      "A A:80 AG:26\n",
      "TG TG:90 T:64\n",
      "CTGAAAG CTGAAAG:26 ATAAAAA:0\n",
      "CT CT:24 AA:0\n",
      "TA TA:8 T:0\n",
      "A A:8 G:0\n",
      "A A:8 T:0\n",
      "A A:10 G:0\n",
      "AAAGCTGGAGTG AAAGCTGGAGTG:40 GAAAGTAGGAGTA:24\n",
      "TT TT:112 AA:14\n",
      "A A:106 AAG:0\n",
      "A A:88 AC:62\n",
      "T T:32 TA:0\n",
      "G G:66 A:0\n",
      "A A:68 T:0\n",
      "G G:46 C:0\n",
      "GATTTTCCT GATTTTCCT:6 ATTTTTC:0\n",
      "GAA GAA:8 AT:0\n",
      "AAAGCTGGAGTG AAAGCTGGAGTG:40 GAAAGTAGGAGTA:24\n",
      "TT TT:112 AA:14\n",
      "A A:106 AAG:0\n",
      "A A:88 AC:62\n",
      "T T:32 TA:0\n",
      "A A:80 AG:26\n",
      "TG TG:90 T:64\n",
      "CTGAAAG CTGAAAG:26 ATAAAAA:0\n",
      "CT CT:24 AA:0\n",
      "TA TA:8 T:0\n",
      "A A:8 G:0\n",
      "A A:8 T:0\n",
      "A A:10 G:0\n",
      "T T:74 TC:58\n",
      "G G:16 A:8\n",
      "CATAACAATAAATT CATAACAATAAATT:22 ATAAGAATAAATA:0\n",
      "TG TG:36 T:24\n",
      "C C:66 T:0\n",
      "C C:110 G:0\n",
      "T T:82 TG:66\n",
      "GCATCATCCTCCTTTATTT GCATCATCCTCCTTTATTT:36 G:2\n",
      "A A:148 AG:32\n",
      "A A:162 AG:150\n",
      "CTGAAAG CTGAAAG:40 ATAAAAA:0\n",
      "CT CT:38 AA:0\n",
      "TACTAGTTA TACTAGTTA:72 GAATACTTG:4\n",
      "C C:106 T:4\n",
      "T T:114 C:4\n",
      "AAAGCTGGAGTG AAAGCTGGAGTG:80 GAAAGTAGGAGTA:4\n",
      "TT TT:160 AA:16\n",
      "A A:154 AAG:0\n",
      "T T:176 TA:0\n",
      "C C:170 CG:50\n",
      "CG CG:130 C:44\n",
      "G G:112 A:0\n",
      "A A:6 T:0\n",
      "A A:168 AC:148\n",
      "A A:170 AC:114\n",
      "A A:146 AC:14\n",
      "A A:92 T:0\n",
      "G G:72 C:0\n",
      "GA GA:34 AT:0\n",
      "CCT CCT:32 C:0\n",
      "A A:10 T:0\n",
      "GAA GAA:8 AT:0\n",
      "TACTAGTTA TACTAGTTA:72 GAATACTTG:4\n",
      "C C:106 T:4\n",
      "T T:114 C:4\n",
      "AAAGCTGGAGTG AAAGCTGGAGTG:80 GAAAGTAGGAGTA:4\n",
      "TT TT:160 AA:16\n",
      "A A:154 AAG:0\n",
      "T T:176 TA:0\n",
      "C C:170 CG:50\n",
      "CG CG:130 C:44\n",
      "A A:148 AG:32\n",
      "A A:162 AG:150\n",
      "CTGAAAG CTGAAAG:40 ATAAAAA:0\n",
      "CT CT:38 AA:0\n",
      "T T:126 TC:86\n",
      "TTC TTC:8 AT:0\n",
      "T T:10 A:0\n",
      "AGGAAAATCATAACAATAAATT AGGAAAATCATAACAATAAATT:32 GAAAAATATAAGAATAAATA:0\n",
      "CGATGAAATAAAGGAGGAT CGATGAAATAAAGGAGGAT:36 C:0\n",
      "A A:168 AG:116\n",
      "T T:168 TG:148\n",
      "T T:6 A:2\n",
      "C C:112 T:0\n",
      "G G:6 T:0\n",
      "C C:124 G:0\n",
      "T T:166 TG:18\n",
      "A A:84 AG:44\n",
      "CTGAAAG CTGAAAG:26 ATAAAAA:0\n",
      "CT CT:24 AA:0\n",
      "T T:64 TC:48\n",
      "AAAGCTGGAGTG AAAGCTGGAGTG:44 GAAAGTAGGAGTA:26\n",
      "TT TT:116 AA:0\n",
      "A A:108 AAG:0\n",
      "A A:80 AC:68\n",
      "T T:34 TA:0\n",
      "G G:68 A:0\n",
      "A A:70 T:0\n",
      "G G:44 C:0\n",
      "GATTTTCCT GATTTTCCT:6 ATTTTTC:0\n",
      "GAA GAA:8 AT:0\n",
      "AAAGCTGGAGTG AAAGCTGGAGTG:44 GAAAGTAGGAGTA:26\n",
      "TT TT:116 AA:0\n",
      "A A:108 AAG:0\n",
      "A A:80 AC:68\n",
      "T T:34 TA:0\n",
      "A A:84 AG:44\n",
      "CTGAAAG CTGAAAG:26 ATAAAAA:0\n",
      "CT CT:24 AA:0\n",
      "T T:78 TC:52\n",
      "TTC TTC:8 AT:0\n",
      "GAAAATCATAACAATAAATT GAAAATCATAACAATAAATT:16 AAAAATATAAGAATAAATA:0\n",
      "TG TG:48 T:34\n",
      "C C:68 T:0\n",
      "C C:130 G:0\n",
      "T T:32 C:24\n",
      "T T:100 TG:90\n",
      "GCATCATCCTCCTTTATTT GCATCATCCTCCTTTATTT:34 G:2\n",
      "CGG CGG:120 C:18\n",
      "A A:144 AT:0\n",
      "CTGAAAG CTGAAAG:38 ATAAAAA:0\n",
      "CT CT:36 AA:0\n",
      "TACTAGTTA TACTAGTTA:70 GAATACTTG:4\n",
      "C C:104 T:4\n",
      "T T:114 C:4\n",
      "AAAGCTGGAGTG AAAGCTGGAGTG:80 GAAAGTAGGAGTA:4\n",
      "TT TT:158 AA:12\n",
      "A A:152 AAG:0\n",
      "A A:142 AC:22\n",
      "T T:174 TA:0\n",
      "C C:156 CG:18\n",
      "AG AG:112 A:104\n",
      "G G:112 A:0\n",
      "A A:146 AC:14\n",
      "A A:98 T:0\n",
      "G G:76 C:0\n",
      "GA GA:36 AT:0\n",
      "CCT CCT:32 C:0\n",
      "A A:12 T:0\n",
      "GAA GAA:10 AT:0\n",
      "TACTAGTTA TACTAGTTA:70 GAATACTTG:4\n",
      "C C:104 T:4\n",
      "T T:114 C:4\n",
      "AAAGCTGGAGTG AAAGCTGGAGTG:80 GAAAGTAGGAGTA:4\n",
      "TT TT:158 AA:12\n",
      "A A:152 AAG:0\n",
      "A A:142 AC:22\n",
      "T T:172 TA:0\n",
      "C C:156 CG:18\n",
      "AG AG:112 A:104\n",
      "CTGAAAG CTGAAAG:38 ATAAAAA:0\n",
      "CT CT:36 AA:0\n",
      "TTC TTC:10 AT:0\n",
      "T T:12 A:0\n",
      "AGGAAAATCATAACAATAAATT AGGAAAATCATAACAATAAATT:32 GAAAAATATAAGAATAAATA:0\n",
      "A A:144 AT:0\n",
      "CGATGAAATAAAGGAGGAT CGATGAAATAAAGGAGGAT:34 C:0\n",
      "TG TG:120 T:82\n",
      "C C:112 T:0\n",
      "G G:6 T:0\n",
      "C C:124 G:0\n",
      "ACC ACC:136 A:20\n",
      "T T:166 TG:14\n",
      "A A:78 AG:52\n",
      "CTGAAAG CTGAAAG:26 ATAAAAA:0\n",
      "CT CT:24 AA:0\n",
      "T T:28 TC:18\n",
      "AAAGCTGGAGTG AAAGCTGGAGTG:42 GAAAGTAGGAGTA:32\n",
      "TT TT:114 AA:14\n",
      "A A:106 AAG:0\n",
      "T T:32 TA:2\n",
      "G G:68 A:0\n",
      "A A:66 T:0\n",
      "G G:44 C:0\n",
      "TG TG:22 T:0\n",
      "C C:16 T:10\n",
      "CCT CCT:6 C:0\n",
      "GAA GAA:8 AT:0\n",
      "AAAGCTGGAGTG AAAGCTGGAGTG:42 GAAAGTAGGAGTA:32\n",
      "TT TT:114 AA:14\n",
      "A A:106 AAG:0\n",
      "T T:32 TA:2\n",
      "A A:78 AG:52\n",
      "CTGAAAG CTGAAAG:26 ATAAAAA:0\n",
      "CT CT:24 AA:0\n",
      "T T:84 TC:48\n",
      "GAAAATCATAACAATAAATT GAAAATCATAACAATAAATT:16 AAAAATATAAGAATAAATA:0\n",
      "TG TG:40 T:36\n",
      "C C:68 T:0\n",
      "C C:126 G:0\n",
      "T T:106 TG:98\n",
      "TC TC:56 T:22\n",
      "A A:62 AG:58\n",
      "TG TG:98 T:44\n",
      "CTGAAAG CTGAAAG:26 ATAAAAA:0\n",
      "CT CT:24 AA:0\n",
      "C C:102 CG:94\n",
      "T T:34 TC:14\n",
      "AAAGCTGGAGTG AAAGCTGGAGTG:44 GAAAGTAGGAGTA:28\n",
      "TT TT:114 AA:22\n",
      "A A:108 AAG:0\n",
      "T T:32 TA:0\n",
      "G G:68 A:0\n",
      "A A:94 AC:86\n",
      "A A:62 T:0\n",
      "G G:42 C:0\n",
      "GATTTTCCT GATTTTCCT:6 ATTTTTC:0\n",
      "GC GC:6 T:0\n",
      "GAA GAA:10 AT:0\n",
      "AAAGCTGGAGTG AAAGCTGGAGTG:44 GAAAGTAGGAGTA:28\n",
      "TT TT:114 AA:22\n",
      "A A:108 AAG:0\n",
      "T T:32 TA:0\n",
      "A A:62 AG:58\n",
      "TG TG:98 T:44\n",
      "CTGAAAG CTGAAAG:26 ATAAAAA:0\n",
      "CT CT:24 AA:0\n",
      "C C:102 CG:94\n",
      "T T:80 TC:16\n",
      "GAAAATCATAACAATAAATT GAAAATCATAACAATAAATT:16 AAAAATATAAGAATAAATA:0\n",
      "TG TG:34 T:14\n",
      "C C:70 T:0\n",
      "C C:112 G:0\n",
      "T T:114 TG:100\n",
      "TC TC:104 T:18\n",
      "GCATCATCCTCCTTTATTT GCATCATCCTCCTTTATTT:36 G:2\n",
      "A A:110 AG:96\n",
      "TG TG:150 T:60\n",
      "CTGAAAG CTGAAAG:38 ATAAAAA:0\n",
      "CT CT:38 AA:0\n",
      "T T:72 TC:38\n",
      "TACTAGTTA TACTAGTTA:74 GAATACTTG:4\n",
      "C C:108 T:4\n",
      "T T:116 C:4\n",
      "AAAGCTGGAGTG AAAGCTGGAGTG:80 GAAAGTAGGAGTA:4\n",
      "TT TT:162 AA:12\n",
      "A A:156 AAG:0\n",
      "T T:172 C:32\n",
      "T T:172 TA:0\n",
      "CG CG:114 C:76\n",
      "G G:114 A:0\n",
      "C C:134 CTTAAATTTT:18\n",
      "A A:92 T:0\n",
      "G G:70 C:0\n",
      "GA GA:36 AT:0\n",
      "CCT CCT:32 C:0\n",
      "A A:12 T:0\n",
      "GAA GAA:10 AT:0\n",
      "TACTAGTTA TACTAGTTA:74 GAATACTTG:4\n",
      "C C:108 T:4\n",
      "T T:116 C:4\n",
      "AAAGCTGGAGTG AAAGCTGGAGTG:80 GAAAGTAGGAGTA:4\n",
      "TT TT:162 AA:12\n",
      "A A:156 AAG:0\n",
      "T T:172 C:32\n",
      "T T:172 TA:0\n",
      "CG CG:114 C:76\n",
      "A A:110 AG:96\n",
      "TG TG:144 T:56\n",
      "CTGAAAG CTGAAAG:38 ATAAAAA:0\n",
      "CT CT:38 AA:0\n",
      "T T:150 TC:60\n",
      "TTC TTC:10 AT:0\n",
      "T T:12 A:0\n",
      "AGGAAAATCATAACAATAAATT AGGAAAATCATAACAATAAATT:32 GAAAAATATAAGAATAAATA:0\n",
      "C C:134 CTTAAAAAAT:20\n",
      "CGATGAAATAAAGGAGGAT CGATGAAATAAAGGAGGAT:36 C:0\n",
      "TG TG:102 T:18\n",
      "C C:114 T:0\n",
      "G G:6 T:0\n",
      "C C:130 G:0\n",
      "A A:78 AG:12\n",
      "TG TG:86 T:58\n",
      "CTGAAAG CTGAAAG:26 ATAAAAA:0\n",
      "CT CT:24 AA:0\n",
      "T T:24 TC:8\n",
      "AAAGCTGGAGTG AAAGCTGGAGTG:44 GAAAGTAGGAGTA:26\n",
      "TT TT:116 AA:14\n",
      "A A:108 AAG:0\n",
      "A A:74 AC:62\n",
      "T T:32 TA:0\n",
      "G G:66 A:0\n",
      "C C:82 CTTAAATTTT:46\n",
      "A A:68 T:0\n",
      "G G:44 C:0\n",
      "TG TG:22 T:0\n",
      "C C:16 T:8\n",
      "CCT CCT:6 C:0\n",
      "GAA GAA:8 AT:0\n",
      "AAAGCTGGAGTG AAAGCTGGAGTG:44 GAAAGTAGGAGTA:26\n",
      "TT TT:116 AA:14\n",
      "A A:108 AAG:0\n",
      "A A:74 AC:62\n",
      "T T:32 TA:0\n",
      "A A:78 AG:12\n",
      "TG TG:86 T:58\n",
      "CTGAAAG CTGAAAG:26 ATAAAAA:0\n",
      "CT CT:24 AA:0\n",
      "T T:82 TC:16\n",
      "TTC TTC:8 AT:0\n",
      "GAAAATCATAACAATAAATT GAAAATCATAACAATAAATT:16 AAAAATATAAGAATAAATA:0\n",
      "C C:82 CTTAAAAAAT:44\n",
      "C C:66 T:0\n",
      "C C:112 G:0\n",
      "TC TC:68 T:34\n",
      "A A:6 G:0\n",
      "A A:82 AG:50\n",
      "TG TG:88 T:66\n",
      "CTGAAAG CTGAAAG:26 ATAAAAA:0\n",
      "CT CT:24 AA:0\n",
      "AAAGCTGGAGTG AAAGCTGGAGTG:42 GAAAGTAGGAGTA:26\n",
      "TT TT:120 AA:18\n",
      "A A:106 AAG:0\n",
      "A A:88 AC:82\n",
      "T T:34 TA:0\n",
      "G G:66 A:0\n",
      "A A:66 T:0\n",
      "G G:46 C:0\n",
      "GAA GAA:8 AT:0\n",
      "AAAGCTGGAGTG AAAGCTGGAGTG:42 GAAAGTAGGAGTA:26\n",
      "TT TT:120 AA:18\n",
      "A A:106 AAG:0\n",
      "A A:88 AC:82\n",
      "T T:34 TA:0\n",
      "A A:82 AG:50\n",
      "TG TG:88 T:66\n",
      "CTGAAAG CTGAAAG:26 ATAAAAA:0\n",
      "CT CT:24 AA:0\n",
      "T T:80 TC:46\n",
      "GAAAATCATAACAATAAATT GAAAATCATAACAATAAATT:16 AAAAATATAAGAATAAATA:0\n",
      "T T:6 C:0\n",
      "TG TG:56 T:30\n",
      "A A:108 AG:106\n",
      "C C:66 T:0\n",
      "C C:122 G:0\n",
      "T T:28 C:18\n",
      "T T:106 TG:96\n",
      "TC TC:72 T:42\n",
      "A A:94 AG:48\n",
      "TG TG:94 T:80\n",
      "CTGAAAG CTGAAAG:26 ATAAAAA:0\n",
      "CT CT:22 AA:0\n",
      "AAAGCTGGAGTG AAAGCTGGAGTG:44 GAAAGTAGGAGTA:24\n",
      "TT TT:116 AA:16\n",
      "A A:106 AAG:0\n",
      "T T:34 TA:0\n",
      "G G:70 A:0\n",
      "A A:88 AC:42\n",
      "A A:70 T:0\n",
      "G G:46 C:0\n",
      "GATTTTCCT GATTTTCCT:6 ATTTTTC:0\n",
      "GAA GAA:8 AT:0\n",
      "AAAGCTGGAGTG AAAGCTGGAGTG:44 GAAAGTAGGAGTA:24\n",
      "TT TT:116 AA:16\n",
      "A A:106 AAG:0\n",
      "T T:34 TA:0\n",
      "A A:94 AG:48\n",
      "TG TG:94 T:80\n",
      "CTGAAAG CTGAAAG:26 ATAAAAA:0\n",
      "CT CT:22 AA:0\n",
      "T T:80 TC:48\n",
      "GAAAATCATAACAATAAATT GAAAATCATAACAATAAATT:20 AAAAATATAAGAATAAATA:0\n",
      "TG TG:60 T:36\n",
      "C C:70 T:0\n",
      "C C:122 G:0\n",
      "T T:36 C:16\n",
      "T T:110 TG:42\n",
      "A A:90 AG:44\n",
      "CTGAAAG CTGAAAG:26 ATAAAAA:0\n",
      "CT CT:24 AA:0\n",
      "AAAGCTGGAGTG AAAGCTGGAGTG:44 GAAAGTAGGAGTA:26\n",
      "TT TT:112 AA:14\n",
      "A A:106 AAG:0\n",
      "T T:32 TA:0\n",
      "G G:68 A:0\n",
      "A A:82 AC:80\n",
      "A A:64 T:0\n",
      "G G:46 C:0\n",
      "TG TG:20 T:0\n",
      "C C:18 T:10\n",
      "CCT CCT:6 C:0\n",
      "A A:6 T:0\n",
      "GAA GAA:12 AT:0\n",
      "AAAGCTGGAGTG AAAGCTGGAGTG:44 GAAAGTAGGAGTA:26\n",
      "TT TT:112 AA:14\n",
      "A A:106 AAG:0\n",
      "T T:32 TA:0\n",
      "A A:90 AG:44\n",
      "CTGAAAG CTGAAAG:26 ATAAAAA:0\n",
      "CT CT:24 AA:0\n",
      "T T:80 TC:50\n",
      "G G:18 A:10\n",
      "CATAACAATAAATT CATAACAATAAATT:20 ATAAGAATAAATA:0\n",
      "C C:68 T:0\n",
      "C C:112 G:0\n",
      "T T:110 TG:76\n",
      "AC AC:32 A:10\n",
      "TCC TCC:58 T:6\n",
      "GCATCATCCTCCTTTATTT GCATCATCCTCCTTTATTT:36 G:2\n",
      "C C:156 CG:84\n",
      "A A:134 AGG:0\n",
      "CTGAAAG CTGAAAG:38 ATAAAAA:0\n",
      "CT CT:36 AA:0\n",
      "A A:6 G:2\n",
      "TACTAGTTA TACTAGTTA:72 GAATACTTG:4\n",
      "C C:106 T:4\n",
      "T T:112 C:4\n",
      "AAAGCTGGAGTG AAAGCTGGAGTG:78 GAAAGTAGGAGTA:4\n",
      "TT TT:154 AA:14\n",
      "A A:150 AAG:0\n",
      "A A:114 AC:42\n",
      "C C:114 CG:64\n",
      "T T:160 TA:0\n",
      "C C:140 CG:20\n",
      "CG CG:84 C:18\n",
      "TG TG:146 T:0\n",
      "G G:114 A:0\n",
      "AC AC:124 A:28\n",
      "A A:140 AC:82\n",
      "A A:136 ACC:0\n",
      "A A:92 T:0\n",
      "G G:72 C:0\n",
      "GA GA:34 AT:0\n",
      "CCT CCT:32 C:0\n",
      "A A:8 T:0\n",
      "GAA GAA:6 AT:0\n",
      "TACTAGTTA TACTAGTTA:72 GAATACTTG:4\n",
      "C C:106 T:4\n",
      "T T:112 C:4\n",
      "AAAGCTGGAGTG AAAGCTGGAGTG:78 GAAAGTAGGAGTA:4\n",
      "TT TT:154 AA:14\n",
      "A A:150 AAG:0\n",
      "A A:122 AC:50\n",
      "C C:114 CG:64\n",
      "T T:164 TA:0\n",
      "C C:140 CG:20\n",
      "CG CG:78 C:18\n",
      "A A:134 AGG:0\n",
      "CTGAAAG CTGAAAG:38 ATAAAAA:0\n",
      "CT CT:36 AA:0\n",
      "A A:6 G:2\n",
      "TTC TTC:6 AT:0\n",
      "T T:8 A:0\n",
      "AGGAAAATCATAACAATAAATT AGGAAAATCATAACAATAAATT:32 GAAAAATATAAGAATAAATA:0\n",
      "CGATGAAATAAAGGAGGAT CGATGAAATAAAGGAGGAT:36 C:0\n",
      "TGG TGG:56 T:6\n",
      "AG AG:132 A:34\n",
      "C C:114 T:0\n",
      "G G:6 T:0\n",
      "C C:124 G:0\n",
      "TC TC:170 T:0\n",
      "A A:162 AC:94\n",
      "T T:158 TGG:0\n",
      "T T:146 TG:84\n",
      "GTCT GTCT:6 ATCA:2\n",
      "A A:10 G:9\n",
      "A A:43 G:39\n",
      "T T:42 C:36\n",
      "C C:9 CG:1\n",
      "AC AC:8 A:1\n",
      "G G:6 T:0\n",
      "A A:18 AC:0\n",
      "A A:5 G:3\n",
      "T T:5 A:4\n",
      "A A:12 AG:6\n",
      "A A:6 C:0\n",
      "G G:5 T:0\n",
      "T T:22 G:0\n",
      "T T:11 C:3\n",
      "G G:18 T:0\n",
      "T T:18 G:0\n",
      "G G:6 A:0\n",
      "CTTAG CTTAG:12 ATTAT:0\n",
      "G G:10 T:0\n",
      "C C:10 T:0\n",
      "A A:10 T:0\n",
      "GCG GCG:10 ACA:0\n",
      "C C:6 A:0\n",
      "A A:5 T:0\n",
      "C C:8 A:0\n",
      "CT CT:14 TA:0\n",
      "T T:12 A:0\n",
      "G G:10 T:0\n",
      "A A:11 T:0\n",
      "G G:5 T:0\n",
      "A A:6 C:0\n",
      "G G:9 T:0\n",
      "A A:8 AC:4\n",
      "G G:18 A:13\n",
      "T T:11 C:9\n",
      "A A:11 T:0\n",
      "C C:8 T:0\n",
      "G G:10 T:0\n",
      "G G:14 A:0\n",
      "A A:14 G:0\n",
      "A A:10 C:0\n",
      "AG AG:8 TA:0\n",
      "G G:8 A:0\n",
      "TGAGG TGAGG:7 GGAGA:0\n",
      "G G:7 A:0\n",
      "G G:6 T:0\n",
      "T T:16 C:14\n",
      "A A:9 T:0\n",
      "A A:10 C:0\n",
      "G G:8 T:0\n",
      "G G:6 T:0\n",
      "T T:8 A:0\n",
      "ATTTC ATTTC:14 CTTTT:0\n",
      "AT AT:15 GA:0\n",
      "C C:15 A:0\n",
      "C C:5 A:0\n",
      "T T:11 G:0\n",
      "A A:5 G:3\n",
      "A A:15 C:0\n",
      "C C:6 A:0\n",
      "TTG TTG:34 CTT:8\n",
      "AT AT:31 GC:12\n",
      "G G:32 A:13\n",
      "C C:8 A:0\n",
      "G G:7 T:0\n",
      "ATTTGTAA ATTTGTAA:6 TTTTTTAT:0\n",
      "C C:5 T:0\n",
      "G G:7 A:0\n",
      "A A:7 C:0\n",
      "CTGAAAG CTGAAAG:26 ATAAAAA:0\n",
      "CT CT:24 AA:0\n",
      "AAAGCTGGAGTG AAAGCTGGAGTG:44 GAAAGTAGGAGTA:26\n",
      "TT TT:116 AA:2\n",
      "A A:108 AAG:0\n",
      "T T:32 TA:0\n",
      "G G:70 A:0\n",
      "C C:82 CTTAAATTTT:64\n",
      "A A:64 T:0\n",
      "G G:46 C:0\n",
      "GC GC:6 T:0\n",
      "GAA GAA:10 AT:0\n",
      "AAAGCTGGAGTG AAAGCTGGAGTG:44 GAAAGTAGGAGTA:26\n",
      "TT TT:116 AA:2\n",
      "A A:108 AAG:0\n",
      "T T:32 TA:0\n",
      "CTGAAAG CTGAAAG:26 ATAAAAA:0\n",
      "CT CT:24 AA:0\n",
      "T T:82 TC:60\n",
      "GAAAATCATAACAATAAATT GAAAATCATAACAATAAATT:16 AAAAATATAAGAATAAATA:0\n",
      "C C:82 CTTAAAAAAT:60\n",
      "C C:70 T:0\n",
      "C C:126 G:0\n",
      "T T:32 C:20\n",
      "A A:8 G:0\n",
      "A A:76 AG:26\n",
      "TG TG:84 T:62\n",
      "CTGAAAG CTGAAAG:26 ATAAAAA:0\n",
      "CT CT:24 AA:0\n",
      "T T:42 TC:24\n",
      "AAAGCTGGAGTG AAAGCTGGAGTG:42 GAAAGTAGGAGTA:24\n",
      "TT TT:120 AA:18\n",
      "A A:110 AAG:0\n",
      "A A:80 AC:78\n",
      "T T:34 TA:0\n",
      "G G:66 A:0\n",
      "A A:116 AC:116\n",
      "A A:88 AC:80\n",
      "A A:66 T:0\n",
      "G G:42 C:2\n",
      "GAA GAA:8 AT:0\n",
      "AAAGCTGGAGTG AAAGCTGGAGTG:42 GAAAGTAGGAGTA:24\n",
      "TT TT:120 AA:18\n",
      "A A:110 AAG:0\n",
      "A A:80 AC:78\n",
      "T T:34 TA:0\n",
      "A A:76 AG:26\n",
      "TG TG:84 T:62\n",
      "CTGAAAG CTGAAAG:26 ATAAAAA:0\n",
      "CT CT:24 AA:0\n",
      "T T:82 TC:40\n",
      "TTC TTC:8 AT:0\n",
      "GAAAATCATAACAATAAATT GAAAATCATAACAATAAATT:16 AAAAATATAAGAATAAATA:0\n",
      "T T:8 C:0\n",
      "A A:112 AG:110\n",
      "C C:66 T:0\n",
      "C C:122 G:0\n",
      "T T:28 C:18\n",
      "T T:108 TG:88\n",
      "T T:6 G:0\n",
      "A A:88 AG:42\n",
      "TG TG:90 T:76\n",
      "CTGAAAG CTGAAAG:26 ATAAAAA:0\n",
      "CT CT:24 AA:0\n",
      "AAAGCTGGAGTG AAAGCTGGAGTG:44 GAAAGTAGGAGTA:26\n",
      "TT TT:116 AA:2\n",
      "A A:108 AAG:0\n",
      "T T:34 TA:0\n",
      "G G:78 A:0\n",
      "A A:68 T:0\n",
      "G G:44 C:0\n",
      "GATTTTCCT GATTTTCCT:6 ATTTTTC:0\n",
      "GAA GAA:10 AT:0\n",
      "AAAGCTGGAGTG AAAGCTGGAGTG:44 GAAAGTAGGAGTA:26\n",
      "TT TT:116 AA:2\n",
      "A A:108 AAG:0\n",
      "T T:34 TA:0\n",
      "A A:88 AG:42\n",
      "TG TG:90 T:76\n",
      "CTGAAAG CTGAAAG:26 ATAAAAA:0\n",
      "CT CT:24 AA:0\n",
      "T T:82 TC:68\n",
      "GAAAATCATAACAATAAATT GAAAATCATAACAATAAATT:16 AAAAATATAAGAATAAATA:0\n",
      "T T:6 G:0\n",
      "TG TG:72 T:42\n",
      "C C:78 T:0\n",
      "C C:134 G:0\n",
      "A A:6 G:0\n",
      "A A:86 AG:40\n",
      "CTGAAAG CTGAAAG:26 ATAAAAA:0\n",
      "CT CT:26 AA:0\n",
      "TA TA:6 T:0\n",
      "A A:6 G:0\n",
      "GA GA:6 TG:0\n",
      "AAAGCTGGAGTG AAAGCTGGAGTG:42 GAAAGTAGGAGTA:30\n",
      "TT TT:116 AA:0\n",
      "A A:104 AAG:0\n",
      "A A:82 AC:68\n",
      "T T:32 TA:0\n",
      "G G:68 A:0\n",
      "C C:88 CTTAAATTTT:0\n",
      "A A:68 T:0\n",
      "G G:42 C:0\n",
      "TG TG:22 T:0\n",
      "C C:16 T:8\n",
      "CCT CCT:6 C:0\n",
      "GAA GAA:8 AT:0\n",
      "AAAGCTGGAGTG AAAGCTGGAGTG:42 GAAAGTAGGAGTA:30\n",
      "TT TT:116 AA:0\n",
      "A A:104 AAG:0\n",
      "A A:82 AC:68\n",
      "T T:32 TA:0\n",
      "A A:86 AG:40\n",
      "CTGAAAG CTGAAAG:26 ATAAAAA:0\n",
      "CT CT:26 AA:0\n",
      "TA TA:6 T:0\n",
      "A A:6 G:0\n",
      "GA GA:6 TG:0\n",
      "T T:78 TC:58\n",
      "G G:16 A:8\n",
      "CATAACAATAAATT CATAACAATAAATT:22 ATAAGAATAAATA:0\n",
      "C C:88 CTTAAAAAAT:0\n",
      "T T:6 C:0\n",
      "C C:68 T:0\n",
      "C C:122 G:0\n",
      "GCATCATCCTCCTTTATTT GCATCATCCTCCTTTATTT:36 G:2\n",
      "CGG CGG:114 C:6\n",
      "A A:172 AT:2\n",
      "CA CA:160 C:0\n",
      "A A:142 AG:34\n",
      "CTGAAAG CTGAAAG:40 ATAAAAA:0\n",
      "CT CT:38 AA:0\n",
      "TACTAGTTA TACTAGTTA:72 GAATACTTG:4\n",
      "C C:106 T:4\n",
      "T T:114 C:4\n",
      "AAAGCTGGAGTG AAAGCTGGAGTG:78 GAAAGTAGGAGTA:4\n",
      "TT TT:158 AA:14\n",
      "A A:152 AAG:0\n",
      "A A:164 AG:20\n",
      "T T:174 TA:0\n",
      "C C:156 CG:32\n",
      "A A:154 AG:106\n",
      "G G:116 A:0\n",
      "AC AC:158 A:80\n",
      "A A:140 AC:64\n",
      "A A:94 T:0\n",
      "G G:74 C:0\n",
      "GA GA:34 AT:0\n",
      "CCT CCT:32 C:0\n",
      "A A:10 T:0\n",
      "GAA GAA:8 AT:0\n",
      "TACTAGTTA TACTAGTTA:72 GAATACTTG:4\n",
      "C C:106 T:4\n",
      "T T:114 C:4\n",
      "AAAGCTGGAGTG AAAGCTGGAGTG:78 GAAAGTAGGAGTA:4\n",
      "TT TT:158 AA:14\n",
      "A A:152 AAG:0\n",
      "A A:164 AG:20\n",
      "T T:180 TA:0\n",
      "C C:156 CG:32\n",
      "A A:154 AG:106\n",
      "A A:142 AG:34\n",
      "CTGAAAG CTGAAAG:40 ATAAAAA:0\n",
      "CT CT:38 AA:0\n",
      "TTC TTC:8 AT:0\n",
      "T T:10 A:0\n",
      "AGGAAAATCATAACAATAAATT AGGAAAATCATAACAATAAATT:30 GAAAAATATAAGAATAAATA:0\n",
      "A A:184 AT:2\n",
      "CA CA:160 C:0\n",
      "CGATGAAATAAAGGAGGAT CGATGAAATAAAGGAGGAT:36 C:0\n",
      "AG AG:152 A:80\n",
      "C C:116 T:0\n",
      "G G:6 T:0\n",
      "C C:126 G:0\n",
      "ACC ACC:130 A:6\n",
      "T T:164 TG:74\n",
      "TC TC:54 T:26\n",
      "A A:86 AG:36\n",
      "TG TG:88 T:50\n",
      "CTGAAAG CTGAAAG:26 ATAAAAA:0\n",
      "CT CT:26 AA:0\n",
      "T T:44 TC:18\n",
      "AAAGCTGGAGTG AAAGCTGGAGTG:44 GAAAGTAGGAGTA:24\n",
      "TT TT:112 AA:2\n",
      "A A:106 AAG:0\n",
      "T T:34 TA:0\n",
      "G G:68 A:0\n",
      "A A:88 AC:80\n",
      "A A:66 T:0\n",
      "G G:46 C:0\n",
      "GATTTTCCT GATTTTCCT:6 ATTTTTC:0\n",
      "AATGC AATGC:6 TATT:0\n",
      "GAA GAA:8 AT:0\n",
      "AAAGCTGGAGTG AAAGCTGGAGTG:44 GAAAGTAGGAGTA:24\n",
      "TT TT:112 AA:2\n",
      "A A:106 AAG:0\n",
      "T T:34 TA:0\n",
      "A A:86 AG:36\n",
      "TG TG:88 T:50\n",
      "CTGAAAG CTGAAAG:26 ATAAAAA:0\n",
      "CT CT:26 AA:0\n",
      "T T:78 TC:18\n",
      "GAAAATCATAACAATAAATT GAAAATCATAACAATAAATT:16 AAAAATATAAGAATAAATA:0\n",
      "TG TG:36 T:28\n",
      "C C:68 T:0\n",
      "C C:114 G:0\n",
      "T T:110 TG:86\n",
      "TC TC:40 T:12\n",
      "A A:40 AG:12\n",
      "TG TG:80 T:42\n",
      "CTGAAAG CTGAAAG:26 ATAAAAA:0\n",
      "CT CT:24 AA:0\n",
      "T T:62 TC:14\n",
      "AAAGCTGGAGTG AAAGCTGGAGTG:44 GAAAGTAGGAGTA:24\n",
      "TT TT:112 AA:14\n",
      "A A:106 AAG:0\n",
      "T T:32 TA:0\n",
      "G G:66 A:0\n",
      "A A:86 AC:30\n",
      "A A:62 T:0\n",
      "G G:42 C:0\n",
      "GATTTTCCT GATTTTCCT:6 ATTTTTC:0\n",
      "GAA GAA:8 AT:0\n",
      "AAAGCTGGAGTG AAAGCTGGAGTG:44 GAAAGTAGGAGTA:24\n",
      "TT TT:112 AA:14\n",
      "A A:106 AAG:0\n",
      "T T:32 TA:0\n",
      "A A:40 AG:12\n",
      "TG TG:80 T:42\n",
      "CTGAAAG CTGAAAG:26 ATAAAAA:0\n",
      "CT CT:24 AA:0\n",
      "T T:104 TC:20\n",
      "TTC TTC:8 AT:0\n",
      "GAAAATCATAACAATAAATT GAAAATCATAACAATAAATT:16 AAAAATATAAGAATAAATA:0\n",
      "TG TG:34 T:14\n",
      "C C:66 T:0\n",
      "C C:110 G:0\n",
      "T T:106 TG:34\n",
      "TC TC:32 T:4\n",
      "T T:6 G:0\n",
      "TG TG:82 T:44\n",
      "CTGAAAG CTGAAAG:26 ATAAAAA:0\n",
      "CT CT:24 AA:0\n",
      "A A:6 G:0\n",
      "A A:6 T:0\n",
      "A A:10 G:0\n",
      "T T:32 TC:16\n",
      "AAAGCTGGAGTG AAAGCTGGAGTG:44 GAAAGTAGGAGTA:26\n",
      "TT TT:120 AA:2\n",
      "A A:110 AAG:0\n",
      "A A:94 AC:14\n",
      "T T:34 TA:0\n",
      "C C:74 CG:46\n",
      "A A:88 AG:72\n",
      "G G:70 A:0\n",
      "A A:116 AC:108\n",
      "A A:64 T:0\n",
      "G G:44 C:0\n",
      "GATTTTCCT GATTTTCCT:6 ATTTTTC:0\n",
      "GAA GAA:8 AT:0\n",
      "AAAGCTGGAGTG AAAGCTGGAGTG:44 GAAAGTAGGAGTA:26\n",
      "TT TT:120 AA:2\n",
      "A A:110 AAG:0\n",
      "A A:94 AC:14\n",
      "T T:34 TA:0\n",
      "C C:74 CG:46\n",
      "A A:88 AG:72\n",
      "TG TG:82 T:44\n",
      "CTGAAAG CTGAAAG:26 ATAAAAA:0\n",
      "CT CT:24 AA:0\n",
      "A A:6 G:0\n",
      "A A:6 T:0\n",
      "A A:10 G:0\n",
      "T T:80 TC:18\n",
      "TTC TTC:8 AT:0\n",
      "GAAAATCATAACAATAAATT GAAAATCATAACAATAAATT:16 AAAAATATAAGAATAAATA:0\n",
      "T T:6 G:0\n",
      "TG TG:30 T:4\n",
      "A A:110 AG:106\n",
      "C C:70 T:0\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "C C:112 G:0\n",
      "T T:30 C:28\n",
      "A A:15 G:0\n",
      "C C:14 G:0\n",
      "G G:6 A:6\n",
      "A A:10 T:0\n",
      "A A:12 C:1\n",
      "C C:6 A:0\n",
      "C C:11 T:5\n",
      "T T:6 A:3\n",
      "(924, 17)\n",
      "                  CHROM     POS     TYPE      REF                  ALT  \\\n",
      "0  CG8421-NZ_CP005388-x  149010      ins        G  GCATCATCCTCCTTTATTT   \n",
      "1  CG8421-NZ_CP005388-x  565751      del       AG                    A   \n",
      "2  CG8421-NZ_CP005388-x  649500  complex  ATAAAAA              CTGAAAG   \n",
      "3  CG8421-NZ_CP005388-x  649511  complex       AA                   CT   \n",
      "4  CG8421-NZ_CP005388-x  649530  complex   GTTATG               AATAGA   \n",
      "\n",
      "                     EVIDENCE FTYPE STRAND NT_POS AA_POS EFFECT LOCUS_TAG  \\\n",
      "0  GCATCATCCTCCTTTATTT:34 G:2   NaN    NaN    NaN    NaN    NaN       NaN   \n",
      "1                 A:148 AG:10   NaN    NaN    NaN    NaN    NaN       NaN   \n",
      "2        CTGAAAG:42 ATAAAAA:0   NaN    NaN    NaN    NaN    NaN       NaN   \n",
      "3                  CT:40 AA:0   NaN    NaN    NaN    NaN    NaN       NaN   \n",
      "4           AATAGA:6 GTTATG:0   NaN    NaN    NaN    NaN    NaN       NaN   \n",
      "\n",
      "  GENE PRODUCT ref_assem  mapped_lib  evidence_pct  \n",
      "0  NaN     NaN    CG8421  crofts_8R1      0.944444  \n",
      "1  NaN     NaN    CG8421  crofts_8R1      0.936709  \n",
      "2  NaN     NaN    CG8421  crofts_8R1      1.000000  \n",
      "3  NaN     NaN    CG8421  crofts_8R1      1.000000  \n",
      "4  NaN     NaN    CG8421  crofts_8R1      1.000000  \n"
     ]
    },
    {
     "ename": "AssertionError",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mAssertionError\u001b[0m                            Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-55-9782ae77fbf0>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m     28\u001b[0m \u001b[0;31m# determine whether it matches the location and type of a recorded mutation\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     29\u001b[0m \u001b[0;32mfor\u001b[0m \u001b[0manch_idx\u001b[0m \u001b[0;32min\u001b[0m \u001b[0me_anchor_df\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mindex\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 30\u001b[0;31m     \u001b[0;32massert\u001b[0m \u001b[0;32mFalse\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     31\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     32\u001b[0m \u001b[0;32mfor\u001b[0m \u001b[0mres_row\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mres_df\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mindex\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mAssertionError\u001b[0m: "
     ]
    }
   ],
   "source": [
    "# clean up the blast hits \n",
    "select_blast = blast_df[blast_df['sseqid'].isin(contigs_identified)]\n",
    "\n",
    "# make variant types match \n",
    "e_anchor_df['var_type'] = e_anchor_df['Variant Type'].map({'Insertion':'ins', 'Deletion':'del', 'SNV':'snp'})\n",
    "\n",
    "def calc_evidence(x):\n",
    "    res_alt, res_evid = x\n",
    "    print(res_alt, res_evid)\n",
    "    evid_nums = [int(i.split(\":\")[-1].strip()) for i in res_evid.split()]\n",
    "    assert res_alt == res_evid.split(\":\")[0].strip()\n",
    "    return float(evid_nums[0])/float(sum(evid_nums))\n",
    "\n",
    "# make one result df\n",
    "res_pile = []\n",
    "for a_res, res_dfx in results_df.items():\n",
    "    ref, query = list(mapping_pairs2.loc[a_res, ['ref_assem', 'mapped_lib']])\n",
    "    res_df = res_dfx.copy()\n",
    "    res_df['ref_assem'] = pd.Series(index=res_df.index, data=[ref]*len(res_df))\n",
    "    res_df['mapped_lib'] = pd.Series(index=res_df.index, data=[query]*len(res_df))\n",
    "    res_pile.append(res_df)\n",
    "\n",
    "all_results = pd.concat(res_pile, axis=0, ignore_index=True, verify_integrity=True)\n",
    "all_results['evidence_pct'] = all_results[['ALT', 'EVIDENCE']].apply(calc_evidence, axis=1)\n",
    "print(all_results.shape)\n",
    "print(all_results.head())\n",
    "\n",
    "# determine whether it matches the location and type of a recorded mutation \n",
    "for anch_idx in e_anchor_df.index:\n",
    "    all_results['mut_start'] \n",
    "    assert False\n",
    "    \n",
    "for res_row in res_df.index:\n",
    "    res_cntg, res_pos = res_df.loc[res_row, ['CHROM', 'POS']]\n",
    "    q1 = select_blast['sseqid'] == res_cntg\n",
    "    q2 = select_blast['sstart2'] <= res_pos\n",
    "    q3 = select_blast['send2'] >= res_pos\n",
    "\n",
    "    subhit_df = select_blast[((q1) & (q2) & (q3))]\n",
    "    if (subhit_df.shape[0] > 0) and (evidence_pct > 0.25) and ref == 'CG8421':\n",
    "\n",
    "        \n",
    "\n",
    "\n",
    "        # confirm whether its upstream, downstream, or in CDS based on primer name \n",
    "\n",
    "        # \n",
    "\n",
    "        # \n",
    "        print(subhit_df.T)\n",
    "        assert False\n",
    "        res_df.loc[res_row, 'primer_match'] = \", \".join(list(subhit_df['primer']))\n",
    "        l_cols2 = ['primer', 'send', 'sstart', 'qstart', 'qend', 'qlen']\n",
    "\n",
    "    # add \n",
    "    if ref == 'CG8421':\n",
    "        possible_affected = []\n",
    "        # if its within a CDS\n",
    "        within = (ngb_df['start'] <= res_pos) & (ngb_df['end'] >= res_pos)\n",
    "\n",
    "        # closest downstream watson strand encoded gene \n",
    "        downstream = (ngb_df['strand'] == 1) & ((ngb_df['start'] - res_pos) > 0)\n",
    "        if downstream.sum() > 0:\n",
    "            min_start = (ngb_df.loc[downstream, 'start'] - res_pos).astype(int).idxmin()\n",
    "            dist_to_start = ngb_df.loc[min_start, 'start'] - res_pos\n",
    "            res_df.loc[res_row, 'nearest_downstream'] = ngb_df.loc[min_start, 'locus_tag']\n",
    "            res_df.loc[res_row, 'dist_downstream'] = dist_to_start\n",
    "            if abs(dist_to_start) <= 50:\n",
    "                 possible_affected.append(res_df.loc[res_row, 'nearest_downstream'])\n",
    "\n",
    "        # closest upstream crick strand encoded gene\n",
    "        upstream = (ngb_df['strand'] == -1) & ((res_pos - ngb_df['end']) > 0)\n",
    "        if upstream.sum() > 0:\n",
    "            min_end = (res_pos - ngb_df.loc[upstream, 'end']).astype(int).idxmin()\n",
    "            dist_to_end = res_pos - ngb_df.loc[min_start, 'end']\n",
    "            res_df.loc[res_row, 'nearest_upstream'] = ngb_df.loc[min_end, 'locus_tag']\n",
    "            res_df.loc[res_row, 'dist_upstream'] = dist_to_end\n",
    "            if abs(dist_to_end) <= 50:\n",
    "                 possible_affected.append(res_df.loc[res_row, 'nearest_upstream'])\n",
    "\n",
    "        if within.sum() > 0:\n",
    "            possible_affected += list(ngb_df.loc[within, 'locus_tag'])\n",
    "            res_df.loc[res_row, 'affected_genes'] = \", \".join(possible_affected)\n",
    "            res_df.loc[res_row, 'within_gene'] = \", \".join(list(ngb_df.loc[within, 'locus_tag']))\n",
    "            #print(\"\\t\", res_row, res_cntg, res_pos, res_df.loc[res_row, 'within_gene'], list(ngb_df.loc[within, ['start', 'end', 'gene', 'product']].values))\n",
    "\n",
    "        aff_gene_cntr.update(possible_affected)\n",
    "\n",
    "results_df2[a_res] = res_df.copy()\n",
    "print(\"mapping {} to {}, gives {} matched variants {} affected genes\".format(ref, query, hit_counter, \n",
    "                                                                             len(aff_gene_cntr)))\n",
    "print(\"\\t{}\\n\".format(aff_gene_cntr))\n",
    "          \n",
    "\n",
    "drp_cols = ['AA_POS', 'EFFECT', 'FTYPE', 'GENE', 'LOCUS_TAG', 'NT_POS', 'PRODUCT', 'STRAND']\n",
    "snp_df = pd.concat(list(results_df2.values()), axis=0, ignore_index=True, sort=True).drop(drp_cols, axis=1)\n",
    "good_snps_df = snp_df.copy()[snp_df['primer_match'].notnull()]\n",
    "good_snps_df['primer_set'] = good_snps_df['primer_match'].apply(lambda x: set(x.split(\", \")))\n",
    "good_snps_df.sample(10)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "substandard_repeats = []\n",
    "for primer in sim_check_df.primer.unique():\n",
    "    matched_locations = sim_check_df[sim_check_df['primer'] == primer]\n",
    "    pident_adj = matched_locations['pident'] * ((matched_locations['qend'] - matched_locations['qstart'])/matched_locations['qend'])\n",
    "    best_locations = matched_locations[pident_adj == pident_adj.max()]\n",
    "    substandard_repeats += list(set(matched_locations.index) - set(best_locations.index))\n",
    "\n",
    "sim_check_df2 = sim_check_df.drop(substandard_repeats, axis=0)\n",
    "\n",
    "def convert_sim_locs(x):\n",
    "    sscdf = sim_check_df2[sim_check_df2['primer'] == x[0]]\n",
    "    if sscdf.shape[0] > 1:\n",
    "        return np.nan\n",
    "    elif sscdf.shape[0] == 1:\n",
    "        assert sscdf['qstart'].values[0] < sscdf['qend'].values[0]\n",
    "        if sscdf['sstart'].values[0] > sscdf['send'].values[0]:\n",
    "            return sscdf['sstart'].values[0]  - (x[1] - sscdf['qstart'].values[0])\n",
    "        else:\n",
    "            return sscdf['sstart'].values[0]  + (x[1] - sscdf['qstart'].values[0])\n",
    "    else:\n",
    "        raise ValueError(\"undefined contig mapped\")\n",
    "\n",
    "# \n",
    "falsePos, falseNeg, tru_pos = 0, 0, 0\n",
    "for cg in crofts_genomes:\n",
    "    this_lib = snp_df[snp_df['mapped_lib'] == cg]\n",
    "    positiveHits = this_lib[this_lib['ref_assem'] == 'CG8421']\n",
    "    print(cg)\n",
    "    questionableHits = this_lib.copy()[this_lib['ref_assem'] == 'CG8421_sim']\n",
    "    questionableHits['contig'] = questionableHits['CHROM'].apply(lambda x: x.split(\"-\")[1])\n",
    "    questionableHits['real_pos'] = questionableHits[['contig', 'POS']].apply(convert_sim_locs, axis=1)\n",
    "    print(questionableHits[['POS', 'real_pos', 'CG8421_POS', 'primer_match']])\n",
    "    questionableHits.sort_values(by='real_pos', inplace=True)\n",
    "    hitpool = set(positiveHits.index)\n",
    "    sourcepool = set(questionableHits.index)\n",
    "    for qh in questionableHits.index:\n",
    "        hit_dist = abs(positiveHits.loc[hitpool, 'POS'] - questionableHits.loc[qh, 'real_pos'])\n",
    "        closest_var = hit_dist[hit_dist == hit_dist.min()]\n",
    "        if len(closest_var) > 0:\n",
    "            closest_pos = positiveHits.loc[closest_var.index[0], 'POS']\n",
    "            other_candidates = [abs(questionableHits.loc[sp, 'real_pos']-closest_pos) for sp in sourcepool]\n",
    "            print(len(sourcepool), len(hitpool), closest_var.min(), min(other_candidates), closest_pos)        \n",
    "            if (closest_var.min() < 23) and (closest_var.min() <= min(other_candidates)):\n",
    "                hitpool.remove(closest_var.index[0])\n",
    "                assert len(closest_var) == 1\n",
    "                tru_pos += 1\n",
    "            else:\n",
    "                falsePos += 1\n",
    "        else:\n",
    "            falsePos += 1\n",
    "    \n",
    "        sourcepool.remove(qh)\n",
    "        \n",
    "    falseNeg += len(hitpool)\n",
    "    print(\"{} CG8421 {}, Sim {}, Left Over {}, TP {}, FP {}, FN {}\".format(cg, len(questionableHits), len(positiveHits), len(hitpool), \n",
    "                                                                           tru_pos, falsePos, falseNeg))\n",
    "\n",
    "precision = tru_pos/ (tru_pos + falsePos)\n",
    "recall = tru_pos/ (tru_pos + falseNeg)\n",
    "f1score=  2*(precision*recall)/(precision+recall)\n",
    "\n",
    "print(\"Using assemblies is has a F1 score of {:4f}\".format(f1score))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "primer_to_var_region = {}\n",
    "# map the region back to the variant index\n",
    "for e_ix, h, mr in e_anchor_df[['headers', 'mutation_rows']].reset_index().values:\n",
    "    primer_to_var_region[h] = set([e_ix])\n",
    "    \n",
    "# map the gene to the variation to the variant index\n",
    "for h, mr, sr in g_anchor_df[['headers', 'mutation_rows', 'support_rows']].values:\n",
    "    if mr != []:\n",
    "        e_ix = list(e_anchor_df[e_anchor_df['mutation_rows'].apply(lambda x: len(set(x).intersection(set(mr))) > 0 )].index)\n",
    "        primer_to_var_region[h] = set(e_ix)\n",
    "\n",
    "all_primers_possible = set(primer_to_var_region.keys())\n",
    "ontarget_hits = good_snps_df['primer_set'].apply(lambda x: len(x.intersection(all_primers_possible)) > 0)\n",
    "hit_snps = good_snps_df[ ontarget_hits & (good_snps_df['ref_assem'] == 'CG8421')]\n",
    "\n",
    "snp_locs = {i:set() for i in e_anchor_df.index}\n",
    "snp_populations = {i:set() for i in e_anchor_df.index}\n",
    "snp_ref = {i:set() for i in e_anchor_df.index}\n",
    "\n",
    "for ix in good_snps_df.index:\n",
    "    if good_snps_df.loc[ix, 'ref_assem'] != 'CG8421_sim':\n",
    "        these_matches = good_snps_df.loc[ix, 'primer_set']\n",
    "        these_anchors = [primer_to_var_region[i] for i in these_matches if i in primer_to_var_region.keys()]\n",
    "        these_anchors = set([i for j in these_anchors for i in j])\n",
    "        for anch in these_anchors:\n",
    "            snp_locs[anch].update([good_snps_df.loc[ix, 'CG8421_POS']])    \n",
    "            snp_populations[anch].update([good_snps_df.loc[ix, 'mapped_lib']])\n",
    "            snp_ref[anch].update([good_snps_df.loc[ix, 'ref_assem']])\n",
    "        \n",
    "flipped_df['SNPs_positions'] = pd.Series(snp_locs)\n",
    "flipped_df['SNPd_populations'] = pd.Series(snp_populations)\n",
    "flipped_df['SNP_references'] = pd.Series(snp_ref)\n",
    "\n",
    "group_keys = [['crofts_innoc'] , primary_crofts, relapse_crofts, peru_relapse]\n",
    "group_names = ['Innoc', 'Crofts_Primary', 'Crofts_Relapse', 'Iquitos_Relapse']\n",
    "new_col_names = [gn+\"_w_Variant\" for gn in group_names]\n",
    "new_col_names2 = [gn+\"_wo_Variant\" for gn in group_names]\n",
    "\n",
    "for gk, ncn, ncn2 in zip(group_keys, new_col_names, new_col_names2):\n",
    "    flipped_df[ncn2] = pd.Series({k:len(set(gk) - set(v)) for k, v in snp_populations.items()})\n",
    "    flipped_df[ncn] = pd.Series({k:len(set(gk).intersection(set(v))) for k, v in snp_populations.items()})\n",
    "\n",
    "expectation_cols = ['Matched_to_Crofts_Data', 'Missing_in_our_Data', 'Misreported_by_them']\n",
    "for ec in expectation_cols:\n",
    "    flipped_df[ec] = pd.Series(index=flipped_df.index)\n",
    "\n",
    "for ix in flipped_df.index:\n",
    "    snpd_pops = flipped_df.loc[ix, 'SNPd_populations']\n",
    "    pops_expctd = e_anchor_df.loc[ix, 'Samps_w_Mutation']\n",
    "    pops_not_expctd = e_anchor_df.loc[ix, 'Samps_off_Target']\n",
    "    flipped_df.loc[ix, expectation_cols[0]] = len(pops_expctd.intersection(snpd_pops))\n",
    "    flipped_df.loc[ix, expectation_cols[1]] = len(pops_expctd - snpd_pops)\n",
    "    flipped_df.loc[ix, expectation_cols[2]] = len(pops_not_expctd.intersection(snpd_pops))\n",
    "\n",
    "print(flipped_df[expectation_cols].sum())\n",
    "\n",
    "flipped_df.to_csv('/Volumes/KeithSSD/variant-calling/misc_data/Region_and_Snp_Data.txt', sep=\"\\t\")\n",
    "\n",
    "#valid_snps = hit_snps[['mapped_lib', 'primer_match', 'ref_assem']].groupby(['mapped_lib', 'primer_match'])\n",
    "#sim_and_not = lambda x: 'CG8421' in list(x) and 'CG8421_sim' in list(x)\n",
    "#print(valid_snps.agg(set))#.astype(int).sum().values[0])\n",
    "\n",
    "disp_cols = sorted(['SNPs_positions', 'SNPd_populations', 'SNP_references']+new_col_names + new_col_names2)\n",
    "flipped_df[disp_cols]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "good_snps_df[good_snps_df['mapped_lib'] == 'crofts_1D2'].sort_values(by='POS')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(e_anchor_df.columns)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
